// @instantdb/core/dist/standalone/index.js
var Kn = Object.defineProperty;
var Qn = (t, e, n) => e in t ? Kn(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n;
var y = (t, e, n) => Qn(t, typeof e != "symbol" ? e + "" : e, n);
function U(t) {
  if (typeof t == "number")
    return (Math.abs(t * 2654435761) >>> 0).toString(16);
  if (typeof t == "boolean") return t ? "1" : "0";
  if (t === null) return "null";
  if (t === void 0) return "undefined";
  if (typeof t == "string") {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++)
      e ^= t.charCodeAt(n), e += (e << 1) + (e << 4) + (e << 7) + (e << 8) + (e << 24), e = e >>> 0;
    return e.toString(16);
  }
  if (Array.isArray(t)) {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++) {
      e ^= (n + 1) * 2654435761;
      const r = U(t[n]);
      for (let s = 0; s < r.length; s++)
        e ^= r.charCodeAt(s), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  if (typeof t == "object") {
    let e = 2166136261;
    const n = Object.keys(t).sort();
    for (let r = 0; r < n.length; r++) {
      const s = n[r], i = U(s);
      e ^= parseInt(i, 16), e *= 16777619, e = e >>> 0;
      const o = U(t[s]);
      e ^= parseInt(o, 16), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  return U(String(t));
}
const E = {
  Remove: "remove",
  Replace: "replace",
  Add: "add"
}, rn = Symbol.for("__MUTATIVE_PROXY_DRAFT__"), Gn = Symbol("__MUTATIVE_RAW_RETURN_SYMBOL__"), Oe = Symbol.iterator, L = {
  mutable: "mutable",
  immutable: "immutable"
}, ct = {};
function he(t, e) {
  return t instanceof Map ? t.has(e) : Object.prototype.hasOwnProperty.call(t, e);
}
function kt(t, e) {
  if (e in t) {
    let n = Reflect.getPrototypeOf(t);
    for (; n; ) {
      const r = Reflect.getOwnPropertyDescriptor(n, e);
      if (r)
        return r;
      n = Reflect.getPrototypeOf(n);
    }
  }
}
function ut(t) {
  return Object.getPrototypeOf(t) === Set.prototype;
}
function ft(t) {
  return Object.getPrototypeOf(t) === Map.prototype;
}
function x(t) {
  var e;
  return (e = t.copy) !== null && e !== void 0 ? e : t.original;
}
function te(t) {
  return !!g(t);
}
function g(t) {
  return typeof t != "object" ? null : t == null ? void 0 : t[rn];
}
function lt(t) {
  var e;
  const n = g(t);
  return n ? (e = n.copy) !== null && e !== void 0 ? e : n.original : t;
}
function V(t, e) {
  if (!t || typeof t != "object")
    return !1;
  let n;
  return Object.getPrototypeOf(t) === Object.prototype || Array.isArray(t) || t instanceof Map || t instanceof Set || !!(e != null && e.mark) && ((n = e.mark(t, L)) === L.immutable || typeof n == "function");
}
function sn(t, e = []) {
  if (Object.hasOwnProperty.call(t, "key")) {
    const n = t.parent.copy, r = g(B(n, t.key));
    if (r !== null && (r == null ? void 0 : r.original) !== t.original)
      return null;
    const s = t.parent.type === 3, i = s ? Array.from(t.parent.setMap.keys()).indexOf(t.key) : t.key;
    if (!(s && n.size > i || he(n, i)))
      return null;
    e.push(i);
  }
  if (t.parent)
    return sn(t.parent, e);
  e.reverse();
  try {
    Bn(t.copy, e);
  } catch {
    return null;
  }
  return e;
}
function ne(t) {
  return Array.isArray(t) ? 1 : t instanceof Map ? 2 : t instanceof Set ? 3 : 0;
}
function B(t, e) {
  return ne(t) === 2 ? t.get(e) : t[e];
}
function Te(t, e, n) {
  ne(t) === 2 ? t.set(e, n) : t[e] = n;
}
function Ue(t, e) {
  const n = g(t);
  return (n ? x(n) : t)[e];
}
function G(t, e) {
  return t === e ? t !== 0 || 1 / t === 1 / e : t !== t && e !== e;
}
function ze(t) {
  if (t)
    for (; t.finalities.revoke.length > 0; )
      t.finalities.revoke.pop()();
}
function X(t, e) {
  return e ? t : [""].concat(t).map((n) => {
    const r = `${n}`;
    return r.indexOf("/") === -1 && r.indexOf("~") === -1 ? r : r.replace(/~/g, "~0").replace(/\//g, "~1");
  }).join("/");
}
function Bn(t, e) {
  for (let n = 0; n < e.length - 1; n += 1) {
    const r = e[n];
    if (t = B(ne(t) === 3 ? Array.from(t) : t, r), typeof t != "object")
      throw new Error(`Cannot resolve patch at '${e.join("/")}'.`);
  }
  return t;
}
function Hn(t) {
  const e = Object.create(Object.getPrototypeOf(t));
  return Reflect.ownKeys(t).forEach((n) => {
    let r = Reflect.getOwnPropertyDescriptor(t, n);
    if (r.enumerable && r.configurable && r.writable) {
      e[n] = t[n];
      return;
    }
    r.writable || (r.writable = !0, r.configurable = !0), (r.get || r.set) && (r = {
      configurable: !0,
      writable: !0,
      enumerable: r.enumerable,
      value: t[n]
    }), Reflect.defineProperty(e, n, r);
  }), e;
}
const Jn = Object.prototype.propertyIsEnumerable;
function on(t, e) {
  let n;
  if (Array.isArray(t))
    return Array.prototype.concat.call(t);
  if (t instanceof Set) {
    if (!ut(t)) {
      const r = Object.getPrototypeOf(t).constructor;
      return new r(t.values());
    }
    return Set.prototype.difference ? Set.prototype.difference.call(t, /* @__PURE__ */ new Set()) : new Set(t.values());
  } else if (t instanceof Map) {
    if (!ft(t)) {
      const r = Object.getPrototypeOf(t).constructor;
      return new r(t);
    }
    return new Map(t);
  } else if (e != null && e.mark && (n = e.mark(t, L), n !== void 0) && n !== L.mutable) {
    if (n === L.immutable)
      return Hn(t);
    if (typeof n == "function") {
      if (e.enablePatches || e.enableAutoFreeze)
        throw new Error("You can't use mark and patches or auto freeze together.");
      return n();
    }
    throw new Error(`Unsupported mark result: ${n}`);
  } else if (typeof t == "object" && Object.getPrototypeOf(t) === Object.prototype) {
    const r = {};
    return Object.keys(t).forEach((s) => {
      r[s] = t[s];
    }), Object.getOwnPropertySymbols(t).forEach((s) => {
      Jn.call(t, s) && (r[s] = t[s]);
    }), r;
  } else
    throw new Error("Please check mark() to ensure that it is a stable marker draftable function.");
}
function v(t) {
  t.copy || (t.copy = on(t.original, t.options));
}
function le(t) {
  if (!V(t))
    return lt(t);
  if (Array.isArray(t))
    return t.map(le);
  if (t instanceof Map) {
    const n = Array.from(t.entries()).map(([r, s]) => [
      r,
      le(s)
    ]);
    if (!ft(t)) {
      const r = Object.getPrototypeOf(t).constructor;
      return new r(n);
    }
    return new Map(n);
  }
  if (t instanceof Set) {
    const n = Array.from(t).map(le);
    if (!ut(t)) {
      const r = Object.getPrototypeOf(t).constructor;
      return new r(n);
    }
    return new Set(n);
  }
  const e = Object.create(Object.getPrototypeOf(t));
  for (const n in t)
    e[n] = le(t[n]);
  return e;
}
function Ae(t) {
  return te(t) ? le(t) : t;
}
function K(t) {
  var e;
  t.assignedMap = (e = t.assignedMap) !== null && e !== void 0 ? e : /* @__PURE__ */ new Map(), t.operated || (t.operated = !0, t.parent && K(t.parent));
}
function Pt() {
  throw new Error("Cannot modify frozen object");
}
function ie(t, e, n, r, s) {
  {
    n = n ?? /* @__PURE__ */ new WeakMap(), r = r ?? [], s = s ?? [];
    const o = n.has(t) ? n.get(t) : t;
    if (r.length > 0) {
      const a = r.indexOf(o);
      if (o && typeof o == "object" && a !== -1)
        throw r[0] === o ? new Error("Forbids circular reference") : new Error(`Forbids circular reference: ~/${s.slice(0, a).map((c, u) => {
          if (typeof c == "symbol")
            return `[${c.toString()}]`;
          const f = r[u];
          return typeof c == "object" && (f instanceof Map || f instanceof Set) ? Array.from(f.keys()).indexOf(c) : c;
        }).join("/")}`);
      r.push(o), s.push(e);
    } else
      r.push(o);
  }
  if (Object.isFrozen(t) || te(t)) {
    r.pop(), s.pop();
    return;
  }
  switch (ne(t)) {
    case 2:
      for (const [a, c] of t)
        ie(a, a, n, r, s), ie(c, a, n, r, s);
      t.set = t.clear = t.delete = Pt;
      break;
    case 3:
      for (const a of t)
        ie(a, a, n, r, s);
      t.add = t.clear = t.delete = Pt;
      break;
    case 1:
      Object.freeze(t);
      let o = 0;
      for (const a of t)
        ie(a, o, n, r, s), o += 1;
      break;
    default:
      Object.freeze(t), Object.keys(t).forEach((a) => {
        const c = t[a];
        ie(c, a, n, r, s);
      });
  }
  r.pop(), s.pop();
}
function dt(t, e) {
  const n = ne(t);
  if (n === 0)
    Reflect.ownKeys(t).forEach((r) => {
      e(r, t[r], t);
    });
  else if (n === 1) {
    let r = 0;
    for (const s of t)
      e(r, s, t), r += 1;
  } else
    t.forEach((r, s) => e(s, r, t));
}
function an(t, e, n) {
  if (te(t) || !V(t, n) || e.has(t) || Object.isFrozen(t))
    return;
  const r = t instanceof Set, s = r ? /* @__PURE__ */ new Map() : void 0;
  if (e.add(t), dt(t, (i, o) => {
    var a;
    if (te(o)) {
      const c = g(o);
      v(c);
      const u = !((a = c.assignedMap) === null || a === void 0) && a.size || c.operated ? c.copy : c.original;
      Te(r ? s : t, i, u);
    } else
      an(o, e, n);
  }), s) {
    const i = t, o = Array.from(i);
    i.clear(), o.forEach((a) => {
      i.add(s.has(a) ? s.get(a) : a);
    });
  }
}
function Yn(t, e) {
  const n = t.type === 3 ? t.setMap : t.copy;
  t.finalities.revoke.length > 1 && t.assignedMap.get(e) && n && an(B(n, e), t.finalities.handledSet, t.options);
}
function We(t) {
  t.type === 3 && t.copy && (t.copy.clear(), t.setMap.forEach((e) => {
    t.copy.add(lt(e));
  }));
}
function Ke(t, e, n, r) {
  if (t.operated && t.assignedMap && t.assignedMap.size > 0 && !t.finalized) {
    if (n && r) {
      const i = sn(t);
      i && e(t, i, n, r);
    }
    t.finalized = !0;
  }
}
function ht(t, e, n, r) {
  const s = g(n);
  s && (s.callbacks || (s.callbacks = []), s.callbacks.push((i, o) => {
    var a;
    const c = t.type === 3 ? t.setMap : t.copy;
    if (G(B(c, e), n)) {
      let u = s.original;
      s.copy && (u = s.copy), We(t), Ke(t, r, i, o), t.options.enableAutoFreeze && (t.options.updatedValues = (a = t.options.updatedValues) !== null && a !== void 0 ? a : /* @__PURE__ */ new WeakMap(), t.options.updatedValues.set(u, s.original)), Te(c, e, u);
    }
  }), t.options.enableAutoFreeze && s.finalities !== t.finalities && (t.options.enableAutoFreeze = !1)), V(n, t.options) && t.finalities.draft.push(() => {
    const i = t.type === 3 ? t.setMap : t.copy;
    G(B(i, e), n) && Yn(t, e);
  });
}
function Zn(t, e, n, r, s) {
  let { original: i, assignedMap: o, options: a } = t, c = t.copy;
  c.length < i.length && ([i, c] = [c, i], [n, r] = [r, n]);
  for (let u = 0; u < i.length; u += 1)
    if (o.get(u.toString()) && c[u] !== i[u]) {
      const f = e.concat([u]), l = X(f, s);
      n.push({
        op: E.Replace,
        path: l,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: Ae(c[u])
      }), r.push({
        op: E.Replace,
        path: l,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: Ae(i[u])
      });
    }
  for (let u = i.length; u < c.length; u += 1) {
    const f = e.concat([u]), l = X(f, s);
    n.push({
      op: E.Add,
      path: l,
      // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
      value: Ae(c[u])
    });
  }
  if (i.length < c.length) {
    const { arrayLengthAssignment: u = !0 } = a.enablePatches;
    if (u) {
      const f = e.concat(["length"]), l = X(f, s);
      r.push({
        op: E.Replace,
        path: l,
        value: i.length
      });
    } else
      for (let f = c.length; i.length < f; f -= 1) {
        const l = e.concat([f - 1]), d = X(l, s);
        r.push({
          op: E.Remove,
          path: d
        });
      }
  }
}
function Xn({ original: t, copy: e, assignedMap: n }, r, s, i, o) {
  n.forEach((a, c) => {
    const u = B(t, c), f = Ae(B(e, c)), l = a ? he(t, c) ? E.Replace : E.Add : E.Remove;
    if (G(u, f) && l === E.Replace)
      return;
    const d = r.concat(c), h = X(d, o);
    s.push(l === E.Remove ? { op: l, path: h } : { op: l, path: h, value: f }), i.push(l === E.Add ? { op: E.Remove, path: h } : l === E.Remove ? { op: E.Add, path: h, value: u } : { op: E.Replace, path: h, value: u });
  });
}
function er({ original: t, copy: e }, n, r, s, i) {
  let o = 0;
  t.forEach((a) => {
    if (!e.has(a)) {
      const c = n.concat([o]), u = X(c, i);
      r.push({
        op: E.Remove,
        path: u,
        value: a
      }), s.unshift({
        op: E.Add,
        path: u,
        value: a
      });
    }
    o += 1;
  }), o = 0, e.forEach((a) => {
    if (!t.has(a)) {
      const c = n.concat([o]), u = X(c, i);
      r.push({
        op: E.Add,
        path: u,
        value: a
      }), s.unshift({
        op: E.Remove,
        path: u,
        value: a
      });
    }
    o += 1;
  });
}
function ye(t, e, n, r) {
  const { pathAsArray: s = !0 } = t.options.enablePatches;
  switch (t.type) {
    case 0:
    case 2:
      return Xn(t, e, n, r, s);
    case 1:
      return Zn(t, e, n, r, s);
    case 3:
      return er(t, e, n, r, s);
  }
}
const Me = (t, e, n = !1) => {
  if (typeof t == "object" && t !== null && (!V(t, e) || n))
    throw new Error("Strict mode: Mutable data cannot be accessed directly, please use 'unsafe(callback)' wrap.");
}, Qe = {
  get size() {
    return x(g(this)).size;
  },
  has(t) {
    return x(g(this)).has(t);
  },
  set(t, e) {
    const n = g(this), r = x(n);
    return (!r.has(t) || !G(r.get(t), e)) && (v(n), K(n), n.assignedMap.set(t, !0), n.copy.set(t, e), ht(n, t, e, ye)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = g(this);
    return v(e), K(e), e.original.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.copy.delete(t), !0;
  },
  clear() {
    const t = g(this);
    if (this.size) {
      v(t), K(t), t.assignedMap = /* @__PURE__ */ new Map();
      for (const [e] of t.original)
        t.assignedMap.set(e, !1);
      t.copy.clear();
    }
  },
  forEach(t, e) {
    const n = g(this);
    x(n).forEach((r, s) => {
      t.call(e, this.get(s), s, this);
    });
  },
  get(t) {
    var e, n;
    const r = g(this), s = x(r).get(t), i = ((n = (e = r.options).mark) === null || n === void 0 ? void 0 : n.call(e, s, L)) === L.mutable;
    if (r.options.strict && Me(s, r.options, i), i || r.finalized || !V(s, r.options) || s !== r.original.get(t))
      return s;
    const o = ct.createDraft({
      original: s,
      parentDraft: r,
      key: t,
      finalities: r.finalities,
      options: r.options
    });
    return v(r), r.copy.set(t, o), o;
  },
  keys() {
    return x(g(this)).keys();
  },
  values() {
    const t = this.keys();
    return {
      [Oe]: () => this.values(),
      next: () => {
        const e = t.next();
        return e.done ? e : {
          done: !1,
          value: this.get(e.value)
        };
      }
    };
  },
  entries() {
    const t = this.keys();
    return {
      [Oe]: () => this.entries(),
      next: () => {
        const e = t.next();
        if (e.done)
          return e;
        const n = this.get(e.value);
        return {
          done: !1,
          value: [e.value, n]
        };
      }
    };
  },
  [Oe]() {
    return this.entries();
  }
}, tr = Reflect.ownKeys(Qe), vt = (t, e, { isValuesIterator: n }) => () => {
  var r, s;
  const i = e.next();
  if (i.done)
    return i;
  const o = i.value;
  let a = t.setMap.get(o);
  const c = g(a), u = ((s = (r = t.options).mark) === null || s === void 0 ? void 0 : s.call(r, a, L)) === L.mutable;
  if (t.options.strict && Me(o, t.options, u), !u && !c && V(o, t.options) && !t.finalized && t.original.has(o)) {
    const f = ct.createDraft({
      original: o,
      parentDraft: t,
      key: o,
      finalities: t.finalities,
      options: t.options
    });
    t.setMap.set(o, f), a = f;
  } else c && (a = c.proxy);
  return {
    done: !1,
    value: n ? a : [a, a]
  };
}, Ce = {
  get size() {
    return g(this).setMap.size;
  },
  has(t) {
    const e = g(this);
    if (e.setMap.has(t))
      return !0;
    v(e);
    const n = g(t);
    return !!(n && e.setMap.has(n.original));
  },
  add(t) {
    const e = g(this);
    return this.has(t) || (v(e), K(e), e.assignedMap.set(t, !0), e.setMap.set(t, t), ht(e, t, t, ye)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = g(this);
    v(e), K(e);
    const n = g(t);
    return n && e.setMap.has(n.original) ? (e.assignedMap.set(n.original, !1), e.setMap.delete(n.original)) : (!n && e.setMap.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.setMap.delete(t));
  },
  clear() {
    if (!this.size)
      return;
    const t = g(this);
    v(t), K(t);
    for (const e of t.original)
      t.assignedMap.set(e, !1);
    t.setMap.clear();
  },
  values() {
    const t = g(this);
    v(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.values(),
      next: vt(t, e, { isValuesIterator: !0 })
    };
  },
  entries() {
    const t = g(this);
    v(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.entries(),
      next: vt(t, e, {
        isValuesIterator: !1
      })
    };
  },
  keys() {
    return this.values();
  },
  [Oe]() {
    return this.values();
  },
  forEach(t, e) {
    const n = this.values();
    let r = n.next();
    for (; !r.done; )
      t.call(e, r.value, r.value, this), r = n.next();
  }
};
Set.prototype.difference && Object.assign(Ce, {
  intersection(t) {
    return Set.prototype.intersection.call(new Set(this.values()), t);
  },
  union(t) {
    return Set.prototype.union.call(new Set(this.values()), t);
  },
  difference(t) {
    return Set.prototype.difference.call(new Set(this.values()), t);
  },
  symmetricDifference(t) {
    return Set.prototype.symmetricDifference.call(new Set(this.values()), t);
  },
  isSubsetOf(t) {
    return Set.prototype.isSubsetOf.call(new Set(this.values()), t);
  },
  isSupersetOf(t) {
    return Set.prototype.isSupersetOf.call(new Set(this.values()), t);
  },
  isDisjointFrom(t) {
    return Set.prototype.isDisjointFrom.call(new Set(this.values()), t);
  }
});
const nr = Reflect.ownKeys(Ce), cn = /* @__PURE__ */ new WeakSet(), un = {
  get(t, e, n) {
    var r, s;
    const i = (r = t.copy) === null || r === void 0 ? void 0 : r[e];
    if (i && cn.has(i))
      return i;
    if (e === rn)
      return t;
    let o;
    if (t.options.mark) {
      const u = e === "size" && (t.original instanceof Map || t.original instanceof Set) ? Reflect.get(t.original, e) : Reflect.get(t.original, e, n);
      if (o = t.options.mark(u, L), o === L.mutable)
        return t.options.strict && Me(u, t.options, !0), u;
    }
    const a = x(t);
    if (a instanceof Map && tr.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(Qe, "size").get.call(t.proxy);
      const u = Qe[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (a instanceof Set && nr.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(Ce, "size").get.call(t.proxy);
      const u = Ce[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (!he(a, e)) {
      const u = kt(a, e);
      return u ? "value" in u ? u.value : (
        // !case: support for getter
        (s = u.get) === null || s === void 0 ? void 0 : s.call(t.proxy)
      ) : void 0;
    }
    const c = a[e];
    if (t.options.strict && Me(c, t.options), t.finalized || !V(c, t.options))
      return c;
    if (c === Ue(t.original, e)) {
      if (v(t), t.copy[e] = pt({
        original: t.original[e],
        parentDraft: t,
        key: t.type === 1 ? Number(e) : e,
        finalities: t.finalities,
        options: t.options
      }), typeof o == "function") {
        const u = g(t.copy[e]);
        return v(u), K(u), u.copy;
      }
      return t.copy[e];
    }
    return c;
  },
  set(t, e, n) {
    var r;
    if (t.type === 3 || t.type === 2)
      throw new Error("Map/Set draft does not support any property assignment.");
    let s;
    if (t.type === 1 && e !== "length" && !(Number.isInteger(s = Number(e)) && s >= 0 && (e === 0 || s === 0 || String(s) === String(e))))
      throw new Error("Only supports setting array indices and the 'length' property.");
    const i = kt(x(t), e);
    if (i != null && i.set)
      return i.set.call(t.proxy, n), !0;
    const o = Ue(x(t), e), a = g(o);
    return a && G(a.original, n) ? (t.copy[e] = n, t.assignedMap = (r = t.assignedMap) !== null && r !== void 0 ? r : /* @__PURE__ */ new Map(), t.assignedMap.set(e, !1), !0) : (G(n, o) && (n !== void 0 || he(t.original, e)) || (v(t), K(t), he(t.original, e) && G(n, t.original[e]) ? t.assignedMap.delete(e) : t.assignedMap.set(e, !0), t.copy[e] = n, ht(t, e, n, ye)), !0);
  },
  has(t, e) {
    return e in x(t);
  },
  ownKeys(t) {
    return Reflect.ownKeys(x(t));
  },
  getOwnPropertyDescriptor(t, e) {
    const n = x(t), r = Reflect.getOwnPropertyDescriptor(n, e);
    return r && {
      writable: !0,
      configurable: t.type !== 1 || e !== "length",
      enumerable: r.enumerable,
      value: n[e]
    };
  },
  getPrototypeOf(t) {
    return Reflect.getPrototypeOf(t.original);
  },
  setPrototypeOf() {
    throw new Error("Cannot call 'setPrototypeOf()' on drafts");
  },
  defineProperty() {
    throw new Error("Cannot call 'defineProperty()' on drafts");
  },
  deleteProperty(t, e) {
    var n;
    return t.type === 1 ? un.set.call(this, t, e, void 0, t.proxy) : (Ue(t.original, e) !== void 0 || e in t.original ? (v(t), K(t), t.assignedMap.set(e, !1)) : (t.assignedMap = (n = t.assignedMap) !== null && n !== void 0 ? n : /* @__PURE__ */ new Map(), t.assignedMap.delete(e)), t.copy && delete t.copy[e], !0);
  }
};
function pt(t) {
  const { original: e, parentDraft: n, key: r, finalities: s, options: i } = t, o = ne(e), a = {
    type: o,
    finalized: !1,
    parent: n,
    original: e,
    copy: null,
    proxy: null,
    finalities: s,
    options: i,
    // Mapping of draft Set items to their corresponding draft values.
    setMap: o === 3 ? new Map(e.entries()) : void 0
  };
  (r || "key" in t) && (a.key = r);
  const { proxy: c, revoke: u } = Proxy.revocable(o === 1 ? Object.assign([], a) : a, un);
  if (s.revoke.push(u), cn.add(c), a.proxy = c, n) {
    const f = n;
    f.finalities.draft.push((l, d) => {
      var h, p;
      const T = g(c);
      let w = f.type === 3 ? f.setMap : f.copy;
      const b = B(w, r), _ = g(b);
      if (_) {
        let S = _.original;
        _.operated && (S = lt(b)), We(_), Ke(_, ye, l, d), f.options.enableAutoFreeze && (f.options.updatedValues = (h = f.options.updatedValues) !== null && h !== void 0 ? h : /* @__PURE__ */ new WeakMap(), f.options.updatedValues.set(S, _.original)), Te(w, r, S);
      }
      (p = T.callbacks) === null || p === void 0 || p.forEach((S) => {
        S(l, d);
      });
    });
  } else {
    const f = g(c);
    f.finalities.draft.push((l, d) => {
      We(f), Ke(f, ye, l, d);
    });
  }
  return c;
}
ct.createDraft = pt;
function rr(t, e, n, r, s) {
  var i;
  const o = g(t), a = (i = o == null ? void 0 : o.original) !== null && i !== void 0 ? i : t, c = !!e.length;
  if (o != null && o.operated)
    for (; o.finalities.draft.length > 0; )
      o.finalities.draft.pop()(n, r);
  const u = c ? e[0] : o ? o.operated ? o.copy : o.original : t;
  return o && ze(o), s && ie(u, u, o == null ? void 0 : o.options.updatedValues), [
    u,
    n && c ? [{ op: E.Replace, path: [], value: e[0] }] : n,
    r && c ? [{ op: E.Replace, path: [], value: a }] : r
  ];
}
function sr(t, e) {
  var n;
  const r = {
    draft: [],
    revoke: [],
    handledSet: /* @__PURE__ */ new WeakSet()
  };
  let s, i;
  e.enablePatches && (s = [], i = []);
  const a = ((n = e.mark) === null || n === void 0 ? void 0 : n.call(e, t, L)) === L.mutable || !V(t, e) ? t : pt({
    original: t,
    parentDraft: null,
    finalities: r,
    options: e
  });
  return [
    a,
    (c = []) => {
      const [u, f, l] = rr(a, c, s, i, e.enableAutoFreeze);
      return e.enablePatches ? [u, f, l] : u;
    }
  ];
}
function Ge(t) {
  const { rootDraft: e, value: n, useRawReturn: r = !1, isRoot: s = !0 } = t;
  dt(n, (i, o, a) => {
    const c = g(o);
    if (c && e && c.finalities === e.finalities) {
      t.isContainDraft = !0;
      const u = c.original;
      if (a instanceof Set) {
        const f = Array.from(a);
        a.clear(), f.forEach((l) => a.add(i === l ? u : l));
      } else
        Te(a, i, u);
    } else typeof o == "object" && o !== null && (t.value = o, t.isRoot = !1, Ge(t));
  }), s && (t.isContainDraft || console.warn("The return value does not contain any draft, please use 'rawReturn()' to wrap the return value to improve performance."), r && console.warn("The return value contains drafts, please don't use 'rawReturn()' to wrap the return value."));
}
function fn(t) {
  var e;
  const n = g(t);
  if (!V(t, n == null ? void 0 : n.options))
    return t;
  const r = ne(t);
  if (n && !n.operated)
    return n.original;
  let s;
  function i() {
    s = r === 2 ? ft(t) ? new Map(t) : new (Object.getPrototypeOf(t)).constructor(t) : r === 3 ? Array.from(n.setMap.values()) : on(t, n == null ? void 0 : n.options);
  }
  if (n) {
    n.finalized = !0;
    try {
      i();
    } finally {
      n.finalized = !1;
    }
  } else
    s = t;
  if (dt(s, (o, a) => {
    if (n && G(B(n.original, o), a))
      return;
    const c = fn(a);
    c !== a && (s === t && i(), Te(s, o, c));
  }), r === 3) {
    const o = (e = n == null ? void 0 : n.original) !== null && e !== void 0 ? e : s;
    return ut(o) ? new Set(s) : new (Object.getPrototypeOf(o)).constructor(s);
  }
  return s;
}
function It(t) {
  if (!te(t))
    throw new Error(`current() is only used for Draft, parameter: ${t}`);
  return fn(t);
}
const ir = (t) => function e(n, r, s) {
  var i, o, a;
  if (typeof n == "function" && typeof r != "function")
    return function(m, ...P) {
      return e(m, (R) => n.call(this, R, ...P), r);
    };
  const c = n, u = r;
  let f = s;
  if (typeof r != "function" && (f = r), f !== void 0 && Object.prototype.toString.call(f) !== "[object Object]")
    throw new Error(`Invalid options: ${f}, 'options' should be an object.`);
  f = Object.assign(Object.assign({}, t), f);
  const l = te(c) ? It(c) : c, d = Array.isArray(f.mark) ? (m, P) => {
    for (const R of f.mark) {
      if (typeof R != "function")
        throw new Error(`Invalid mark: ${R}, 'mark' should be a function.`);
      const Y = R(m, P);
      if (Y)
        return Y;
    }
  } : f.mark, h = (i = f.enablePatches) !== null && i !== void 0 ? i : !1, p = (o = f.strict) !== null && o !== void 0 ? o : !1, w = {
    enableAutoFreeze: (a = f.enableAutoFreeze) !== null && a !== void 0 ? a : !1,
    mark: d,
    strict: p,
    enablePatches: h
  };
  if (!V(l, w) && typeof l == "object" && l !== null)
    throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
  const [b, _] = sr(l, w);
  if (typeof r != "function") {
    if (!V(l, w))
      throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
    return [b, _];
  }
  let S;
  try {
    S = u(b);
  } catch (m) {
    throw ze(g(b)), m;
  }
  const N = (m) => {
    const P = g(b);
    if (!te(m)) {
      if (m !== void 0 && !G(m, b) && (P != null && P.operated))
        throw new Error("Either the value is returned as a new non-draft value, or only the draft is modified without returning any value.");
      const Y = m == null ? void 0 : m[Gn];
      if (Y) {
        const H = Y[0];
        return w.strict && typeof m == "object" && m !== null && Ge({
          rootDraft: P,
          value: m,
          useRawReturn: !0
        }), _([H]);
      }
      if (m !== void 0)
        return typeof m == "object" && m !== null && Ge({ rootDraft: P, value: m }), _([m]);
    }
    if (m === b || m === void 0)
      return _([]);
    const R = g(m);
    if (w === R.options) {
      if (R.operated)
        throw new Error("Cannot return a modified child draft.");
      return _([It(m)]);
    }
    return _([m]);
  };
  return S instanceof Promise ? S.then(N, (m) => {
    throw ze(g(b)), m;
  }) : N(S);
}, Be = ir();
Object.prototype.constructor.toString();
function ln(t, e) {
  const n = Object.keys(t), r = Object.keys(e);
  return n.length === r.length && Object.keys(t).every((s) => e.hasOwnProperty(s));
}
function $t(t, e) {
  return Object.keys(t).length === Object.keys(e).length && Object.keys(t).every(
    (n) => e.hasOwnProperty(n) && t[n] === e[n]
  );
}
function Ee(t, e) {
  return typeof t != "object" || typeof e != "object" || t === null || e === null ? t === e : ln(t, e) ? Object.keys(t).every(
    (n) => Ee(t[n], e[n])
  ) : !1;
}
function yt(t) {
  if (!de(t))
    return t;
  const e = {};
  for (const [n, r] of Object.entries(t))
    r !== void 0 && (e[n] = r);
  return e;
}
function dn(t, e) {
  if (!de(t) || !de(e))
    return e;
  const n = { ...t };
  for (const r of Object.keys(e)) {
    if (e[r] === void 0) continue;
    if (e[r] === null) {
      delete n[r];
      continue;
    }
    const s = de(t[r]) && de(e[r]);
    n[r] = s ? dn(t[r], e[r]) : e[r];
  }
  return n;
}
function de(t) {
  return typeof t == "object" && t !== null && !Array.isArray(t);
}
function or(t, e, n) {
  if (!t || e.length === 0)
    return;
  let r = t || {};
  for (let i = 0; i < e.length - 1; i++) {
    const o = e[i];
    (!(o in r) || typeof r[o] != "object") && (r[o] = typeof e[i + 1] == "number" ? [] : {}), r = r[o];
  }
  const s = e[e.length - 1];
  Array.isArray(r) && typeof s == "number" ? r.splice(s, 0, n) : r[s] = n;
}
function jt(t, e, n) {
  if (!t || e.length === 0)
    return;
  let r = t || {};
  for (let s = 0; s < e.length - 1; s++) {
    const i = e[s];
    (!(i in r) || typeof r[i] != "object") && (r[i] = typeof e[s + 1] == "number" ? [] : {}), r = r[i];
  }
  r[e[e.length - 1]] = n;
}
function hn(t, e) {
  if (!t || e.length === 0)
    return;
  const [n, ...r] = e;
  if (n in t) {
    if (r.length === 0) {
      Array.isArray(t) ? t.splice(n, 1) : delete t[n];
      return;
    }
    hn(t[n], r), ar(t[n]) && delete t[n];
  }
}
function ar(t) {
  return t && Object.keys(t).length === 0;
}
const Dt = /ZULU|YEKT|YEKST|YAPT|YAKT|YAKST|XJT|WGT|WGST|WFT|WETDST|WET|WDT|WAT|WAST|WAKT|WADT|VUT|VOLT|VLAT|VLAST|VET|UZT|UZST|UYT|UYST|UTC|UT|ULAT|ULAST|UCT|TVT|TRUT|TOT|TMT|TKT|TJT|TFT|TAHT|SGT|SCT|SAST|SADT|RET|PYT|PYST|PWT|PST|PONT|PMST|PMDT|PKT|PKST|PHT|PGT|PETT|PETST|PET|PDT|OMST|OMSST|NZT|NZST|NZDT|NUT|NST|NPT|NOVT|NOVST|NFT|NDT|MYT|MVT|MUT|MUST|MST|MSK|MSD|MPT|MMT|MHT|MEZ|METDST|MET|MESZ|MEST|MDT|MAWT|MART|MAGT|MAGST|LKT|LINT|LIGT|LHST|LHDT|KST|KRAT|KRAST|KOST|KGT|KGST|KDT|JST|JAYT|IST|IRT|IRKT|IRKST|IOT|IDT|ICT|HST|HKT|GYT|GMT|GILT|GFT|GET|GEST|GAMT|GALT|FNT|FNST|FKT|FKST|FJT|FJST|FET|EST|EGT|EGST|EETDST|EET|EEST|EDT|EAT|EAST|EASST|DDUT|DAVT|CXT|CST|COT|CLT|CLST|CKT|CHUT|CHAST|CHADT|CETDST|CET|CEST|CDT|CCT|CAST|CADT|BTT|BST|BRT|BRST|BRA|BOT|BORT|BNT|BDT|BDST|AZT|AZST|AZOT|AZOST|AWST|AWSST|AST|ART|ARST|ANAT|ANAST|AMT|AMST|ALMT|ALMST|AKST|AKDT|AFT|AEST|AESST|AEDT|ADT|ACWST|ACT|ACST|ACSST|ACDT$/, cr = {
  ZULU: 0,
  YEKT: 18e3,
  YEKST: 21600,
  YAPT: 36e3,
  YAKT: 32400,
  YAKST: 32400,
  XJT: 21600,
  WGT: -10800,
  WGST: -7200,
  WFT: 43200,
  WETDST: 3600,
  WET: 0,
  WDT: 32400,
  WAT: 3600,
  WAST: 25200,
  WAKT: 43200,
  WADT: 28800,
  VUT: 39600,
  VOLT: 10800,
  VLAT: 36e3,
  VLAST: 36e3,
  VET: -14400,
  UZT: 18e3,
  UZST: 21600,
  UYT: -10800,
  UYST: -7200,
  UTC: 0,
  UT: 0,
  ULAT: 28800,
  ULAST: 32400,
  UCT: 0,
  TVT: 43200,
  TRUT: 36e3,
  TOT: 46800,
  TMT: 18e3,
  TKT: 46800,
  TJT: 18e3,
  TFT: 18e3,
  TAHT: -36e3,
  SGT: 28800,
  SCT: 14400,
  SAST: 7200,
  SADT: 37800,
  RET: 14400,
  PYT: -14400,
  PYST: -10800,
  PWT: 32400,
  PST: -28800,
  PONT: 39600,
  PMST: -10800,
  PMDT: -7200,
  PKT: 18e3,
  PKST: 21600,
  PHT: 28800,
  PGT: 36e3,
  PETT: 43200,
  PETST: 43200,
  PET: -18e3,
  PDT: -25200,
  OMST: 21600,
  OMSST: 21600,
  NZT: 43200,
  NZST: 43200,
  NZDT: 46800,
  NUT: -39600,
  NST: -12600,
  NPT: 20700,
  NOVT: 25200,
  NOVST: 25200,
  NFT: -12600,
  NDT: -9e3,
  MYT: 28800,
  MVT: 18e3,
  MUT: 14400,
  MUST: 18e3,
  MST: -25200,
  MSK: 10800,
  MSD: 14400,
  MPT: 36e3,
  MMT: 23400,
  MHT: 43200,
  MEZ: 3600,
  METDST: 7200,
  MET: 3600,
  MESZ: 7200,
  MEST: 7200,
  MDT: -21600,
  MAWT: 18e3,
  MART: -34200,
  MAGT: 39600,
  MAGST: 39600,
  LKT: 19800,
  LINT: 50400,
  LIGT: 36e3,
  LHST: 37800,
  LHDT: 37800,
  KST: 32400,
  KRAT: 25200,
  KRAST: 25200,
  KOST: 39600,
  KGT: 21600,
  KGST: 21600,
  KDT: 36e3,
  JST: 32400,
  JAYT: 32400,
  IST: 7200,
  IRT: 12600,
  IRKT: 28800,
  IRKST: 28800,
  IOT: 21600,
  IDT: 10800,
  ICT: 25200,
  HST: -36e3,
  HKT: 28800,
  GYT: -14400,
  GMT: 0,
  GILT: 43200,
  GFT: -10800,
  GET: 14400,
  GEST: 14400,
  GAMT: -32400,
  GALT: -21600,
  FNT: -7200,
  FNST: -3600,
  FKT: -10800,
  FKST: -10800,
  FJT: 43200,
  FJST: 46800,
  FET: 10800,
  EST: -18e3,
  EGT: -3600,
  EGST: 0,
  EETDST: 10800,
  EET: 7200,
  EEST: 10800,
  EDT: -14400,
  EAT: 10800,
  EAST: -21600,
  EASST: -21600,
  DDUT: 36e3,
  DAVT: 25200,
  CXT: 25200,
  CST: -21600,
  COT: -18e3,
  CLT: -14400,
  CLST: -10800,
  CKT: -36e3,
  CHUT: 36e3,
  CHAST: 45900,
  CHADT: 49500,
  CETDST: 7200,
  CET: 3600,
  CEST: 7200,
  CDT: -18e3,
  CCT: 28800,
  CAST: 34200,
  CADT: 37800,
  BTT: 21600,
  BST: 3600,
  BRT: -10800,
  BRST: -7200,
  BRA: -10800,
  BOT: -14400,
  BORT: 28800,
  BNT: 28800,
  BDT: 21600,
  BDST: 7200,
  AZT: 14400,
  AZST: 14400,
  AZOT: -3600,
  AZOST: 0,
  AWST: 28800,
  AWSST: 32400,
  AST: -14400,
  ART: -10800,
  ARST: -10800,
  ANAT: 43200,
  ANAST: 43200,
  AMT: -14400,
  AMST: 14400,
  ALMT: 21600,
  ALMST: 25200,
  AKST: -32400,
  AKDT: -28800,
  AFT: 16200,
  AEST: 36e3,
  AESST: 39600,
  AEDT: 39600,
  ADT: -10800,
  ACWST: 31500,
  ACT: -18e3,
  ACST: 34200,
  ACSST: 37800,
  ACDT: 37800
};
function ur(t) {
  return new Date(t);
}
function fr(t) {
  return /* @__PURE__ */ new Date(t + "Z");
}
const lr = /^(\d+)[\./-](\d+)[\./-](\d+)$/;
function dr(t) {
  const e = t.match(lr);
  if (!e)
    return null;
  const [n, r, s, i] = e;
  return r <= 0 || s <= 0 || i <= 0 ? null : r > 999 ? new Date(Date.UTC(r, s - 1, i, 0, 0, 0, 0)) : new Date(Date.UTC(i, r - 1, s, 0, 0, 0, 0));
}
function hr(t) {
  const [e, n] = t.split(" ");
  return /* @__PURE__ */ new Date(e + "T" + n + "Z");
}
function pr(t) {
  const [e, n] = t.split(" ");
  return /* @__PURE__ */ new Date(e + "T" + n + "Z");
}
function yr(t) {
  return new Date(t);
}
function br(t) {
  const e = /^(\w{3}) (\w{3}) (\d{2}) (\d{4})$/;
  if (!t.match(e))
    throw new Error(`Unable to parse \`${t}\` as a date.`);
  const r = /* @__PURE__ */ new Date(t + " UTC");
  return new Date(
    Date.UTC(
      r.getUTCFullYear(),
      r.getUTCMonth(),
      r.getUTCDate(),
      0,
      0,
      0,
      0
    )
  );
}
function Tr(t) {
  const e = /^(.+T.+)([+-])(\d{2})$/, n = t.match(e);
  if (n) {
    const [, r, s, i] = n, o = `${r}${s}${i}:00`;
    return new Date(o);
  }
  return null;
}
function wr(t) {
  const [e, n] = t.split(", "), [r, s, i] = e.split("/").map(Number), o = n.match(/(\d{1,2}):(\d{2}):(\d{2}) (AM|PM)/);
  if (!o)
    throw new Error(`Unable to parse time from: ${t}`);
  let [, a, c, u, f] = o;
  return a = Number(a), c = Number(c), u = Number(u), f === "PM" && a !== 12 ? a += 12 : f === "AM" && a === 12 && (a = 0), new Date(Date.UTC(i, r - 1, s, a, c, u));
}
function gr(t) {
  switch (t) {
    case "epoch":
      return /* @__PURE__ */ new Date(0);
    case "infinity":
    case "-infinity":
    case "today":
    case "tomorrow":
    case "yesterday":
      return null;
  }
}
function _r(t) {
  const e = t.match(Dt);
  if (!e)
    return null;
  const [n] = e, r = cr[n], s = new Date(t.replace(Dt, "Z"));
  return new Date(s.getTime() - r * 1e3);
}
const mr = [
  dr,
  pr,
  br,
  wr,
  yr,
  fr,
  Tr,
  hr,
  ur,
  gr,
  _r
];
function Sr(t, e) {
  try {
    const n = t(e);
    return n instanceof Date && !isNaN(n.getTime()) ? n : null;
  } catch {
    return null;
  }
}
function He(t) {
  for (const e of mr) {
    const n = Sr(e, t);
    if (n)
      return n;
  }
  return null;
}
function Or(t) {
  try {
    const e = JSON.parse(t);
    return typeof e == "string" ? He(e) : null;
  } catch {
    return null;
  }
}
function ve(t) {
  if (t !== void 0) {
    if (t === null)
      return null;
    if (t instanceof Date)
      return t;
    if (typeof t == "string") {
      const e = He(t) || Or(t) || He(t.trim());
      if (!e)
        throw new Error(`Unable to parse \`${t}\` as a date.`);
      return e;
    } else if (typeof t == "number")
      return new Date(t);
    throw new Error(
      `Invalid date value \`${t}\`. Expected a date, number, or string, got type ${typeof t}.`
    );
  }
}
function Ar(t) {
  return t.cardinality === "one";
}
function bt(t) {
  return t["value-type"] === "ref";
}
function Tt(t) {
  return t["value-type"] === "blob";
}
function fe(t, e) {
  return t[e];
}
function Ie(t, e) {
  return e.reduce((n, r) => n && n.get(r), t);
}
function q(t, e) {
  if (e.length === 0) throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.delete(e[0]);
    return;
  }
  const [n, ...r] = e;
  t.has(n) && q(t.get(n), r);
}
function $(t, e, n) {
  if (e.length === 0) throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.set(e[0], n);
    return;
  }
  const [r, ...s] = e;
  let i = t.get(r);
  i || (i = /* @__PURE__ */ new Map(), t.set(r, i)), $(i, s, n);
}
function pn(t, e, n) {
  const r = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map(), i = /* @__PURE__ */ new Map();
  for (const o of e) {
    let [a, c, u, f] = o;
    const l = fe(t, c);
    if (!l) {
      console.warn("no such attr", a, t);
      continue;
    }
    l["checked-data-type"] === "date" && n && (u = ve(u), o[2] = u), bt(l) && $(i, [u, c, a], o), $(r, [a, c, u], o), $(s, [c, a, u], o);
  }
  return { eav: r, aev: s, vae: i };
}
function yn(t) {
  const e = /* @__PURE__ */ new Map(), n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map();
  for (const i of Object.values(t)) {
    const o = i["forward-identity"], [a, c, u] = o, f = i["reverse-identity"];
    if ($(r, [c, u], i), Tt(i) && $(e, [c, u], i), i["primary?"] && $(n, [c], i), f) {
      const [l, d, h] = f;
      $(s, [d, h], i);
    }
  }
  return { blobAttrs: e, primaryKeys: n, forwardIdents: r, revIdents: s };
}
function Er(t) {
  return {
    __type: t.__type,
    attrs: t.attrs,
    triples: j(t.eav, 3),
    cardinalityInference: t.cardinalityInference,
    linkIndex: t.linkIndex,
    useDateObjects: t.useDateObjects
  };
}
function Mr(t) {
  return Je(
    t.attrs,
    t.triples,
    t.cardinalityInference,
    t.linkIndex,
    t.useDateObjects
  );
}
function wt(t) {
  t.attrIndexes = yn(t.attrs);
}
function Je(t, e, n, r, s) {
  const i = pn(t, e, s);
  return i.useDateObjects = s, i.attrs = t, i.attrIndexes = yn(t), i.cardinalityInference = n, i.linkIndex = r, i.__type = "store", i;
}
function we(t, e) {
  var s, i;
  let n;
  if (Array.isArray(e[0])) {
    const [o, a] = e[0], c = t.aev.get(o);
    if (!c)
      return null;
    n = (s = j(c, 2).find((f) => f[2] === a)) == null ? void 0 : s[0];
  } else
    n = e[0];
  if (!n)
    return null;
  const r = e[2];
  if (Array.isArray(r) && r.length === 2 && t.aev.get(r[0])) {
    const [o, a] = r, c = t.aev.get(o);
    if (!c)
      return null;
    const f = (i = j(c, 2).find((T) => T[2] === a)) == null ? void 0 : i[0];
    if (!f)
      return null;
    const [l, d, h, ...p] = e;
    return [n, d, f, ...p];
  } else {
    const [o, ...a] = e;
    return [n, ...a];
  }
}
function Cr(t, e) {
  const n = we(t, e);
  if (!n)
    return;
  const [r, s, i] = n, o = fe(t.attrs, s);
  o && (q(t.eav, [r, s, i]), q(t.aev, [s, r, i]), bt(o) && q(t.vae, [i, s, r]));
}
let kr = 0;
function bn(t, e, n) {
  const [r, s, i] = n;
  let o;
  const a = Ie(t.ea, [r, s, i]);
  return a && (o = a[3]), o || Date.now() * 10 + kr++;
}
function Pr(t, e) {
  const n = we(t, e);
  if (!n)
    return;
  let [r, s, i] = n;
  const o = fe(t.attrs, s);
  if (!o)
    return;
  o["checked-data-type"] === "date" && t.useDateObjects && (i = ve(i));
  const a = Ie(t.eav, [r, s, i]), c = (a == null ? void 0 : a[3]) ?? bn(t, o, n), u = [r, s, i, c];
  Ar(o) ? ($(t.eav, [r, s], /* @__PURE__ */ new Map([[i, u]])), $(t.aev, [s, r], /* @__PURE__ */ new Map([[i, u]]))) : ($(t.eav, [r, s, i], u), $(t.aev, [s, r, i], u)), bt(o) && $(t.vae, [i, s, r], u);
}
function vr(t, e) {
  var d;
  const n = we(t, e);
  if (!n)
    return;
  const [r, s, i] = n, o = fe(t.attrs, s);
  if (!o) return;
  if (!Tt(o))
    throw new Error("merge operation is not supported for links");
  const a = Ie(t.eav, [r, s]);
  if (!a) return;
  const c = (d = a.values().next()) == null ? void 0 : d.value;
  if (!c) return;
  const u = c[2], f = dn(u, i), l = [
    r,
    s,
    f,
    bn(t, o, c)
  ];
  $(t.eav, [r, s], /* @__PURE__ */ new Map([[f, l]]));
}
function Ye(t, e) {
  var c, u;
  const [n, r] = e, s = we(t, [n]);
  if (!s)
    return;
  const [i] = s, o = t.eav.get(i);
  if (o) {
    for (const f of o.keys()) {
      const l = t.attrs[f];
      l && l["on-delete-reverse"] === "cascade" && j(o.get(f), 1).forEach(
        ([d, h, p]) => {
          var T;
          return Ye(t, [p, (T = l["reverse-identity"]) == null ? void 0 : T[1]]);
        }
      ), // Fall back to deleting everything if we've rehydrated tx-steps from
      // the store that didn't set `etype` in deleteEntity
      (!r || // If we don't know about the attr, let's just get rid of it
      !l || // Make sure it matches the etype
      ((c = l["forward-identity"]) == null ? void 0 : c[1]) === r) && (q(t.aev, [f, i]), q(t.eav, [i, f]));
    }
    o.size === 0 && q(t.eav, [i]);
  }
  const a = t.vae.get(i) && j(t.vae.get(i), 2);
  a && a.forEach((f) => {
    var T, w, b;
    const [l, d, h] = f, p = t.attrs[d];
    (!r || !p || ((T = p["reverse-identity"]) == null ? void 0 : T[1]) === r) && (q(t.eav, [l, d, h]), q(t.aev, [d, l, h]), q(t.vae, [h, d, l])), p && p["on-delete"] === "cascade" && ((w = p["reverse-identity"]) == null ? void 0 : w[1]) === r && Ye(t, [l, (b = p["forward-identity"]) == null ? void 0 : b[1]]);
  }), ((u = t.vae.get(i)) == null ? void 0 : u.size) === 0 && q(t.vae, [i]);
}
function Tn(t, e) {
  const n = pn(
    t.attrs,
    e,
    t.useDateObjects
  );
  Object.keys(n).forEach((r) => {
    t[r] = n[r];
  });
}
function Ir(t, [e]) {
  t.attrs[e.id] = e, wt(t);
}
function wn(t) {
  return j(t.eav, 3);
}
function $r(t, [e]) {
  if (!t.attrs[e]) return;
  const n = wn(t).filter(([r, s]) => s !== e);
  delete t.attrs[e], wt(t), Tn(t, n);
}
function jr(t, [e]) {
  const n = t.attrs[e.id];
  n && (t.attrs[e.id] = { ...n, ...e }, wt(t), Tn(t, wn(t)));
}
function Dr(t, e) {
  const [n, ...r] = e;
  switch (n) {
    case "add-triple":
      Pr(t, r);
      break;
    case "deep-merge-triple":
      vr(t, r);
      break;
    case "retract-triple":
      Cr(t, r);
      break;
    case "delete-entity":
      Ye(t, r);
      break;
    case "add-attr":
      Ir(t, r);
      break;
    case "delete-attr":
      $r(t, r);
      break;
    case "update-attr":
      jr(t, r);
      break;
    case "restore-attr":
      break;
    case "rule-params":
      break;
    default:
      throw new Error(`unhandled transaction action: ${n}`);
  }
}
function j(t, e, n = []) {
  if (!t || e === 0)
    return n;
  if (e === 1) {
    for (const r of t.values())
      n.push(r);
    return n;
  }
  for (const r of t.values())
    j(r, e - 1, n);
  return n;
}
function _e(t, e, n) {
  var i, o;
  const r = [];
  if (n != null && n.hasOwnProperty("$not")) {
    for (const a of e.keys())
      n.$not !== a && r.push(e.get(a));
    return r;
  }
  if (n != null && n.hasOwnProperty("$isNull")) {
    const { attrId: a, isNull: c, reverse: u } = n.$isNull;
    if (u)
      for (const f of e.keys()) {
        const l = t.vae.get(f), d = !l || ((i = l.get(a)) == null ? void 0 : i.get(null)) || !l.get(a);
        (c ? d : !d) && r.push(e.get(f));
      }
    else {
      const f = t.aev.get(a);
      for (const l of e.keys()) {
        const d = !f || ((o = f.get(l)) == null ? void 0 : o.get(null)) || !f.get(l);
        (c ? d : !d) && r.push(e.get(l));
      }
    }
    return r;
  }
  if (n != null && n.$comparator)
    return j(e, 1).filter(n.$op);
  const s = n.in || n.$in || [n];
  for (const a of s) {
    const c = e.get(a);
    c && r.push(c);
  }
  return r;
}
function Rr(t, e, n) {
  let r = "";
  return t !== void 0 && (r += "e"), e !== void 0 && (r += "a"), n !== void 0 && (r += "v"), r;
}
function Ur(t, [e, n, r]) {
  var i, o;
  switch (Rr(e, n, r)) {
    case "e": {
      const a = t.eav.get(e);
      return j(a, 2);
    }
    case "ea": {
      const a = (i = t.eav.get(e)) == null ? void 0 : i.get(n);
      return j(a, 1);
    }
    case "eav": {
      const a = (o = t.eav.get(e)) == null ? void 0 : o.get(n);
      return a ? _e(t, a, r) : [];
    }
    case "ev": {
      const a = t.eav.get(e);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(..._e(t, u, r));
      return c;
    }
    case "a": {
      const a = t.aev.get(n);
      return j(a, 2);
    }
    case "av": {
      const a = t.aev.get(n);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(..._e(t, u, r));
      return c;
    }
    case "v": {
      const a = [];
      for (const c of t.eav.values())
        for (const u of c.values())
          a.push(..._e(t, u, r));
      return a;
    }
    default:
      return j(t.eav, 3);
  }
}
function xr(t, e, n) {
  var s;
  const r = {};
  for (const [i, o] of e.entries()) {
    const a = (s = t.eav.get(n)) == null ? void 0 : s.get(o.id), c = j(a, 1);
    for (const u of c)
      r[i] = u[2];
  }
  return r;
}
function ae(t, e, n) {
  var r;
  return (r = t.attrIndexes.forwardIdents.get(e)) == null ? void 0 : r.get(n);
}
function gn(t, e, n) {
  var r;
  return (r = t.attrIndexes.revIdents.get(e)) == null ? void 0 : r.get(n);
}
function Lr(t, e) {
  return t.attrIndexes.blobAttrs.get(e);
}
function _n(t, e) {
  var r;
  const n = t.attrIndexes.primaryKeys.get(e);
  return n || ((r = t.attrIndexes.forwardIdents.get(e)) == null ? void 0 : r.get("id"));
}
function Nr(t, e) {
  const n = we(t, e);
  if (!n)
    return;
  const [r, s, i] = n;
  if (fe(t.attrs, s))
    return Ie(t.eav, [r, s]);
}
function Fr(t, e) {
  const n = e.filter(
    ([r, s, i, o, a]) => {
      if (r !== "add-triple" && r !== "deep-merge-triple")
        return !0;
      const c = a == null ? void 0 : a.mode;
      if (c !== "create" && c !== "update")
        return !0;
      let u = !1;
      const f = fe(t.attrs, i);
      if (f) {
        const l = _n(t, f["forward-identity"][1]);
        u = !!Nr(t, [s, l.id, s]);
      }
      return !(c === "create" && u || c === "update" && !u);
    }
  );
  return Be(t, (r) => {
    n.forEach((s) => {
      Dr(r, s);
    });
  });
}
function qr(t) {
  return typeof t == "string" && t.startsWith("?");
}
function Vr(t, e, n) {
  if (n.hasOwnProperty(t)) {
    const r = n[t];
    return mn(r, e, n);
  }
  return { ...n, [t]: e };
}
function Rt(t, e, n) {
  return t === e ? n : null;
}
function zr(t) {
  switch (typeof t) {
    case "string":
      return t.startsWith("?") ? Vr : Rt;
    default:
      return Rt;
  }
}
const Wr = [
  "in",
  "$in",
  "$not",
  "$isNull",
  "$comparator"
  // covers all of $gt, $lt, etc.
];
function Kr(t) {
  for (const e of Wr)
    if (t.hasOwnProperty(e))
      return !0;
  return !1;
}
function mn(t, e, n) {
  return n ? typeof t == "object" ? Kr(t) ? n : null : zr(t)(t, e, n) : null;
}
function Qr(t, e, n) {
  return t.reduce((r, s, i) => {
    const o = e[i];
    return mn(s, o, r);
  }, n);
}
function Gr(t, e, n) {
  return Jr(t, e, n).map((r) => Qr(e, r, n)).filter((r) => r);
}
function Br(t, e, n) {
  return e.or ? e.or.patterns.flatMap((r) => Ze(t, r, n)) : e.and ? e.and.patterns.reduce((r, s) => Ze(t, s, r), n) : n.flatMap((r) => Gr(t, e, r));
}
function Ze(t, e, n = [{}]) {
  return e.reduce((r, s) => Br(t, s, r), n);
}
function gt(t, e) {
  return Array.isArray(e) ? e.map((n) => gt(t, n)) : qr(e) ? t[e] : e;
}
function Hr(t, { find: e, where: n }) {
  return Ze(t, n).map((s) => gt(s, e));
}
function Jr(t, e, n) {
  return Ur(t, gt(n, e));
}
const Yr = /^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000|ffffffff-ffff-ffff-ffff-ffffffffffff)$/i;
function ce(t) {
  return typeof t == "string" && Yr.test(t);
}
const M = [];
for (let t = 0; t < 256; ++t)
  M.push((t + 256).toString(16).slice(1));
function Zr(t, e = 0) {
  return (M[t[e + 0]] + M[t[e + 1]] + M[t[e + 2]] + M[t[e + 3]] + "-" + M[t[e + 4]] + M[t[e + 5]] + "-" + M[t[e + 6]] + M[t[e + 7]] + "-" + M[t[e + 8]] + M[t[e + 9]] + "-" + M[t[e + 10]] + M[t[e + 11]] + M[t[e + 12]] + M[t[e + 13]] + M[t[e + 14]] + M[t[e + 15]]).toLowerCase();
}
let xe;
const Xr = new Uint8Array(16);
function es() {
  if (!xe) {
    if (typeof crypto > "u" || !crypto.getRandomValues)
      throw new Error("crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported");
    xe = crypto.getRandomValues.bind(crypto);
  }
  return xe(Xr);
}
const ts = typeof crypto < "u" && crypto.randomUUID && crypto.randomUUID.bind(crypto), Ut = { randomUUID: ts };
function ns(t, e, n) {
  var s;
  if (Ut.randomUUID && !t)
    return Ut.randomUUID();
  t = t || {};
  const r = t.random ?? ((s = t.rng) == null ? void 0 : s.call(t)) ?? es();
  if (r.length < 16)
    throw new Error("Random bytes length must be >= 16");
  return r[6] = r[6] & 15 | 64, r[8] = r[8] & 63 | 128, Zr(r);
}
function xt(t) {
  const e = t.replace(/-/g, ""), n = [];
  for (let r = 0; r < e.length; r += 2)
    n.push(parseInt(e.substring(r, r + 2), 16));
  return n;
}
function rs(t, e) {
  for (let n = 0; n < t.length; n++) {
    if (t[n] < e[n]) return -1;
    if (t[n] > e[n]) return 1;
  }
  return 0;
}
function ss(t, e) {
  return rs(xt(t), xt(e));
}
function k() {
  return ns();
}
function is(t, e) {
  return t.localeCompare(e);
}
function os() {
  let t = is;
  if (typeof Intl == "object" && Intl.hasOwnProperty("Collator"))
    try {
      t = Intl.Collator("en-US").compare;
    } catch {
    }
  return t;
}
const as = os();
let cs = 0;
function pe(t) {
  return $e(`_${t}`, cs++);
}
function $e(t, e) {
  return `?${t}-${e}`;
}
class ue extends Error {
  constructor(e) {
    super(e), this.name = "AttrNotFoundError";
  }
}
function us(t, e) {
  const n = _n(t, e);
  if (!n)
    throw new ue(`Could not find id attr for ${e}`);
  return n;
}
function Lt(t, e, n, r) {
  return [fs(t, e, n, r)];
}
function fs(t, e, n, r) {
  return [
    t(n, r),
    us(e, n).id,
    t(n, r),
    t("time", r)
  ];
}
function ls(t, e, n) {
  return t.map((r) => r === e ? n : r);
}
function Sn(t, e, n, r, s) {
  const i = ae(e, n, s), o = gn(e, n, s), a = i || o;
  if (!a)
    throw new ue(`Could not find attr for ${[n, s]}`);
  if (a["value-type"] !== "ref")
    throw new Error(`Attr ${a.id} is not a ref`);
  const [c, u] = a["forward-identity"], [f, l] = a["reverse-identity"], d = r + 1, h = i ? [
    t(u, r),
    a.id,
    t(l, d),
    pe("time")
  ] : [
    t(u, d),
    a.id,
    t(l, r),
    pe("time")
  ];
  return [i ? l : u, d, h, a, !!i];
}
function Nt(t, e) {
  if (typeof e != "string")
    return function(o) {
      return !1;
    };
  const r = e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&").replace(/%/g, ".*").replace(/_/g, "."), s = new RegExp(
    `^${r}$`,
    t ? void 0 : "i"
  );
  return function(o) {
    return typeof o != "string" ? !1 : s.test(o);
  };
}
function ds(t, e) {
  if (typeof e != "object" || e.hasOwnProperty("$in") || e.hasOwnProperty("in"))
    return e;
  const n = t["checked-data-type"] === "date";
  if (e.hasOwnProperty("$gt"))
    return {
      $comparator: !0,
      $op: n ? function(s) {
        return new Date(s[2]) > new Date(e.$gt);
      } : function(s) {
        return s[2] > e.$gt;
      }
    };
  if (e.hasOwnProperty("$gte"))
    return {
      $comparator: !0,
      $op: n ? function(s) {
        return new Date(s[2]) >= new Date(e.$gte);
      } : function(s) {
        return s[2] >= e.$gte;
      }
    };
  if (e.hasOwnProperty("$lt"))
    return {
      $comparator: !0,
      $op: n ? function(s) {
        return new Date(s[2]) < new Date(e.$lt);
      } : function(s) {
        return s[2] < e.$lt;
      }
    };
  if (e.hasOwnProperty("$lte"))
    return {
      $comparator: !0,
      $op: n ? function(s) {
        return new Date(s[2]) <= new Date(e.$lte);
      } : function(s) {
        return s[2] <= e.$lte;
      }
    };
  if (e.hasOwnProperty("$like")) {
    const r = Nt(!0, e.$like);
    return {
      $comparator: !0,
      $op: function(i) {
        return r(i[2]);
      }
    };
  }
  if (e.hasOwnProperty("$ilike")) {
    const r = Nt(!1, e.$ilike);
    return {
      $comparator: !0,
      $op: function(i) {
        return r(i[2]);
      }
    };
  }
  return e;
}
function hs(t, e, n, r, s, i) {
  const o = ae(e, n, s), a = gn(e, n, s), c = o || a;
  if (!c)
    throw new ue(
      `No attr for etype = ${n} label = ${s}`
    );
  if (i != null && i.hasOwnProperty("$isNull")) {
    const u = ae(e, n, "id");
    if (!u)
      throw new ue(
        `No attr for etype = ${n} label = id`
      );
    return [
      t(n, r),
      u.id,
      { $isNull: { attrId: c.id, isNull: i.$isNull, reverse: !o } },
      pe("time")
    ];
  }
  return o ? [
    t(n, r),
    c.id,
    ds(c, i),
    pe("time")
  ] : [i, c.id, t(n, r), pe("time")];
}
function ps(t, e, n, r, s) {
  const [i, o, a] = s.reduce(
    (c, u) => {
      const [f, l, d] = c, [h, p, T] = Sn(
        t,
        e,
        f,
        l,
        u
      );
      return [h, p, [...d, T]];
    },
    [n, r, []]
  );
  return [i, o, a];
}
function Xe(t, e, n, r, s, i) {
  const o = s.slice(0, s.length - 1), a = s[s.length - 1], [c, u, f] = ps(
    t,
    e,
    n,
    r,
    o
  ), l = hs(
    t,
    e,
    c,
    u,
    a,
    i
  );
  return f.concat([l]);
}
function ys(t, e) {
  return e ? [e].concat(t) : t;
}
function bs([t, e]) {
  return t === "or" && Array.isArray(e);
}
function Ts([t, e]) {
  return t === "and" && Array.isArray(e);
}
function ws(t, e, n) {
  return (r, s) => {
    const i = t(r, s);
    return e == i ? i : `${i}-${n}`;
  };
}
function Ft(t, e, n, r, s, i) {
  const o = t(r, s), a = i.map((c, u) => {
    const f = ws(t, o, u);
    return On(f, n, r, s, c);
  });
  return { [e]: { patterns: a, joinSym: o } };
}
function gs(t) {
  const e = [];
  for (let n = 1; n <= t.length; n++)
    e.push(t.slice(0, n));
  return e;
}
function qt(t, e, n, r, s) {
  return gs(s).map(
    (i) => Xe(t, e, n, r, i, { $isNull: !0 })
  );
}
function On(t, e, n, r, s) {
  return Object.entries(s).flatMap(([i, o]) => {
    if (bs([i, o]))
      return Ft(t, "or", e, n, r, o);
    if (Ts([i, o]))
      return Ft(t, "and", e, n, r, o);
    if (i === "$entityIdStartsWith")
      return [];
    const a = i.split(".");
    if (o != null && o.hasOwnProperty("$ne") && (o = { ...o, $not: o.$ne }, delete o.$ne), o != null && o.hasOwnProperty("$not")) {
      const c = Xe(t, e, n, r, a, o), u = qt(
        t,
        e,
        n,
        r,
        a
      );
      return [
        {
          or: {
            patterns: [c, ...u],
            joinSym: t(n, r)
          }
        }
      ];
    }
    return o != null && o.hasOwnProperty("$isNull") && o.$isNull === !0 && a.length > 1 ? [
      {
        or: {
          patterns: qt(
            t,
            e,
            n,
            r,
            a
          ),
          joinSym: t(n, r)
        }
      }
    ] : Xe(t, e, n, r, a, o);
  });
}
function _s(t, e, n, r) {
  const s = $e;
  return r ? On(s, t, e, n, r).concat(Lt(s, t, e, n)) : Lt(s, t, e, n);
}
function ms(t, e, n) {
  return [t(e, n), t("time", n)];
}
function Ss(t, e, n, r, s, i) {
  const [o, a, c, u, f] = Sn(
    t,
    e,
    n,
    r,
    s
  ), l = ls(c, t(n, r), i);
  return [o, a, l, u, f];
}
function Os(t, e, { etype: n, level: r, form: s }, i) {
  const o = Object.keys(s).filter((a) => a !== "$");
  return o.length ? Object.entries(i).map(function([c, u]) {
    return o.map(function(d) {
      var p, T, w;
      const h = !!(e.cardinalityInference && ((w = (T = (p = e.linkIndex) == null ? void 0 : p[n]) == null ? void 0 : T[d]) != null && w.isSingular));
      try {
        const [b, _, S] = Ss(
          t,
          e,
          n,
          r,
          d,
          c
        ), N = En(e, {
          etype: b,
          level: _,
          form: s[d],
          join: S
        }), m = h ? N[0] : N;
        return { [d]: m };
      } catch (b) {
        if (b instanceof ue)
          return { [d]: h ? void 0 : [] };
        throw b;
      }
    }).reduce(function(d, h) {
      return { ...d, ...h };
    }, u);
  }) : Object.values(i);
}
function As(t, e, n) {
  return n === "string" ? as(t, e) : t > e ? 1 : -1;
}
function ke([t, e], [n, r], s) {
  return e === r || e == null && r == null ? ss(t, n) : r == null ? 1 : e == null ? -1 : As(e, r, s);
}
function et(t) {
  return t == null ? t : new Date(t).getTime();
}
function Es(t, e, n, r) {
  var p;
  const [s, i, o, a] = t, c = n === "desc" ? 1 : -1;
  if (((p = e["forward-identity"]) == null ? void 0 : p[2]) === "id")
    return ke(r, [s, a], null) === c;
  const [u, f] = r, l = e["checked-data-type"], d = l === "date" ? et(f) : f, h = l === "date" ? et(o) : o;
  return ke([u, d], [s, h], l) === c;
}
function Ms(t, e) {
  const n = e[1];
  return t.attrs[n];
}
function Cs(t, e, n) {
  const r = Object.keys(n)[0];
  return ae(t, e, r);
}
function ks(t, e, n, r) {
  if (n)
    return Ms(t, n);
  if (r)
    return Cs(t, e, r);
}
function Ps(t, e, n) {
  var s, i;
  if (!Array.isArray(n.fields))
    return Lr(t, e);
  const r = /* @__PURE__ */ new Map();
  for (const o of n.fields) {
    const a = ae(t, e, o), c = (s = a == null ? void 0 : a["forward-identity"]) == null ? void 0 : s[2];
    c && Tt(a) && r.set(c, a);
  }
  if (!r.has("id")) {
    const o = ae(t, e, "id"), a = (i = o == null ? void 0 : o["forward-identity"]) == null ? void 0 : i[2];
    a && r.set(a, o);
  }
  return r;
}
function vs(t, { etype: e, pageInfo: n, dq: r, form: s }) {
  var h, p;
  const i = (h = s == null ? void 0 : s.$) == null ? void 0 : h.order, o = An(s), a = Is(s);
  let c = Hr(t, r);
  const u = n == null ? void 0 : n["start-cursor"], f = ks(t, e, u, i);
  if (f && ((p = f == null ? void 0 : f["forward-identity"]) == null ? void 0 : p[2]) !== "id") {
    const T = f["checked-data-type"] === "date", w = f.id;
    c = c.map(([b]) => {
      var S, N, m, P, R;
      let _ = (R = (P = (m = (N = (S = t.eav.get(b)) == null ? void 0 : S.get(w)) == null ? void 0 : N.values()) == null ? void 0 : m.next()) == null ? void 0 : P.value) == null ? void 0 : R[2];
      return T && (_ = et(_)), [b, _];
    });
  }
  c.sort(
    a === "asc" ? function(w, b) {
      return ke(w, b, f == null ? void 0 : f["checked-data-type"]);
    } : function(w, b) {
      return ke(b, w, f == null ? void 0 : f["checked-data-type"]);
    }
  );
  let l = {};
  const d = Ps(t, e, r);
  for (const T of c) {
    const [w] = T;
    if (l[w] || !o && u && f && Es(u, f, a, T))
      continue;
    const b = xr(t, d, w);
    b && (l[w] = b);
  }
  return l;
}
function Is(t) {
  var n;
  const e = (n = t.$) == null ? void 0 : n.order;
  return e && e[Object.keys(e)[0]] || "asc";
}
function An(t) {
  var s, i, o;
  const e = (s = t.$) == null ? void 0 : s.offset, n = (i = t.$) == null ? void 0 : i.before, r = (o = t.$) == null ? void 0 : o.after;
  return !e && !n && !r;
}
function $s(t, { etype: e, level: n, form: r, join: s, pageInfo: i }) {
  var l, d, h, p, T;
  if (!An(r) && (!i || !i["start-cursor"]))
    return [];
  const o = ys(_s(t, e, n, (l = r.$) == null ? void 0 : l.where), s), a = ms($e, e, n), c = (d = r.$) == null ? void 0 : d.fields, u = vs(t, {
    etype: e,
    pageInfo: i,
    form: r,
    dq: { where: o, find: a, fields: c }
  }), f = ((h = r.$) == null ? void 0 : h.limit) || ((p = r.$) == null ? void 0 : p.first) || ((T = r.$) == null ? void 0 : T.last);
  if (f != null) {
    const w = Object.entries(u);
    return w.length <= f ? u : Object.fromEntries(w.slice(0, f));
  }
  return u;
}
function js(t, e) {
  try {
    return $s(t, e);
  } catch (n) {
    if (n instanceof ue)
      return {};
    throw n;
  }
}
function En(t, e) {
  const n = js(t, e);
  return Os($e, t, e, n);
}
function Ds(t) {
  const e = {};
  for (const [n, r] of Object.entries(t))
    e[n] = {
      startCursor: r["start-cursor"],
      endCursor: r["end-cursor"],
      hasNextPage: r["has-next-page?"],
      hasPreviousPage: r["has-previous-page?"]
    };
  return e;
}
function Rs({ store: t, pageInfo: e, aggregate: n }, r) {
  const i = { data: Object.keys(r).reduce(function(a, c) {
    return n != null && n[c] || c === "$$ruleParams" || (a[c] = En(t, {
      etype: c,
      form: r[c],
      level: 0,
      pageInfo: e == null ? void 0 : e[c]
    })), a;
  }, {}) };
  return e && (i.pageInfo = Ds(e)), n && (i.aggregate = n), i;
}
function Us() {
  const e = {
    __etype: 1,
    __ops: 1,
    create: 1,
    update: 1,
    link: 1,
    unlink: 1,
    delete: 1,
    merge: 1,
    ruleParams: 1
  };
  return new Set(Object.keys(e));
}
const xs = Us();
function tt(t, e, n) {
  const r = {
    __etype: t,
    __ops: n
  };
  return new Proxy(r, {
    get: (s, i) => {
      if (i === "__ops") return n;
      if (i === "__etype") return t;
      if (xs.has(i))
        return (o, a) => tt(t, e, [
          ...n,
          a ? [i, t, e, o, a] : [i, t, e, o]
        ]);
    }
  });
}
function wo(t, e) {
  return `lookup__${t}__${JSON.stringify(e)}`;
}
function Pe(t) {
  return t.startsWith("lookup__");
}
function Mn(t) {
  const [e, n, ...r] = t.split("__");
  return [n, JSON.parse(r.join("__"))];
}
function Ls(t) {
  return new Proxy(
    {
      __etype: t
    },
    {
      get(e, n) {
        if (n === "__etype") return t;
        const r = n;
        return Pe(r) ? tt(t, Mn(r), []) : tt(t, r, []);
      }
    }
  );
}
function Cn() {
  return new Proxy(
    {},
    {
      get(t, e) {
        return Ls(e);
      }
    }
  );
}
const go = Cn();
function Ns(t) {
  return t.__ops;
}
function Fs(t, e) {
  const { attrIdMap: n, refSwapAttrIds: r } = t, s = [];
  for (const o of e) {
    const a = n[o];
    if (a)
      s.push(a);
    else if (Array.isArray(o) && o.length == 2 && n[o[0]]) {
      const [c, u] = o;
      s.push([n[c], u]);
    } else
      s.push(o);
  }
  const [i] = e;
  if ((i === "add-triple" || i === "retract-triple") && r.has(e[2])) {
    const o = s[1];
    s[1] = s[3], s[3] = o;
  }
  return s;
}
function C(t, e, n) {
  return Object.values(t).find((r) => {
    const [s, i, o] = r["forward-identity"];
    return i === e && o === n;
  });
}
function ee(t, e, n) {
  return Object.values(t).find((r) => {
    const s = r["reverse-identity"];
    if (!s) return !1;
    const [i, o, a] = s;
    return o === e && a === n;
  });
}
function qs(t) {
  if (Array.isArray(t))
    return t;
  const e = Object.entries(t);
  if (e.length !== 1)
    throw new Error(
      "lookup must be an object with a single unique attr and value."
    );
  return e[0];
}
function nt(t, e, n) {
  return n.indexOf(".") !== -1 && // attr names can have `.` in them, so use the attr we find with a `.`
  // before assuming it's a ref lookup.
  !C(t, e, n);
}
function rt(t) {
  const [e, n, ...r] = t.split(".");
  if (r.length > 0 || n !== "id")
    throw new Error(`${t} is not a valid lookup attribute.`);
  return e;
}
function Vs(t, e, n) {
  if (!nt(t, e, n))
    return C(t, e, n);
  const r = rt(n), s = C(t, e, r) || ee(t, e, r);
  if (s && s["value-type"] !== "ref")
    throw new Error(`${n} does not reference a valid link attribute.`);
  return s;
}
function st(t) {
  return typeof t == "string" && !Pe(t) ? null : typeof t == "string" && Pe(t) ? Mn(t) : qs(t);
}
function D(t, e, n) {
  const r = st(n);
  if (r === null)
    return n;
  const [s, i] = r, o = Vs(t, e, s);
  if (!o || !o["unique?"])
    throw new Error(`${s} is not a unique attribute.`);
  return [o.id, i];
}
function kn(t, e, n, r) {
  const s = D(t, e, n);
  return Array.isArray(s) ? [[
    "add-triple",
    s,
    C(t, e, "id").id,
    s
  ]].concat(r) : r;
}
function zs({ attrs: t }, [e, n, r]) {
  const s = Object.entries(r).flatMap(([i, o]) => {
    const a = Array.isArray(o) ? o : [o], c = C(t, e, i), u = ee(t, e, i);
    return a.map((f) => c ? [
      "add-triple",
      D(t, e, n),
      c.id,
      D(t, c["reverse-identity"][1], f)
    ] : [
      "add-triple",
      D(t, u["forward-identity"][1], f),
      u.id,
      D(t, e, n)
    ]);
  });
  return kn(t, e, n, s);
}
function Ws({ attrs: t }, [e, n, r]) {
  const s = Object.entries(r).flatMap(([i, o]) => {
    const a = Array.isArray(o) ? o : [o], c = C(t, e, i), u = ee(t, e, i);
    return a.map((f) => c ? [
      "retract-triple",
      D(t, e, n),
      c.id,
      D(t, c["reverse-identity"][1], f)
    ] : [
      "retract-triple",
      D(t, u["forward-identity"][1], f),
      u.id,
      D(t, e, n)
    ]);
  });
  return kn(t, e, n, s);
}
function Ks(t, e, n) {
  if (Array.isArray(n)) {
    const [r, s] = n;
    for (const i of t || []) {
      const o = i == null ? void 0 : i.aev.get(r);
      if (o) {
        for (const [a, c, u] of j(o, 2))
          if (u === s)
            return !0;
      }
    }
  } else
    for (const r of t || []) {
      const s = r == null ? void 0 : r.eav.get(n);
      if (s) {
        for (const i of s.keys())
          if (r.attrs[i]["forward-identity"][1] == e)
            return !0;
      }
    }
  return !1;
}
function Pn({ stores: t, attrs: e }, [n, r, s, i]) {
  return (i == null ? void 0 : i.upsert) === !1 ? { mode: "update" } : (i == null ? void 0 : i.upsert) === !0 ? null : Ks(t, n, r) ? { mode: "update" } : null;
}
function Qs(t, e) {
  const { attrs: n } = t, [r, s, i, o] = e, a = yt(i), c = D(n, r, s);
  return [["id", c]].concat(Object.entries(a)).map(([f, l]) => {
    const d = C(n, r, f);
    return d["checked-data-type"] === "date" && t.useDateObjects && (l = ve(l)), ["add-triple", c, d.id, l, { mode: "create" }];
  });
}
function Gs(t, e) {
  const { attrs: n } = t, [r, s, i, o] = e, a = yt(i), c = D(n, r, s), u = Pn(t, [r, c, i, o]);
  return [["id", c]].concat(Object.entries(a)).map(([l, d]) => {
    const h = C(n, r, l);
    return h["checked-data-type"] === "date" && t.useDateObjects && (d = ve(d)), [
      "add-triple",
      c,
      h.id,
      d,
      ...u ? [u] : []
    ];
  });
}
function Bs({ attrs: t }, [e, n]) {
  return [["delete-entity", D(t, e, n), e]];
}
function Hs(t, e) {
  const { attrs: n } = t, [r, s, i, o] = e, a = yt(i), c = D(n, r, s), u = Pn(t, [r, c, i, o]), f = Object.entries(a).map(([d, h]) => {
    const p = C(n, r, d);
    return [
      "deep-merge-triple",
      c,
      p.id,
      h,
      ...u ? [u] : []
    ];
  });
  return [[
    "add-triple",
    c,
    C(n, r, "id").id,
    c,
    ...u ? [u] : []
  ]].concat(f);
}
function Js({ attrs: t }, [e, n, r]) {
  return [["rule-params", D(t, e, n), e, r]];
}
function Ys(t) {
  const [e, n, r, s, i] = t;
  if (!s)
    return t;
  const o = { ...s };
  return delete o.id, [e, n, r, o, ...i ? [i] : []];
}
function Zs(t, e) {
  const [n, ...r] = Ys(e);
  switch (n) {
    case "merge":
      return Hs(t, r);
    case "create":
      return Qs(t, r);
    case "update":
      return Gs(t, r);
    case "link":
      return zs(t, r);
    case "unlink":
      return Ws(t, r);
    case "delete":
      return Bs(t, r);
    case "ruleParams":
      return Js(t, r);
    default:
      throw new Error(`unsupported action ${n}`);
  }
}
function Xs(t) {
  switch (t) {
    case "string":
    case "date":
    case "boolean":
    case "number":
      return t;
    default:
      return;
  }
}
function ei(t, e, n) {
  var a, c;
  const r = (c = (a = t.entities[e]) == null ? void 0 : a.attrs) == null ? void 0 : c[n];
  if (n === "id") return null;
  if (!r)
    throw new Error(`${e}.${n} does not exist in your schema`);
  const { unique: s, indexed: i } = r == null ? void 0 : r.config, o = Xs(r == null ? void 0 : r.valueType);
  return {
    "index?": i,
    "unique?": s,
    "checked-data-type": o
  };
}
function me(t, e, n, r) {
  const s = t ? ei(t, e, n) : null, i = k(), a = [k(), e, n];
  return {
    id: i,
    "forward-identity": a,
    "value-type": "blob",
    cardinality: "one",
    "unique?": !1,
    "index?": !1,
    isUnsynced: !0,
    ...s || {},
    ...r || {}
  };
}
function ti(t, e, n) {
  return Object.values(t.links).find((s) => s.forward.on === e && s.forward.label === n || s.reverse.on === e && s.reverse.label === n);
}
function ni(t, e, n) {
  const r = ti(t, e, n);
  if (!r)
    throw new Error(`Couldn't find the link ${e}.${n} in your schema`);
  const { forward: s, reverse: i } = r;
  return {
    "forward-identity": [k(), s.on, s.label],
    "reverse-identity": [k(), i.on, i.label],
    cardinality: s.has === "one" ? "one" : "many",
    "unique?": i.has === "one",
    "on-delete": s.onDelete,
    "on-delete-reverse": i.onDelete
  };
}
function Vt(t, e, n, r) {
  const s = t ? ni(t, e, n) : null, i = k(), o = [k(), e, n], a = [k(), n, e];
  return {
    id: i,
    "forward-identity": o,
    "reverse-identity": a,
    "value-type": "ref",
    cardinality: "many",
    "unique?": !1,
    "index?": !1,
    isUnsynced: !0,
    ...s || {},
    ...r || {}
  };
}
const ri = /* @__PURE__ */ new Set(["create", "update", "merge", "link", "unlink"]), si = /* @__PURE__ */ new Set(["link", "unlink"]), ii = /* @__PURE__ */ new Set(["create", "update", "merge"]), oi = /* @__PURE__ */ new Set([
  "link",
  "unlink",
  "create",
  "update",
  "merge",
  "delete",
  "ruleParams"
]), it = { "unique?": !0, "index?": !0 }, ai = { ...it, cardinality: "one" };
function ci(t) {
  const e = [], [n, r, s, i] = t;
  if (!oi.has(n))
    return e;
  const o = st(s);
  if (o && e.push({ etype: r, lookupPair: o }), n === "link")
    for (const [a, c] of Object.entries(i)) {
      const u = Array.isArray(c) ? c : [c];
      for (const f of u) {
        const l = st(f);
        l && e.push({
          etype: r,
          lookupPair: l,
          linkLabel: a
        });
      }
    }
  return e;
}
function ui({ attrs: t, schema: e }, n) {
  var u, f;
  const [r, s, i] = [/* @__PURE__ */ new Set(), { ...t }, []];
  function o(l) {
    s[l.id] = l, i.push(["add-attr", l]), r.add(l.id);
  }
  function a(l) {
    l != null && l.isUnsynced && !r.has(l.id) && (i.push(["add-attr", l]), r.add(l.id));
  }
  function c(l, d) {
    const h = C(s, l, d), p = ee(s, l, d);
    a(h), a(p), !h && !p && o(Vt(e, l, d, ai));
  }
  for (const l of n)
    for (const { etype: d, lookupPair: h, linkLabel: p } of ci(l)) {
      const T = h[0];
      if (p) {
        c(d, p);
        const w = C(s, d, p), b = ee(s, d, p);
        a(w), a(b);
        const _ = ((u = w == null ? void 0 : w["reverse-identity"]) == null ? void 0 : u[1]) || ((f = b == null ? void 0 : b["forward-identity"]) == null ? void 0 : f[1]) || p;
        if (nt(s, _, T))
          c(_, rt(T));
        else {
          const S = C(s, _, T);
          S || o(
            me(e, _, T, it)
          ), a(S);
        }
      } else if (nt(s, d, T))
        c(d, rt(T));
      else {
        const w = C(s, d, T);
        w || o(me(e, d, T, it)), a(w);
      }
    }
  for (const l of n) {
    const [d, h, p, T] = l;
    if (ri.has(d)) {
      const w = C(s, h, "id");
      a(w), w || o(me(e, h, "id", { "unique?": !0 }));
      for (const b of Object.keys(T)) {
        const _ = C(s, h, b);
        if (a(_), ii.has(d) && (_ || o(
          me(
            e,
            h,
            b,
            b === "id" ? { "unique?": !0 } : null
          )
        )), si.has(d)) {
          const S = ee(s, h, b);
          !_ && !S && o(Vt(e, h, b)), a(S);
        }
      }
    }
  }
  return [s, i];
}
function fi(t, e) {
  const r = (Array.isArray(e) ? e : [e]).flatMap((c) => Ns(c)), [s, i] = ui(t, r), o = { ...t, attrs: s }, a = r.flatMap((c) => Zs(o, c));
  return [...i, ...a];
}
class vn {
  constructor(e) {
    this.dbName = e, this._storeName = "kv", this._dbPromise = this._init();
  }
  _init() {
    return new Promise((e, n) => {
      const r = indexedDB.open(this.dbName, 1);
      r.onerror = (s) => {
        n(s);
      }, r.onsuccess = (s) => {
        const i = s.target;
        e(i.result);
      }, r.onupgradeneeded = (s) => {
        s.target.result.createObjectStore(this._storeName);
      };
    });
  }
  async getItem(e) {
    const n = await this._dbPromise;
    return new Promise((r, s) => {
      const a = n.transaction([this._storeName], "readonly").objectStore(this._storeName).get(e);
      a.onerror = (c) => {
        s(c);
      }, a.onsuccess = (c) => {
        a.result ? r(a.result) : r(null);
      };
    });
  }
  async setItem(e, n) {
    const r = await this._dbPromise;
    return new Promise((s, i) => {
      const c = r.transaction([this._storeName], "readwrite").objectStore(this._storeName).put(n, e);
      c.onerror = (u) => {
        i(u);
      }, c.onsuccess = (u) => {
        s();
      };
    });
  }
}
class In {
  static async getIsOnline() {
    return navigator.onLine;
  }
  static listen(e) {
    const n = () => {
      e(!0);
    }, r = () => {
      e(!1);
    };
    return addEventListener("online", n), addEventListener("offline", r), () => {
      removeEventListener("online", n), removeEventListener("offline", r);
    };
  }
}
class be extends Error {
  constructor(e, n) {
    super(e), this.hint = n;
    const r = new.target.prototype;
    Object.setPrototypeOf && Object.setPrototypeOf(this, r), Error.captureStackTrace && Error.captureStackTrace(this, be), this.name = "InstantError";
  }
  get [Symbol.toStringTag]() {
    return "InstantError";
  }
}
class je extends be {
  constructor(e) {
    var s;
    const n = ((s = e.body) == null ? void 0 : s.message) || `API Error (${e.status})`;
    super(n, e.body.hint);
    const r = new.target.prototype;
    Object.setPrototypeOf && Object.setPrototypeOf(this, r), Error.captureStackTrace && Error.captureStackTrace(this, je), this.name = "InstantAPIError", this.status = e.status, this.body = e.body;
  }
  get [Symbol.toStringTag]() {
    return "InstantAPIError";
  }
}
async function z(t, e) {
  const n = await fetch(t, e), r = await n.json();
  return n.status === 200 ? Promise.resolve(r) : Promise.reject(new je({ status: n.status, body: r }));
}
function li({
  apiURI: t,
  appId: e,
  email: n
}) {
  return z(`${t}/runtime/auth/send_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ "app-id": e, email: n })
  });
}
async function di({
  apiURI: t,
  appId: e,
  email: n,
  code: r,
  refreshToken: s
}) {
  return await z(`${t}/runtime/auth/verify_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      "app-id": e,
      email: n,
      code: r,
      ...s ? { "refresh-token": s } : {}
    })
  });
}
async function hi({
  apiURI: t,
  appId: e,
  refreshToken: n
}) {
  return await z(`${t}/runtime/auth/verify_refresh_token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      "app-id": e,
      "refresh-token": n
    })
  });
}
async function pi({
  apiURI: t,
  appId: e
}) {
  return await z(`${t}/runtime/auth/sign_in_guest`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      "app-id": e
    })
  });
}
async function zt({
  apiURI: t,
  appId: e,
  code: n,
  codeVerifier: r,
  refreshToken: s
}) {
  return await z(`${t}/runtime/oauth/token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      app_id: e,
      code: n,
      code_verifier: r,
      refresh_token: s
    })
  });
}
async function yi({
  apiURI: t,
  appId: e,
  nonce: n,
  idToken: r,
  clientName: s,
  refreshToken: i
}) {
  return await z(`${t}/runtime/oauth/id_token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      app_id: e,
      nonce: n,
      id_token: r,
      client_name: s,
      refresh_token: i
    })
  });
}
async function bi({
  apiURI: t,
  appId: e,
  refreshToken: n
}) {
  return await z(`${t}/runtime/signout`, {
    method: "POST",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify({
      app_id: e,
      refresh_token: n
    })
  });
}
async function Ti({
  apiURI: t,
  appId: e,
  path: n,
  file: r,
  refreshToken: s,
  contentType: i,
  contentDisposition: o
}) {
  const a = {
    app_id: e,
    path: n,
    authorization: `Bearer ${s}`,
    "content-type": i || r.type
  };
  return o && (a["content-disposition"] = o), await z(`${t}/storage/upload`, {
    method: "PUT",
    headers: a,
    body: r
  });
}
async function wi({
  apiURI: t,
  appId: e,
  path: n,
  refreshToken: r
}) {
  const { data: s } = await z(
    `${t}/storage/files?app_id=${e}&filename=${encodeURIComponent(n)}`,
    {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${r}`
      }
    }
  );
  return s;
}
async function gi({
  apiURI: t,
  appId: e,
  fileName: n,
  refreshToken: r,
  metadata: s = {}
}) {
  const { data: i } = await z(`${t}/storage/signed-upload-url`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${r}`
    },
    body: JSON.stringify({
      app_id: e,
      filename: n
    })
  });
  return i;
}
async function _i(t, e) {
  return (await fetch(t, {
    method: "PUT",
    body: e,
    headers: {
      "Content-Type": e.type
    }
  })).ok;
}
async function mi({
  apiURI: t,
  appId: e,
  path: n,
  refreshToken: r
}) {
  const { data: s } = await z(
    `${t}/storage/signed-download-url?app_id=${e}&filename=${encodeURIComponent(
      n
    )}`,
    {
      method: "GET",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${r}`
      }
    }
  );
  return s;
}
let _t = !1, $n = !1, jn = !1;
typeof window < "u" && typeof window.localStorage < "u" && (_t = !!window.localStorage.getItem("devBackend"), $n = !!window.localStorage.getItem("__instantLogging"), jn = !!window.localStorage.getItem("__devtoolLocalDash"));
function Wt(t, e) {
  if (!e) return t;
  const n = {};
  return e.forEach((r) => {
    n[r] = t[r];
  }), n;
}
function Si(t, e, n) {
  const r = {
    peers: {}
  };
  if (e && "user" in e ? e.user : !0) {
    const i = Wt(t.user ?? {}, e == null ? void 0 : e.keys);
    r.user = { ...i, peerId: n };
  }
  for (const i of Object.keys(t.peers ?? {})) {
    const o = (e == null ? void 0 : e.peers) === void 0, a = Array.isArray(e == null ? void 0 : e.peers) && (e == null ? void 0 : e.peers.includes(i));
    if (o || a) {
      const c = Wt(t.peers[i], e == null ? void 0 : e.keys);
      r.peers[i] = { ...c, peerId: i };
    }
  }
  return r;
}
function Oi(t, e) {
  if (t.isLoading !== e.isLoading || t.error !== e.error || (t.user || e.user) && (!t.user || !e.user || !$t(t.user, e.user)) || !ln(t.peers, e.peers)) return !0;
  for (const r of Object.keys(t.peers))
    if (!$t(t.peers[r], e.peers[r])) return !0;
  return !1;
}
class Kt {
  constructor() {
    y(this, "promise");
    y(this, "_resolve");
    y(this, "_reject");
    this.promise = new Promise((e, n) => {
      this._resolve = e, this._reject = n;
    });
  }
  resolve(e) {
    this._resolve(e);
  }
  reject(e) {
    this._reject(e);
  }
}
function Ai(t, e) {
  typeof requestIdleCallback > "u" ? t() : requestIdleCallback(t, { timeout: e });
}
class Qt {
  constructor(e, n, r, s, i = (u) => u, o = (u) => u, a = 100, c = 1e3) {
    this._subs = [], this._nextSave = null, this._persister = e, this._key = n, this._onMerge = s, this._loadedCbs = [], this._isLoading = !0, this.currentValue = r, this.serialize = i, this.parse = o, this._saveThrottleMs = a, this._pendingSaveCbs = [], this._version = 0, this._idleCallbackMaxWaitMs = c, this._load();
  }
  async _getFromStorage() {
    try {
      return this.parse(await this._persister.getItem(this._key));
    } catch (e) {
      return console.error(`Unable to read from storage for key=${this._key}`, e), null;
    }
  }
  async _load() {
    const e = await this._getFromStorage();
    this._isLoading = !1, this._onMerge(e, this.currentValue);
    for (const n of this._loadedCbs)
      n();
  }
  async waitForLoaded() {
    if (!this._isLoading)
      return;
    await new Promise((n) => {
      this._loadedCbs.push(n);
    });
  }
  isLoading() {
    return this._isLoading;
  }
  version() {
    return this._version;
  }
  async waitForSync() {
    if (!this._nextSave)
      return;
    await new Promise((n) => {
      this._pendingSaveCbs.push(n);
    });
  }
  _writeToStorage() {
    this._persister.setItem(this._key, this.serialize(this.currentValue));
    for (const e of this._pendingSaveCbs)
      e();
    this._pendingSaveCbs.length = 0;
  }
  async flush() {
    this._nextSave && (clearTimeout(this._nextSave), this._writeToStorage());
  }
  _enqueuePersist(e) {
    if (this._nextSave) {
      e && this._pendingSaveCbs.push(e);
      return;
    }
    this._nextSave = setTimeout(() => {
      Ai(() => {
        this._nextSave = null, this._writeToStorage();
      }, this._idleCallbackMaxWaitMs);
    }, this._saveThrottleMs);
  }
  set(e, n) {
    this._version++, this.currentValue = e(this.currentValue), this._isLoading ? this._loadedCbs.push(() => this._enqueuePersist(n)) : this._enqueuePersist(n);
    for (const r of this._subs)
      r(this.currentValue);
  }
  subscribe(e) {
    return this._subs.push(e), e(this.currentValue), () => {
      this._subs = this._subs.filter((n) => n !== e);
    };
  }
}
function Dn(t, e = []) {
  t.forEach((n) => {
    const { data: r } = n, { "datalog-result": s } = r, { "join-rows": i } = s;
    for (const o of i)
      for (const a of o)
        e.push(a);
    Dn(n["child-nodes"], e);
  });
}
function Gt(t) {
  const e = [];
  return Dn(t, e), e;
}
function Bt(t) {
  return Object.values(t.links).reduce((e, n) => {
    var r, s;
    return e[r = n.forward.on] ?? (e[r] = {}), e[n.forward.on][n.forward.label] = {
      isForward: !0,
      isSingular: n.forward.has === "one",
      link: n
    }, e[s = n.reverse.on] ?? (e[s] = {}), e[n.reverse.on][n.reverse.label] = {
      isForward: !1,
      isSingular: n.reverse.has === "one",
      link: n
    }, e;
  }, {});
}
const Rn = "v0.22.34";
function Ei(t) {
  return {
    info: t ? console.info.bind(console) : () => {
    },
    debug: t ? console.debug.bind(console) : () => {
    },
    error: t ? console.error.bind(console) : () => {
    }
  };
}
class I {
  constructor(e, n, r, s = { indexed: !1, unique: !1 }) {
    this.valueType = e, this.required = n, this.isIndexed = r, this.config = s, this.metadata = {};
  }
  /**
   * @deprecated Only use this temporarily for attributes that you want
   * to treat as required in frontend code but can’t yet mark as required
   * and enforced for backend
   */
  clientRequired() {
    return new I(
      this.valueType,
      !1,
      this.isIndexed,
      this.config
    );
  }
  optional() {
    return new I(
      this.valueType,
      !1,
      this.isIndexed,
      this.config
    );
  }
  unique() {
    return new I(
      this.valueType,
      this.required,
      this.isIndexed,
      {
        ...this.config,
        unique: !0
      }
    );
  }
  indexed() {
    return new I(
      this.valueType,
      this.required,
      !0,
      {
        ...this.config,
        indexed: !0
      }
    );
  }
  // clientValidate(clientValidator: (value: ValueType) => boolean) {
  //   return new DataAttrDef(this.valueType, this.required, {
  //     ...this.config,
  //     clientValidator,
  //   });
  // }
}
class De {
  constructor(e, n) {
    this.attrs = e, this.links = n;
  }
  asType() {
    return new De(this.attrs, this.links);
  }
}
class Re {
  constructor(e, n, r) {
    this.entities = e, this.links = n, this.rooms = r;
  }
  /**
   * @deprecated
   * `withRoomSchema` is deprecated. Define your schema in `rooms` directly:
   *
   * @example
   * // Before:
   * const schema = i.schema({
   *   // ...
   * }).withRoomSchema<RoomSchema>()
   *
   * // After
   * const schema = i.schema({
   *  rooms: {
   *    // ...
   *  }
   * })
   *
   * @see https://instantdb.com/docs/presence-and-topics#typesafety
   */
  withRoomSchema() {
    return new Re(
      this.entities,
      this.links,
      {}
    );
  }
}
class O extends Error {
  constructor(e, n) {
    const r = n ? `At path '${n}': ${e}` : e;
    super(r), this.name = "QueryValidationError";
  }
}
const Ht = [
  "where",
  "order",
  "limit",
  "last",
  "first",
  "offset",
  "after",
  "before",
  "fields",
  "aggregate"
], Mi = (t) => t.valueType || "unknown", Un = (t, e, n = !1) => {
  if (n || t == null) return !0;
  switch (e) {
    case "string":
      return typeof t == "string";
    case "number":
      return typeof t == "number" && !isNaN(t);
    case "boolean":
      return typeof t == "boolean";
    case "date":
      return t instanceof Date || typeof t == "string" || typeof t == "number";
    default:
      return !0;
  }
}, Ci = (t, e, n, r, s, i, o) => {
  const a = i.valueType === "json", c = (u, f, l) => {
    if (!Un(l, f, a))
      throw new O(
        `Invalid value for operator '${u}' on attribute '${r}' in entity '${s}'. Expected ${f}, but received: ${typeof l}`,
        o
      );
  };
  switch (t) {
    case "in":
    case "$in":
      if (!Array.isArray(e))
        throw new O(
          `Operator '${t}' for attribute '${r}' in entity '${s}' must be an array, but received: ${typeof e}`,
          o
        );
      for (const u of e)
        c(t, n, u);
      break;
    case "$not":
    case "$ne":
    case "$gt":
    case "$lt":
    case "$gte":
    case "$lte":
      c(t, n, e);
      break;
    case "$like":
    case "$ilike":
      if (c(t, "string", e), t === "$ilike" && !i.isIndexed)
        throw new O(
          `Operator '${t}' can only be used with indexed attributes, but '${r}' in entity '${s}' is not indexed`,
          o
        );
      break;
    case "$isNull":
      c(t, "boolean", e);
      break;
    default:
      throw new O(
        `Unknown operator '${t}' for attribute '${r}' in entity '${s}'`,
        o
      );
  }
}, oe = (t, e, n, r, s) => {
  const i = Mi(n), o = n.valueType === "json";
  if (typeof t == "object" && t !== null && !Array.isArray(t)) {
    if (o)
      return;
    const c = t;
    for (const [u, f] of Object.entries(c))
      Ci(
        u,
        f,
        i,
        e,
        r,
        n,
        `${s}.${u}`
      );
  } else if (!Un(t, i, o))
    throw new O(
      `Invalid value for attribute '${e}' in entity '${r}'. Expected ${i}, but received: ${typeof t}`,
      s
    );
}, ki = (t, e, n, r, s) => {
  const i = t.split(".");
  if (i.length < 2)
    throw new O(
      `Invalid dot notation path '${t}'. Must contain at least one dot.`,
      s
    );
  let o = n;
  for (let f = 0; f < i.length - 1; f++) {
    const l = i[f], d = r.entities[o];
    if (!d)
      throw new O(
        `Entity '${o}' does not exist in schema while traversing dot notation path '${t}'.`,
        s
      );
    const h = d.links[l];
    if (!h) {
      const p = Object.keys(d.links);
      throw new O(
        `Link '${l}' does not exist on entity '${o}' in dot notation path '${t}'. Available links: ${p.length > 0 ? p.join(", ") : "none"}`,
        s
      );
    }
    o = h.entityName;
  }
  const a = i[i.length - 1], c = r.entities[o];
  if (!c)
    throw new O(
      `Target entity '${o}' does not exist in schema for dot notation path '${t}'.`,
      s
    );
  if (a === "id") {
    if (typeof e == "string" && !ce(e))
      throw new O(
        `Invalid value for id field in entity '${o}'. Expected a UUID, but received: ${e}`,
        s
      );
    oe(
      e,
      t,
      new I("string", !1, !0),
      n,
      s
    );
    return;
  }
  const u = c.attrs[a];
  if (Object.keys(c.links).includes(a)) {
    if (typeof e == "string" && !ce(e))
      throw new O(
        `Invalid value for link '${a}' in entity '${o}'. Expected a UUID, but received: ${e}`,
        s
      );
    oe(
      e,
      t,
      new I("string", !1, !0),
      n,
      s
    );
    return;
  }
  if (!u) {
    const f = Object.keys(c.attrs);
    throw new O(
      `Attribute '${a}' does not exist on entity '${o}' in dot notation path '${t}'. Available attributes: ${f.length > 0 ? f.join(", ") + ", id" : "id"}`,
      s
    );
  }
  oe(e, t, u, n, s);
}, xn = (t, e, n, r) => {
  for (const [s, i] of Object.entries(t)) {
    if (s === "or" || s === "and") {
      if (Array.isArray(i))
        for (const u of i)
          typeof u == "object" && u !== null && xn(
            u,
            e,
            n,
            `${r}.${s}[${u}]`
          );
      continue;
    }
    if (s === "id") {
      oe(
        i,
        "id",
        new I("string", !1, !0),
        e,
        `${r}.id`
      );
      continue;
    }
    if (s.includes(".")) {
      ki(
        s,
        i,
        e,
        n,
        `${r}.${s}`
      );
      continue;
    }
    const o = n.entities[e];
    if (!o) continue;
    const a = o.attrs[s], c = o.links[s];
    if (!a && !c) {
      const u = Object.keys(o.attrs), f = Object.keys(o.links);
      throw new O(
        `Attribute or link '${s}' does not exist on entity '${e}'. Available attributes: ${u.length > 0 ? u.join(", ") : "none"}. Available links: ${f.length > 0 ? f.join(", ") : "none"}`,
        `${r}.${s}`
      );
    }
    if (a)
      oe(
        i,
        s,
        a,
        e,
        `${r}.${s}`
      );
    else if (c) {
      if (typeof i == "string" && !ce(i))
        throw new O(
          `Invalid value for link '${s}' in entity '${e}'. Expected a UUID, but received: ${i}`,
          `${r}.${s}`
        );
      const u = new I("string", !1, !0);
      oe(
        i,
        s,
        u,
        e,
        `${r}.${s}`
      );
    }
  }
}, Pi = (t, e, n, r, s = 0) => {
  for (const o of Object.keys(t))
    if (!Ht.includes(o))
      throw new O(
        `Invalid query parameter '${o}' in $ object. Valid parameters are: ${Ht.join(", ")}. Found: ${o}`,
        r
      );
  const i = [
    "limit",
    "offset",
    "before",
    "after",
    "first",
    "last"
  ];
  for (const o of i)
    if (t[o] !== void 0 && s > 0)
      throw new O(
        `'${o}' can only be used on top-level namespaces. It cannot be used in nested queries.`,
        r
      );
  if (t.where && n) {
    if (typeof t.where != "object" || t.where === null)
      throw new O(
        `'where' clause must be an object in entity '${e}', but received: ${typeof t.where}`,
        r ? `${r}.where` : void 0
      );
    xn(
      t.where,
      e,
      n,
      r ? `${r}.where` : "where"
    );
  }
}, Ln = (t, e, n, r, s = 0) => {
  var i;
  if (!t || typeof t != "object")
    throw new O(
      `Query part for entity '${e}' must be an object, but received: ${typeof t}`,
      r
    );
  for (const o of Object.keys(t))
    if (o !== "$") {
      if (n && !(o in n.entities[e].links)) {
        const c = Object.keys(n.entities[e].links);
        throw new O(
          `Link '${o}' does not exist on entity '${e}'. Available links: ${c.length > 0 ? c.join(", ") : "none"}`,
          `${r}.${o}`
        );
      }
      const a = t[o];
      if (typeof a == "object" && a !== null) {
        const c = (i = n == null ? void 0 : n.entities[e].links[o]) == null ? void 0 : i.entityName;
        c && Ln(
          a,
          c,
          n,
          `${r}.${o}`,
          s + 1
        );
      }
    } else {
      const a = t[o];
      if (typeof a != "object" || a === null)
        throw new O(
          `Query parameter '$' must be an object in entity '${e}', but received: ${typeof a}`,
          `${r}.$`
        );
      Pi(
        a,
        e,
        n,
        `${r}.$`,
        s
      );
    }
}, Jt = (t, e) => {
  if (typeof t != "object" || t === null)
    throw new O(
      `Query must be an object, but received: ${typeof t}${t === null ? " (null)" : ""}`
    );
  if (Array.isArray(t))
    throw new O(
      `Query must be an object, but received: ${typeof t}`
    );
  const n = t;
  for (const r of Object.keys(n)) {
    if (Array.isArray(t[r]))
      throw new O(
        `Query keys must be strings, but found key of type: ${typeof r}`,
        r
      );
    if (typeof r != "string")
      throw new O(
        `Query keys must be strings, but found key of type: ${typeof r}`,
        r
      );
    if (r !== "$$ruleParams") {
      if (e && !e.entities[r]) {
        const s = Object.keys(e.entities);
        throw new O(
          `Entity '${r}' does not exist in schema. Available entities: ${s.length > 0 ? s.join(", ") : "none"}`,
          r
        );
      }
      Ln(
        n[r],
        r,
        e,
        r,
        0
        // Start at depth 0 for top-level entities
      );
    }
  }
}, Yt = (t) => typeof t != "string" ? !1 : Pe(t) ? !0 : ce(t);
class F extends Error {
  constructor(e) {
    super(e), this.name = "TransactionValidationError";
  }
}
const Nn = (t) => t.length > 0 ? t.join(", ") : "none", vi = (t, e) => new F(
  `Entity '${t}' does not exist in schema. Available entities: ${Nn(e)}`
), Le = {
  string: (t) => typeof t == "string",
  number: (t) => typeof t == "number" && !isNaN(t),
  boolean: (t) => typeof t == "boolean",
  date: (t) => t instanceof Date || typeof t == "string" || typeof t == "number",
  json: () => !0
}, Ii = (t, e) => {
  var n;
  return t == null ? !0 : ((n = Le[e.valueType]) == null ? void 0 : n.call(Le, t)) ?? !1;
}, Fn = (t, e) => {
  const n = e.entities[t];
  if (!n)
    throw vi(t, Object.keys(e.entities));
  return n;
}, Ne = (t, e, n) => {
  const r = Fn(t, n);
  if (typeof e != "object" || e === null)
    throw new F(
      `Arguments for data operation on entity '${t}' must be an object, but received: ${typeof e}`
    );
  for (const [s, i] of Object.entries(e)) {
    if (s === "id") continue;
    const o = r.attrs[s];
    if (o && !Ii(i, o))
      throw new F(
        `Invalid value for attribute '${s}' in entity '${t}'. Expected ${o.valueType}, but received: ${typeof i}`
      );
  }
}, Zt = (t, e, n) => {
  const r = Fn(t, n);
  if (typeof e != "object" || e === null)
    throw new F(
      `Arguments for link operation on entity '${t}' must be an object, but received: ${typeof e}`
    );
  for (const [s, i] of Object.entries(e)) {
    if (!r.links[s]) {
      const a = Object.keys(r.links);
      throw new F(
        `Link '${s}' does not exist on entity '${t}'. Available links: ${Nn(a)}`
      );
    }
    if (i != null) {
      if (Array.isArray(i)) {
        for (const a of i)
          if (!Yt(a))
            throw new F(
              `Invalid entity ID in link '${s}' for entity '${t}'. Expected a UUID or a lookup, but received: ${a}`
            );
      } else if (!Yt(i))
        throw new F(
          `Invalid UUID in link '${s}' for entity '${t}'. Expected a UUID, but received: ${i}`
        );
    }
  }
}, $i = {
  create: Ne,
  update: Ne,
  merge: Ne,
  link: Zt,
  unlink: Zt,
  delete: () => {
  }
}, ji = (t, e) => {
  if (!e) return;
  const [n, r, s, i] = t;
  if (!Array.isArray(s) && !ce(s))
    throw new F(
      `Invalid id for entity '${r}'. Expected a UUID, but received: ${s}`
    );
  if (typeof r != "string")
    throw new F(
      `Entity name must be a string, but received: ${typeof r}`
    );
  const o = $i[n];
  o && i !== void 0 && o(r, i, e);
}, Di = (t, e) => {
  const n = Array.isArray(t) ? t : [t];
  for (const r of n) {
    if (!r || typeof r != "object")
      throw new F(
        `Transaction chunk must be an object, but received: ${typeof r}`
      );
    if (!Array.isArray(r.__ops))
      throw new F(
        `Transaction chunk must have __ops array, but received: ${typeof r.__ops}`
      );
    for (const s of r.__ops) {
      if (!Array.isArray(s))
        throw new F(
          `Transaction operation must be an array, but received: ${typeof s}`
        );
      ji(s, e);
    }
  }
};
let qn = 0;
class Xt {
  constructor(e) {
    this.type = "ws", this.id = `${this.type}_${qn++}`, this.conn = new WebSocket(e), this.conn.onopen = (n) => {
      this.onopen && this.onopen({ target: this });
    }, this.conn.onmessage = (n) => {
      this.onmessage && this.onmessage({
        target: this,
        message: JSON.parse(n.data.toString())
      });
    }, this.conn.onclose = (n) => {
      this.onclose && this.onclose({ target: this });
    }, this.conn.onerror = (n) => {
      this.onerror && this.onerror({ target: this });
    };
  }
  close() {
    this.conn.close();
  }
  isOpen() {
    return this.conn.readyState === WebSocket.OPEN;
  }
  isConnecting() {
    return this.conn.readyState === WebSocket.CONNECTING;
  }
  send(e) {
    return this.conn.send(JSON.stringify(e));
  }
}
class Ri {
  constructor(e, n) {
    this.type = "sse", this.initParams = null, this.sendQueue = [], this.closeFired = !1, this.sseInitTimeout = null, this.id = `${this.type}_${qn++}`, this.url = n, this.ES = e, this.conn = new e(n), this.sseInitTimeout = setTimeout(() => {
      this.initParams || this.handleError();
    }, 1e4), this.conn.onmessage = (r) => {
      const s = JSON.parse(r.data);
      if (s.op === "sse-init") {
        this.initParams = {
          machineId: s["machine-id"],
          sessionId: s["session-id"],
          sseToken: s["sse-token"]
        }, this.onopen && this.onopen({ target: this }), clearTimeout(this.sseInitTimeout);
        return;
      }
      this.onmessage && this.onmessage({
        target: this,
        message: s
      });
    }, this.conn.onerror = (r) => {
      this.handleError();
    };
  }
  // Runs the onerror and closes the connection
  handleError() {
    try {
      this.onerror && this.onerror({ target: this });
    } finally {
      this.handleClose();
    }
  }
  handleClose() {
    this.conn.close(), this.onclose && !this.closeFired && (this.closeFired = !0, this.onclose({ target: this }));
  }
  async postMessages(e) {
    try {
      (await fetch(this.url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          machine_id: this.initParams.machineId,
          session_id: this.initParams.sessionId,
          sse_token: this.initParams.sseToken,
          messages: e
        })
      })).ok || this.handleError();
    } catch {
      this.handleError();
    }
  }
  async flushQueue() {
    if (this.sendPromise || !this.sendQueue.length) return;
    const e = this.sendQueue;
    this.sendQueue = [];
    const n = this.postMessages(e);
    this.sendPromise = n, n.then(() => {
      this.sendPromise = null, this.flushQueue();
    });
  }
  send(e) {
    if (!this.isOpen() || !this.initParams)
      throw this.isConnecting() ? new Error(
        "Failed to execute 'send' on 'EventSource': Still in CONNECTING state."
      ) : this.conn.readyState === this.ES.CLOSED ? new Error("EventSource is already in CLOSING or CLOSED state.") : new Error("EventSource is in invalid state.");
    this.sendQueue.push(e), this.flushQueue();
  }
  isOpen() {
    return this.conn.readyState === this.ES.OPEN && this.initParams !== null;
  }
  isConnecting() {
    return this.conn.readyState === this.ES.CONNECTING || this.conn.readyState === this.ES.OPEN && this.initParams === null;
  }
  close() {
    this.handleClose();
  }
}
const Q = {
  CONNECTING: "connecting",
  OPENED: "opened",
  AUTHENTICATED: "authenticated",
  CLOSED: "closed",
  ERRORED: "errored"
}, Ui = 3e4, xi = 3e4, Li = 200, Ni = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
}, Fe = "_instant_oauth_redirect", qe = "currentUser";
function Fi({
  transportType: t,
  appId: e,
  apiURI: n,
  wsURI: r,
  EventSourceImpl: s
}) {
  if (!s)
    return new Xt(`${r}?app_id=${e}`);
  switch (t) {
    case "ws":
      return new Xt(`${r}?app_id=${e}`);
    case "sse":
      return new Ri(
        s,
        `${n}/runtime/sse?app_id=${e}`
      );
    default:
      throw new Error("Unknown transport type " + t);
  }
}
function qi() {
  return typeof window < "u" || typeof chrome < "u";
}
const en = {
  "set-presence": !0,
  "set-presence-ok": !0,
  "refresh-presence": !0,
  "patch-presence": !0
};
function Vi(t, e) {
  var r;
  const n = typeof t == "string" ? JSON.parse(t) : t;
  for (const s in n) {
    const i = n[s];
    if ((r = i == null ? void 0 : i.result) != null && r.store) {
      const o = i.result.store;
      i.result.store = Mr({
        ...o,
        useDateObjects: e
      });
    }
  }
  return n;
}
function zi(t) {
  var n;
  const e = {};
  for (const r in t) {
    const s = t[r], i = { ...s };
    (n = s.result) != null && n.store && (i.result = {
      ...s.result,
      store: Er(s.result.store)
    }), e[r] = i;
  }
  return e;
}
function tn(t) {
  return [...t].sort((e, n) => {
    const [r, s] = e, [i, o] = n, a = s.order || 0, c = o.order || 0;
    return a == c ? r < i ? -1 : r > i ? 1 : 0 : a - c;
  });
}
class Wi {
  constructor(e, n = vn, r = In, s, i) {
    y(this, "attrs");
    y(this, "_isOnline", !0);
    y(this, "_isShutdown", !1);
    y(this, "status", Q.CONNECTING);
    /** @type {PersistedObject} */
    y(this, "querySubs");
    /** @type {PersistedObject} */
    y(this, "pendingMutations");
    /** @type {Record<string, Array<{ q: any, cb: (data: any) => any }>>} */
    y(this, "queryCbs", {});
    /** @type {Record<string, Array<{ q: any, eventId: string, dfd: Deferred }>>} */
    y(this, "queryOnceDfds", {});
    y(this, "authCbs", []);
    y(this, "attrsCbs", []);
    y(this, "mutationErrorCbs", []);
    y(this, "connectionStatusCbs", []);
    y(this, "config");
    y(this, "_persister");
    y(this, "mutationDeferredStore", /* @__PURE__ */ new Map());
    y(this, "_reconnectTimeoutId", null);
    y(this, "_reconnectTimeoutMs", 0);
    /** @type {Connection} */
    y(this, "_transport");
    /** @type {TransportType} */
    y(this, "_transportType", "ws");
    /** @type {EventSourceConstructor} */
    y(this, "_EventSource");
    /** @type {boolean | null} */
    y(this, "_wsOk", null);
    y(this, "_localIdPromises", {});
    y(this, "_errorMessage", null);
    /** @type {Promise<null | {error: {message: string}}>}**/
    y(this, "_oauthCallbackResponse", null);
    /** @type {null | import('./utils/linkIndex.ts').LinkIndex}} */
    y(this, "_linkIndex", null);
    /** @type BroadcastChannel | undefined */
    y(this, "_broadcastChannel");
    /** @type {Record<string, {isConnected: boolean; error: any}>} */
    y(this, "_rooms", {});
    /** @type {Record<string, boolean>} */
    y(this, "_roomsPendingLeave", {});
    y(this, "_presence", {});
    y(this, "_broadcastQueue", []);
    y(this, "_broadcastSubs", {});
    y(this, "_currentUserCached", { isLoading: !0, error: void 0, user: void 0 });
    y(this, "_beforeUnloadCbs", []);
    y(this, "_dataForQueryCache", {});
    /** @type {Logger} */
    y(this, "_log");
    y(this, "_pendingTxCleanupTimeout");
    y(this, "_pendingMutationCleanupThreshold");
    /**
     *  merge querySubs from storage and in memory. Has the following side
     *  effects:
     *  - We notify all queryCbs because results may been added during merge
     */
    y(this, "_onMergeQuerySubs", (e, n) => {
      const r = e || {}, s = { ...n };
      Object.entries(n).forEach(([o, a]) => {
        var f;
        const c = (f = r == null ? void 0 : r[o]) == null ? void 0 : f.result, u = a.result;
        c && !u && (s[o].result = c);
      }), Object.keys(r).filter((o) => !n[o]).sort((o, a) => {
        var f, l;
        const c = ((f = r[o]) == null ? void 0 : f.lastAccessed) || 0;
        return (((l = r[a]) == null ? void 0 : l.lastAccessed) || 0) - c;
      }).slice(0, this.queryCacheLimit).forEach((o) => {
        s[o] = r[o];
      }), this.querySubs.set((o) => s), this.loadedNotifyAll();
    });
    /**
     * merge pendingMutations from storage and in memory. Has a side effect of
     * sending mutations that were stored but not acked
     */
    y(this, "_onMergePendingMutations", (e, n) => {
      const r = new Map([...e.entries(), ...n.entries()]);
      this.pendingMutations.set((i) => r), this.loadedNotifyAll(), this._rewriteMutationsSorted(
        this.attrs,
        e
      ).forEach(([i, o]) => {
        !n.has(i) && !o["tx-id"] && this._sendMutation(i, o);
      });
    });
    // ---------------------------
    // Queries
    y(this, "getPreviousResult", (e) => {
      const n = U(e);
      return this.dataForQuery(n);
    });
    /** Re-run instaql and call all callbacks with new data */
    y(this, "notifyOne", (e) => {
      var i;
      const n = this.queryCbs[e] ?? [], r = (i = this._dataForQueryCache[e]) == null ? void 0 : i.data, s = this.dataForQuery(e);
      s && (Ee(s, r) || n.forEach((o) => o.cb(s)));
    });
    y(this, "notifyOneQueryOnce", (e) => {
      const n = this.queryOnceDfds[e] ?? [], r = this.dataForQuery(e);
      n.forEach((s) => {
        this._completeQueryOnce(s.q, e, s.dfd), s.dfd.resolve(r);
      });
    });
    y(this, "notifyQueryError", (e, n) => {
      (this.queryCbs[e] || []).forEach((s) => s.cb({ error: n }));
    });
    /** Applies transactions locally and sends transact message to server */
    y(this, "pushTx", (e) => {
      this.config.disableValidation || Di(e, this.config.schema);
      try {
        const n = fi(
          {
            attrs: this.optimisticAttrs(),
            schema: this.config.schema,
            stores: Object.values(this.querySubs.currentValue).map(
              (r) => {
                var s;
                return (s = r == null ? void 0 : r.result) == null ? void 0 : s.store;
              }
            ),
            useDateObjects: this.config.useDateObjects
          },
          e
        );
        return this.pushOps(n);
      } catch (n) {
        return this.pushOps([], n);
      }
    });
    /**
     * @param {*} txSteps
     * @param {*} [error]
     * @returns
     */
    y(this, "pushOps", (e, n) => {
      const r = k(), s = [...this.pendingMutations.currentValue.values()], i = Math.max(0, ...s.map((c) => c.order || 0)) + 1, o = {
        op: "transact",
        "tx-steps": e,
        created: Date.now(),
        error: n,
        order: i
      };
      this.pendingMutations.set((c) => (c.set(r, o), c));
      const a = new Kt();
      return this.mutationDeferredStore.set(r, a), this._sendMutation(r, o), this.notifyAll(), a.promise;
    });
    y(this, "_transportOnOpen", (e) => {
      const n = e.target;
      if (this._transport !== n) {
        this._log.info(
          "[socket][open]",
          n.id,
          "skip; this is no longer the current transport"
        );
        return;
      }
      this._log.info("[socket][open]", this._transport.id), this._setStatus(Q.OPENED), this.getCurrentUser().then((r) => {
        var s;
        this._trySend(k(), {
          op: "init",
          "app-id": this.config.appId,
          "refresh-token": (s = r.user) == null ? void 0 : s.refresh_token,
          versions: this.versions,
          // If an admin token is provided for an app, we will
          // skip all permission checks. This is an advanced feature,
          // to let users write internal tools
          // This option is not exposed in `Config`, as it's
          // not ready for prime time
          "__admin-token": this.config.__adminToken
        });
      }).catch((r) => {
        this._log.error("[socket][error]", n.id, r);
      });
    });
    y(this, "_transportOnMessage", (e) => {
      const n = e.target, r = e.message;
      if (this._transport !== n) {
        this._log.info(
          "[socket][message]",
          n.id,
          r,
          "skip; this is no longer the current transport"
        );
        return;
      }
      !this._wsOk && n.type === "ws" && (this._wsOk = !0), this._transportType = "ws", this._handleReceive(n.id, e.message);
    });
    y(this, "_transportOnError", (e) => {
      const n = e.target;
      if (this._transport !== n) {
        this._log.info(
          "[socket][error]",
          n.id,
          "skip; this is no longer the current transport"
        );
        return;
      }
      this._log.error("[socket][error]", n.id, e);
    });
    y(this, "_scheduleReconnect", () => {
      !this._wsOk && this._transportType !== "sse" && (this._transportType = "sse", this._reconnectTimeoutMs = 0), setTimeout(() => {
        if (this._reconnectTimeoutMs = Math.min(
          this._reconnectTimeoutMs + 1e3,
          1e4
        ), !this._isOnline) {
          this._log.info(
            "[socket][close]",
            this._transport.id,
            "we are offline, no need to start socket"
          );
          return;
        }
        this._startSocket();
      }, this._reconnectTimeoutMs);
    });
    y(this, "_transportOnClose", (e) => {
      const n = e.target;
      if (this._transport !== n) {
        this._log.info(
          "[socket][close]",
          n.id,
          "skip; this is no longer the current transport"
        );
        return;
      }
      this._setStatus(Q.CLOSED);
      for (const r of Object.values(this._rooms))
        r.isConnected = !1;
      if (this._isShutdown) {
        this._log.info(
          "[socket][close]",
          n.id,
          "Reactor has been shut down and will not reconnect"
        );
        return;
      }
      this._log.info(
        "[socket][close]",
        n.id,
        "schedule reconnect, ms =",
        this._reconnectTimeoutMs
      ), this._scheduleReconnect();
    });
    if (this._EventSource = i, this.config = { ...Ni, ...e }, this.queryCacheLimit = this.config.queryCacheLimit ?? 10, this._pendingTxCleanupTimeout = this.config.pendingTxCleanupTimeout ?? xi, this._pendingMutationCleanupThreshold = this.config.pendingMutationCleanupThreshold ?? Li, this._log = Ei(
      e.verbose || _t || $n
    ), this.versions = { ...s || {}, "@instantdb/core": Rn }, this.config.schema && (this._linkIndex = Bt(this.config.schema)), !!qi()) {
      if (!e.appId)
        throw new Error("Instant must be initialized with an appId.");
      if (!ce(e.appId))
        throw new Error(
          `Instant must be initialized with a valid appId. \`${e.appId}\` is not a valid uuid.`
        );
      typeof BroadcastChannel == "function" && (this._broadcastChannel = new BroadcastChannel("@instantdb"), this._broadcastChannel.addEventListener("message", async (o) => {
        var a;
        try {
          if (((a = o.data) == null ? void 0 : a.type) === "auth") {
            const c = await this.getCurrentUser();
            this.updateUser(c.user);
          }
        } catch (c) {
          this._log.error("[error] handle broadcast channel", c);
        }
      })), this._initStorage(n), this._oauthCallbackResponse = this._oauthLoginInit(), this.getCurrentUser(), r.getIsOnline().then((o) => {
        this._isOnline = o, this._startSocket(), r.listen((a) => {
          a !== this._isOnline && (this._log.info("[network] online =", a), this._isOnline = a, this._isOnline ? this._startSocket() : (this._log.info(
            "Changing status from",
            this.status,
            "to",
            Q.CLOSED
          ), this._setStatus(Q.CLOSED)));
        });
      }), typeof addEventListener < "u" && (this._beforeUnload = this._beforeUnload.bind(this), addEventListener("beforeunload", this._beforeUnload));
    }
  }
  updateSchema(e) {
    this.config = {
      ...this.config,
      schema: e,
      cardinalityInference: !!e
    }, this._linkIndex = e ? Bt(this.config.schema) : null;
  }
  _initStorage(e) {
    this._persister = new e(`instant_${this.config.appId}_5`), this.querySubs = new Qt(
      this._persister,
      "querySubs",
      {},
      this._onMergeQuerySubs,
      zi,
      (n) => Vi(n, this.config.useDateObjects)
    ), this.pendingMutations = new Qt(
      this._persister,
      "pendingMutations",
      /* @__PURE__ */ new Map(),
      this._onMergePendingMutations,
      (n) => [...n.entries()],
      (n) => new Map(typeof n == "string" ? JSON.parse(n) : n),
      100,
      // throttle
      100
      // idlecallback timeout, flush more often for mutations
    ), this._beforeUnloadCbs.push(() => {
      this.pendingMutations.flush(), this.querySubs.flush();
    });
  }
  _beforeUnload() {
    for (const e of this._beforeUnloadCbs)
      e();
  }
  /**
   * @param {'enqueued' | 'pending' | 'synced' | 'timeout' |  'error' } status
   * @param {string} eventId
   * @param {{message?: string, type?: string, status?: number, hint?: unknown}} [errorMsg]
   */
  _finishTransaction(e, n, r) {
    const s = this.mutationDeferredStore.get(n);
    this.mutationDeferredStore.delete(n);
    const i = e !== "error" && e !== "timeout";
    if (!s && !i && console.error("Mutation failed", { status: e, eventId: n, ...r }), !!s)
      if (i)
        s.resolve({ status: e, eventId: n });
      else if (r.type) {
        const { status: o, ...a } = r;
        s.reject(
          new je({
            // @ts-expect-error body.type is not constant typed
            body: a,
            status: o
          })
        );
      } else
        s.reject(new be(r.message, r.hint));
  }
  _setStatus(e, n) {
    this.status = e, this._errorMessage = n, this.notifyConnectionStatusSubs(e);
  }
  _flushEnqueuedRoomData(e) {
    var s, i;
    const n = (i = (s = this._presence[e]) == null ? void 0 : s.result) == null ? void 0 : i.user, r = this._broadcastQueue[e];
    if (this._broadcastQueue[e] = [], n && this._trySetPresence(e, n), r)
      for (const o of r) {
        const { topic: a, roomType: c, data: u } = o;
        this._tryBroadcast(e, c, a, u);
      }
  }
  _handleReceive(e, n) {
    var s, i, o, a, c, u;
    const r = !!this.config.schema && ("cardinalityInference" in this.config ? !!this.config.cardinalityInference : !0);
    switch (en[n.op] || this._log.info("[receive]", e, n.op, n), n.op) {
      case "init-ok":
        this._setStatus(Q.AUTHENTICATED), this._reconnectTimeoutMs = 0, this._setAttrs(n.attrs), this._flushPendingMessages(), this._sessionId = n["session-id"];
        for (const A of Object.keys(this._rooms)) {
          const Z = (i = (s = this._presence[A]) == null ? void 0 : s.result) == null ? void 0 : i.user;
          this._tryJoinRoom(A, Z);
        }
        break;
      case "add-query-exists":
        this.notifyOneQueryOnce(U(n.q));
        break;
      case "add-query-ok":
        const { q: f, result: l } = n, d = U(f), h = (a = (o = l == null ? void 0 : l[0]) == null ? void 0 : o.data) == null ? void 0 : a["page-info"], p = (u = (c = l == null ? void 0 : l[0]) == null ? void 0 : c.data) == null ? void 0 : u.aggregate, T = Gt(l), w = Je(
          this.attrs,
          T,
          r,
          this._linkIndex,
          this.config.useDateObjects
        );
        this.querySubs.set((A) => (A[d].result = {
          store: w,
          pageInfo: h,
          aggregate: p,
          processedTxId: n["processed-tx-id"]
        }, A)), this._cleanupPendingMutationsQueries(), this.notifyOne(d), this.notifyOneQueryOnce(d), this._cleanupPendingMutationsTimeout();
        break;
      case "refresh-ok":
        const { computations: b, attrs: _ } = n, S = n["processed-tx-id"];
        _ && this._setAttrs(_), this._cleanupPendingMutationsTimeout();
        const N = this._rewriteMutations(
          this.attrs,
          this.pendingMutations.currentValue,
          S
        );
        N !== this.pendingMutations.currentValue && this.pendingMutations.set(() => N);
        const m = tn(N.entries()), P = b.map((A) => {
          var At, Et, Mt, Ct;
          const Z = A["instaql-query"], W = A["instaql-result"], re = U(Z), se = Gt(W), J = Je(
            this.attrs,
            se,
            r,
            this._linkIndex,
            this.config.useDateObjects
          ), ge = this._applyOptimisticUpdates(
            J,
            m,
            S
          ), zn = (Et = (At = W == null ? void 0 : W[0]) == null ? void 0 : At.data) == null ? void 0 : Et["page-info"], Wn = (Ct = (Mt = W == null ? void 0 : W[0]) == null ? void 0 : Mt.data) == null ? void 0 : Ct.aggregate;
          return { hash: re, store: ge, pageInfo: zn, aggregate: Wn };
        });
        P.forEach(({ hash: A, store: Z, pageInfo: W, aggregate: re }) => {
          this.querySubs.set((se) => (se[A].result = { store: Z, pageInfo: W, aggregate: re, processedTxId: S }, se));
        }), this._cleanupPendingMutationsQueries(), P.forEach(({ hash: A }) => {
          this.notifyOne(A);
        });
        break;
      case "transact-ok": {
        const { "client-event-id": A, "tx-id": Z } = n, re = this._rewriteMutations(
          this.attrs,
          this.pendingMutations.currentValue
        ).get(A);
        if (!re)
          break;
        this.pendingMutations.set((J) => (J.set(A, {
          ...J.get(A),
          "tx-id": Z,
          confirmed: Date.now()
        }), J));
        const se = re["tx-steps"].filter(([J, ...ge]) => J === "add-attr").map(([J, ge]) => ge).concat(Object.values(this.attrs));
        this._setAttrs(se), this._finishTransaction("synced", A), this._cleanupPendingMutationsTimeout();
        break;
      }
      case "patch-presence": {
        const A = n["room-id"];
        this._patchPresencePeers(A, n.edits), this._notifyPresenceSubs(A);
        break;
      }
      case "refresh-presence": {
        const A = n["room-id"];
        this._setPresencePeers(A, n.data), this._notifyPresenceSubs(A);
        break;
      }
      case "server-broadcast":
        const R = n["room-id"], Y = n.topic;
        this._notifyBroadcastSubs(R, Y, n);
        break;
      case "join-room-ok":
        const H = n["room-id"], mt = this._rooms[H];
        if (!mt) {
          this._roomsPendingLeave[H] && (this._tryLeaveRoom(H), delete this._roomsPendingLeave[H]);
          break;
        }
        mt.isConnected = !0, this._notifyPresenceSubs(H), this._flushEnqueuedRoomData(H);
        break;
      case "join-room-error":
        const St = n["room-id"], Ot = this._rooms[St];
        Ot && (Ot.error = n.error), this._notifyPresenceSubs(St);
        break;
      case "error":
        this._handleReceiveError(n);
        break;
    }
  }
  /**
   * @param {'timeout' | 'error'} status
   * @param {string} eventId
   * @param {{message: string, type?: string, status?: number, hint?: unknown}} errorMsg
   */
  _handleMutationError(e, n, r) {
    const s = this.pendingMutations.currentValue.get(n);
    if (s && (e !== "timeout" || !s["tx-id"])) {
      this.pendingMutations.set((o) => (o.delete(n), o));
      const i = {
        message: r.message,
        hint: r.hint
      };
      this.notifyAll(), this.notifyAttrsSubs(), this.notifyMutationErrorSubs(i), this._finishTransaction(e, n, r);
    }
  }
  _handleReceiveError(e) {
    var a, c, u, f, l;
    console.log("error", e);
    const n = e["client-event-id"], r = this.pendingMutations.currentValue.get(n), s = {
      message: e.message || "Uh-oh, something went wrong. Ping Joe & Stopa."
    };
    if (e.hint && (s.hint = e.hint), r) {
      this._handleMutationError("error", n, e);
      return;
    }
    if ((a = e["original-event"]) != null && a.hasOwnProperty("q") && ((c = e["original-event"]) == null ? void 0 : c.op) === "add-query") {
      const d = (u = e["original-event"]) == null ? void 0 : u.q, h = U(d);
      this.notifyQueryError(U(d), s), this.notifyQueryOnceError(d, h, n, s);
      return;
    }
    if (((f = e["original-event"]) == null ? void 0 : f.op) === "init") {
      if (e.type === "record-not-found" && ((l = e.hint) == null ? void 0 : l["record-type"]) === "app-user") {
        this.changeCurrentUser(null);
        return;
      }
      this._setStatus(Q.ERRORED, s), this.notifyAll();
      return;
    }
    const o = { ...e };
    delete o.message, delete o.hint, console.error(e.message, o), e.hint && console.error(
      `This error comes with some debugging information. Here it is: 
`,
      e.hint
    );
  }
  notifyQueryOnceError(e, n, r, s) {
    var o;
    const i = (o = this.queryOnceDfds[n]) == null ? void 0 : o.find((a) => a.eventId === r);
    i && (i.dfd.reject(s), this._completeQueryOnce(e, n, i.dfd));
  }
  _setAttrs(e) {
    this.attrs = e.reduce((n, r) => (n[r.id] = r, n), {}), this.notifyAttrsSubs();
  }
  _startQuerySub(e, n) {
    const r = k();
    return this.querySubs.set((s) => (s[n] = s[n] || { q: e, result: null, eventId: r }, s[n].lastAccessed = Date.now(), s)), this._trySendAuthed(r, { op: "add-query", q: e }), r;
  }
  /**
   *  When a user subscribes to a query the following side effects occur:
   *
   *  - We update querySubs to include the new query
   *  - We update queryCbs to include the new cb
   *  - If we already have a result for the query we call cb immediately
   *  - We send the server an `add-query` message
   *
   *  Returns an unsubscribe function
   */
  subscribeQuery(e, n, r) {
    this.config.disableValidation || Jt(e, this.config.schema), r && "ruleParams" in r && (e = { $$ruleParams: r.ruleParams, ...e });
    const s = U(e), i = this.getPreviousResult(e);
    return i && n(i), this.queryCbs[s] = this.queryCbs[s] ?? [], this.queryCbs[s].push({ q: e, cb: n }), this._startQuerySub(e, s), () => {
      this._unsubQuery(e, s, n);
    };
  }
  queryOnce(e, n) {
    this.config.disableValidation || Jt(e, this.config.schema), n && "ruleParams" in n && (e = { $$ruleParams: n.ruleParams, ...e });
    const r = new Kt();
    if (!this._isOnline)
      return r.reject(
        new Error("We can't run `queryOnce`, because the device is offline.")
      ), r.promise;
    if (!this.querySubs)
      return r.reject(
        new Error(
          "We can't run `queryOnce` on the backend. Use adminAPI.query instead: https://www.instantdb.com/docs/backend#query"
        )
      ), r.promise;
    const s = U(e), i = this._startQuerySub(e, s);
    return this.queryOnceDfds[s] = this.queryOnceDfds[s] ?? [], this.queryOnceDfds[s].push({ q: e, dfd: r, eventId: i }), setTimeout(
      () => r.reject(new Error("Query timed out")),
      Ui
    ), r.promise;
  }
  _completeQueryOnce(e, n, r) {
    this.queryOnceDfds[n] && (this.queryOnceDfds[n] = this.queryOnceDfds[n].filter(
      (s) => s.dfd !== r
    ), this._cleanupQuery(e, n));
  }
  _unsubQuery(e, n, r) {
    this.queryCbs[n] && (this.queryCbs[n] = this.queryCbs[n].filter((s) => s.cb !== r), this._cleanupQuery(e, n));
  }
  _cleanupQuery(e, n) {
    var s, i;
    (s = this.queryCbs[n]) != null && s.length || (i = this.queryOnceDfds[n]) != null && i.length || (delete this.queryCbs[n], delete this.queryOnceDfds[n], this._trySendAuthed(k(), { op: "remove-query", q: e }));
  }
  // When we `pushTx`, it's possible that we don't yet have `this.attrs`
  // This means that `tx-steps` in `pendingMutations` will include `add-attr`
  // commands for attrs that already exist.
  //
  // This will also affect `add-triple` and `retract-triple` which
  // reference attr-ids that do not match the server.
  //
  // We fix this by rewriting `tx-steps` in each `pendingMutation`.
  // We remove `add-attr` commands for attrs that already exist.
  // We update `add-triple` and `retract-triple` commands to use the
  // server attr-ids.
  _rewriteMutations(e, n, r) {
    if (!e) return n;
    const s = (f) => {
      const [l, d, h] = f["forward-identity"];
      return C(e, d, h);
    }, i = (f) => {
      const [l, d, h] = f["forward-identity"];
      return ee(e, d, h);
    }, o = { attrIdMap: {}, refSwapAttrIds: /* @__PURE__ */ new Set() };
    let a = !1;
    const c = (f, l) => {
      const d = [];
      for (const h of f) {
        const [p] = h;
        if (p === "add-attr") {
          const [w, b] = h, _ = s(b);
          if (_ && b.id !== _.id) {
            o.attrIdMap[b.id] = _.id, a = !0;
            continue;
          }
          if (b["value-type"] === "ref") {
            const S = i(b);
            if (S) {
              o.attrIdMap[b.id] = S.id, o.refSwapAttrIds.add(b.id), a = !0;
              continue;
            }
          }
        }
        if (r && l && r >= l && p === "add-attr" || p === "update-attr" || p === "delete-attr") {
          a = !0;
          continue;
        }
        const T = a ? Fs(o, h) : h;
        d.push(T);
      }
      return a ? d : f;
    }, u = /* @__PURE__ */ new Map();
    for (const [f, l] of n.entries())
      u.set(f, {
        ...l,
        "tx-steps": c(l["tx-steps"], l["tx-id"])
      });
    return a ? u : n;
  }
  _rewriteMutationsSorted(e, n) {
    return tn(this._rewriteMutations(e, n).entries());
  }
  // ---------------------------
  // Transact
  optimisticAttrs() {
    var o;
    const e = [
      ...this.pendingMutations.currentValue.values()
    ].flatMap((a) => a["tx-steps"]), n = new Set(
      e.filter(([a, c]) => a === "delete-attr").map(([a, c]) => c)
    ), r = [];
    for (const [a, c] of e)
      if (a === "add-attr")
        r.push(c);
      else if (a === "update-attr" && c.id && ((o = this.attrs) != null && o[c.id])) {
        const u = { ...this.attrs[c.id], ...c };
        r.push(u);
      }
    const s = [
      ...Object.values(this.attrs || {}),
      ...r
    ].filter((a) => !n.has(a.id));
    return Object.fromEntries(
      s.map((a) => [a.id, a])
    );
  }
  /** Runs instaql on a query and a store */
  dataForQuery(e) {
    const n = this._errorMessage;
    if (n)
      return { error: n };
    if (!this.querySubs || !this.pendingMutations) return;
    const r = this.querySubs.version(), s = this.querySubs.currentValue, i = this.pendingMutations.version(), o = this.pendingMutations.currentValue, { q: a, result: c } = s[e] || {};
    if (!c) return;
    const u = this._dataForQueryCache[e];
    if (u && r === u.querySubVersion && i === u.pendingMutationsVersion)
      return u.data;
    const { store: f, pageInfo: l, aggregate: d, processedTxId: h } = c, p = this._rewriteMutationsSorted(
      f.attrs,
      o
    ), T = this._applyOptimisticUpdates(
      f,
      p,
      h
    ), w = Rs({ store: T, pageInfo: l, aggregate: d }, a);
    return this._dataForQueryCache[e] = {
      querySubVersion: r,
      pendingMutationsVersion: i,
      data: w
    }, w;
  }
  _applyOptimisticUpdates(e, n, r) {
    for (const [s, i] of n)
      (!i["tx-id"] || r && i["tx-id"] > r) && (e = Fr(e, i["tx-steps"]));
    return e;
  }
  /** Re-compute all subscriptions */
  notifyAll() {
    Object.keys(this.queryCbs).forEach((e) => {
      this.notifyOne(e);
    });
  }
  loadedNotifyAll() {
    this.pendingMutations.isLoading() || this.querySubs.isLoading() || this.notifyAll();
  }
  shutdown() {
    var e;
    this._log.info("[shutdown]", this.config.appId), this._isShutdown = !0, (e = this._transport) == null || e.close();
  }
  /**
   * Sends mutation to server and schedules a timeout to cancel it if
   * we don't hear back in time.
   * Note: If we're offline we don't schedule a timeout, we'll schedule it
   * later once we're back online and send the mutation again
   *
   */
  _sendMutation(e, n) {
    if (n.error) {
      this._handleMutationError("error", e, {
        message: n.error.message
      });
      return;
    }
    if (this.status !== Q.AUTHENTICATED) {
      this._finishTransaction("enqueued", e);
      return;
    }
    const r = Math.max(
      5e3,
      this.pendingMutations.currentValue.size * 5e3
    );
    this._isOnline ? (this._trySend(e, n), setTimeout(() => {
      this._isOnline && this._handleMutationError("timeout", e, {
        message: "transaction timed out"
      });
    }, r)) : this._finishTransaction("enqueued", e);
  }
  // ---------------------------
  // Websocket
  /** Send messages we accumulated while we were connecting */
  _flushPendingMessages() {
    Object.keys(this.queryCbs).map((s) => this.querySubs.currentValue[s]).filter((s) => s).forEach(({ eventId: s, q: i }) => {
      this._trySendAuthed(s, { op: "add-query", q: i });
    }), Object.values(this.queryOnceDfds).flat().forEach(({ eventId: s, q: i }) => {
      this._trySendAuthed(s, { op: "add-query", q: i });
    }), this._rewriteMutationsSorted(
      this.attrs,
      this.pendingMutations.currentValue
    ).forEach(([s, i]) => {
      i["tx-id"] || this._sendMutation(s, i);
    });
  }
  /**
   * Clean up pendingMutations that all queries have seen
   */
  _cleanupPendingMutationsQueries() {
    let e = Number.MAX_SAFE_INTEGER;
    for (const { result: n } of Object.values(this.querySubs.currentValue))
      n != null && n.processedTxId && (e = Math.min(e, n == null ? void 0 : n.processedTxId));
    this.pendingMutations.set((n) => {
      for (const [r, s] of Array.from(n.entries()))
        s["tx-id"] && s["tx-id"] <= e && n.delete(r);
      return n;
    });
  }
  /**
   * After mutations is confirmed by server, we give each query 30 sec
   * to update its results. If that doesn't happen, we assume query is
   * unaffected by this mutation and it’s safe to delete it from local queue
   */
  _cleanupPendingMutationsTimeout() {
    if (this.pendingMutations.currentValue.size < this._pendingMutationCleanupThreshold)
      return;
    const e = Date.now();
    this.pendingMutations.set((n) => {
      for (const [r, s] of Array.from(n.entries()))
        s.confirmed && s.confirmed + this._pendingTxCleanupTimeout < e && n.delete(r);
      return n;
    });
  }
  _trySendAuthed(...e) {
    this.status === Q.AUTHENTICATED && this._trySend(...e);
  }
  _trySend(e, n, r) {
    this._transport.isOpen() && (en[n.op] || this._log.info("[send]", this._transport.id, n.op, n), this._transport.send({ "client-event-id": e, ...n }));
  }
  _startSocket() {
    if (this._wsOk = null, this._isShutdown) {
      this._log.info(
        "[socket][start]",
        this.config.appId,
        "Reactor has been shut down and will not start a new socket"
      );
      return;
    }
    if (this._transport && this._transport.isConnecting()) {
      this._log.info(
        "[socket][start]",
        this._transport.id,
        "maintained as current transport, we were still in a connecting state"
      );
      return;
    }
    const e = this._transport;
    this._transport = Fi({
      transportType: this._transportType,
      appId: this.config.appId,
      apiURI: this.config.apiURI,
      wsURI: this.config.websocketURI,
      EventSourceImpl: this._EventSource
    }), this._transport.onopen = this._transportOnOpen, this._transport.onmessage = this._transportOnMessage, this._transport.onclose = this._transportOnClose, this._transport.onerror = this._transportOnError, this._log.info("[socket][start]", this._transport.id), e != null && e.isOpen() && (this._log.info(
      "[socket][start]",
      this._transport.id,
      "close previous transport id = ",
      e.id
    ), e.close());
  }
  /**
   * Given a key, returns a stable local id, unique to this device and app.
   *
   * This can be useful if you want to create guest ids for example.
   *
   * Note: If the user deletes their local storage, this id will change.
   *
   * We use this._localIdPromises to ensure that we only generate a local
   * id once, even if multiple callers call this function concurrently.
   */
  async getLocalId(e) {
    const n = `localToken_${e}`, r = await this._persister.getItem(n);
    if (r) return r;
    if (this._localIdPromises[n])
      return this._localIdPromises[n];
    const s = k();
    return this._localIdPromises[n] = this._persister.setItem(n, s).then(() => s), this._localIdPromises[n];
  }
  // ----
  // Auth
  _replaceUrlAfterOAuth() {
    if (typeof URL > "u")
      return;
    const e = new URL(window.location.href);
    if (e.searchParams.get(Fe)) {
      const n = e.toString();
      e.searchParams.delete(Fe), e.searchParams.delete("code"), e.searchParams.delete("error");
      const r = e.pathname + (e.searchParams.size ? "?" + e.searchParams : "") + e.hash;
      if (history.replaceState(history.state, "", r), // @ts-ignore (waiting for ts support)
      typeof navigation == "object" && // @ts-ignore (waiting for ts support)
      typeof navigation.addEventListener == "function" && // @ts-ignore (waiting for ts support)
      typeof navigation.removeEventListener == "function") {
        let s = !1;
        const i = (o) => {
          var a;
          s || (s = !0, navigation.removeEventListener("navigate", i), !o.userInitiated && o.navigationType === "replace" && ((a = o.destination) == null ? void 0 : a.url) === n && history.replaceState(history.state, "", r));
        };
        navigation.addEventListener("navigate", i);
      }
    }
  }
  /**
   *
   * @returns Promise<null | {error: {message: string}}>
   */
  async _oauthLoginInit() {
    var s, i, o, a;
    if (typeof window > "u" || typeof window.location > "u" || typeof URLSearchParams > "u")
      return null;
    const e = new URLSearchParams(window.location.search);
    if (!e.get(Fe))
      return null;
    const n = e.get("error");
    if (n)
      return this._replaceUrlAfterOAuth(), { error: { message: n } };
    const r = e.get("code");
    if (!r)
      return null;
    this._replaceUrlAfterOAuth();
    try {
      const c = await this._getCurrentUser(), u = (c == null ? void 0 : c.type) === "guest", { user: f } = await zt({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        code: r,
        refreshToken: u ? c.refresh_token : void 0
      });
      return this.setCurrentUser(f), null;
    } catch (c) {
      return ((s = c == null ? void 0 : c.body) == null ? void 0 : s.type) === "record-not-found" && ((o = (i = c == null ? void 0 : c.body) == null ? void 0 : i.hint) == null ? void 0 : o["record-type"]) === "app-oauth-code" && await this._hasCurrentUser() ? null : { error: { message: ((a = c == null ? void 0 : c.body) == null ? void 0 : a.message) || "Error logging in." } };
    }
  }
  async _waitForOAuthCallbackResponse() {
    return await this._oauthCallbackResponse;
  }
  __subscribeMutationErrors(e) {
    return this.mutationErrorCbs.push(e), () => {
      this.mutationErrorCbs = this.mutationErrorCbs.filter((n) => n !== e);
    };
  }
  subscribeAuth(e) {
    this.authCbs.push(e);
    const n = this._currentUserCached;
    n.isLoading || e(this._currentUserCached);
    let r = !1;
    return this.getCurrentUser().then((s) => {
      r || Ee(s, n) || e(s);
    }), () => {
      r = !0, this.authCbs = this.authCbs.filter((s) => s !== e);
    };
  }
  async getAuth() {
    const { user: e, error: n } = await this.getCurrentUser();
    if (n)
      throw new be("Could not get current user: " + n.message);
    return e;
  }
  subscribeConnectionStatus(e) {
    return this.connectionStatusCbs.push(e), () => {
      this.connectionStatusCbs = this.connectionStatusCbs.filter(
        (n) => n !== e
      );
    };
  }
  subscribeAttrs(e) {
    return this.attrsCbs.push(e), this.attrs && e(this.attrs), () => {
      this.attrsCbs = this.attrsCbs.filter((n) => n !== e);
    };
  }
  notifyAuthSubs(e) {
    this.authCbs.forEach((n) => n(e));
  }
  notifyMutationErrorSubs(e) {
    this.mutationErrorCbs.forEach((n) => n(e));
  }
  notifyAttrsSubs() {
    if (!this.attrs) return;
    const e = this.optimisticAttrs();
    this.attrsCbs.forEach((n) => n(e));
  }
  notifyConnectionStatusSubs(e) {
    this.connectionStatusCbs.forEach((n) => n(e));
  }
  async setCurrentUser(e) {
    await this._persister.setItem(qe, e);
  }
  getCurrentUserCached() {
    return this._currentUserCached;
  }
  async _getCurrentUser() {
    const e = await this._persister.getItem(qe);
    return typeof e == "string" ? JSON.parse(e) : e;
  }
  async getCurrentUser() {
    const e = await this._waitForOAuthCallbackResponse();
    if (e != null && e.error) {
      const s = { error: e.error, user: void 0 };
      return this._currentUserCached = { isLoading: !1, ...s }, s;
    }
    const r = { user: await this._getCurrentUser(), error: void 0 };
    return this._currentUserCached = {
      isLoading: !1,
      ...r
    }, r;
  }
  async _hasCurrentUser() {
    const e = await this._persister.getItem(qe);
    return JSON.parse(e) != null;
  }
  async changeCurrentUser(e) {
    var r;
    const { user: n } = await this.getCurrentUser();
    if (!Ee(n, e)) {
      await this.setCurrentUser(e), this.updateUser(e);
      try {
        (r = this._broadcastChannel) == null || r.postMessage({ type: "auth" });
      } catch (s) {
        console.error("Error posting message to broadcast channel", s);
      }
    }
  }
  updateUser(e) {
    const n = { error: void 0, user: e };
    this._currentUserCached = { isLoading: !1, ...n }, this._dataForQueryCache = {}, this.querySubs.set((r) => (Object.keys(r).forEach((s) => {
      delete r[s].result;
    }), r)), this._reconnectTimeoutMs = 0, this._transport.close(), this._oauthCallbackResponse = null, this.notifyAuthSubs(n);
  }
  sendMagicCode({ email: e }) {
    return li({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: e
    });
  }
  async signInWithMagicCode({ email: e, code: n }) {
    var o;
    const r = await this.getCurrentUser(), s = ((o = r == null ? void 0 : r.user) == null ? void 0 : o.type) === "guest", i = await di({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: e,
      code: n,
      refreshToken: s ? r.user.refresh_token : void 0
    });
    return await this.changeCurrentUser(i.user), i;
  }
  async signInWithCustomToken(e) {
    const n = await hi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      refreshToken: e
    });
    return await this.changeCurrentUser(n.user), n;
  }
  async signInAsGuest() {
    const e = await pi({
      apiURI: this.config.apiURI,
      appId: this.config.appId
    });
    return await this.changeCurrentUser(e.user), e;
  }
  potentiallyInvalidateToken(e, n) {
    var i;
    const r = (i = e == null ? void 0 : e.user) == null ? void 0 : i.refresh_token;
    if (!r)
      return;
    if (n.invalidateToken === !1) {
      this._log.info("[auth-invalidate] skipped invalidateToken");
      return;
    }
    bi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      refreshToken: r
    }).then(() => {
      this._log.info("[auth-invalidate] completed invalidateToken");
    }).catch((o) => {
    });
  }
  async signOut(e) {
    const n = await this.getCurrentUser();
    this.potentiallyInvalidateToken(n, e), await this.changeCurrentUser(null);
  }
  /**
   * Creates an OAuth authorization URL.
   *
   * @param {Object} params - The parameters to create the authorization URL.
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.redirectURL - The URL to redirect users to after authorization.
   * @returns {string} The created authorization URL.
   */
  createAuthorizationURL({ clientName: e, redirectURL: n }) {
    const { apiURI: r, appId: s } = this.config;
    return `${r}/runtime/oauth/start?app_id=${s}&client_name=${e}&redirect_uri=${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.code - The code received from the OAuth service.
   * @param {string} [params.codeVerifier] - The code verifier used to generate the code challenge.
   */
  async exchangeCodeForToken({ code: e, codeVerifier: n }) {
    var o;
    const r = await this.getCurrentUser(), s = ((o = r == null ? void 0 : r.user) == null ? void 0 : o.type) === "guest", i = await zt({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      code: e,
      codeVerifier: n,
      refreshToken: s ? r.user.refresh_token : void 0
    });
    return await this.changeCurrentUser(i.user), i;
  }
  issuerURI() {
    const { apiURI: e, appId: n } = this.config;
    return `${e}/runtime/${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.idToken - The id_token from the external service
   * @param {string | null | undefined} [params.nonce] - The nonce used when requesting the id_token from the external service
   */
  async signInWithIdToken({ idToken: e, clientName: n, nonce: r }) {
    var a;
    const s = await this.getCurrentUser(), i = (a = s == null ? void 0 : s.user) == null ? void 0 : a.refresh_token, o = await yi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      idToken: e,
      clientName: n,
      nonce: r,
      refreshToken: i
    });
    return await this.changeCurrentUser(o.user), o;
  }
  // --------
  // Rooms
  /**
   * @param {string} roomId
   * @param {any | null | undefined} [initialData] -- initial presence data to send when joining the room
   * @returns () => void
   */
  joinRoom(e, n) {
    return this._rooms[e] || (this._rooms[e] = {
      isConnected: !1,
      error: void 0
    }), this._presence[e] = this._presence[e] || {}, n && (this._presence[e].result = this._presence[e].result || {}, this._presence[e].result.user = n, this._notifyPresenceSubs(e)), this._tryJoinRoom(e, n), () => {
      this._cleanupRoom(e);
    };
  }
  _cleanupRoom(e) {
    var n, r, s;
    if (!((r = (n = this._presence[e]) == null ? void 0 : n.handlers) != null && r.length) && !Object.keys(this._broadcastSubs[e] ?? {}).length) {
      const i = (s = this._rooms[e]) == null ? void 0 : s.isConnected;
      delete this._rooms[e], delete this._presence[e], delete this._broadcastSubs[e], i ? this._tryLeaveRoom(e) : this._roomsPendingLeave[e] = !0;
    }
  }
  // --------
  // Presence
  // TODO: look into typing again
  getPresence(e, n, r = {}) {
    const s = this._rooms[n], i = this._presence[n];
    return !s || !i || !i.result ? null : {
      ...Si(i.result, r, this._sessionId),
      isLoading: !s.isConnected,
      error: s.error
    };
  }
  // TODO: look into typing again
  publishPresence(e, n, r) {
    const s = this._rooms[n], i = this._presence[n];
    if (!s || !i)
      return;
    i.result = i.result || {};
    const o = {
      ...i.result.user,
      ...r
    };
    i.result.user = o, s.isConnected && (this._trySetPresence(n, o), this._notifyPresenceSubs(n));
  }
  _trySetPresence(e, n) {
    this._trySendAuthed(k(), {
      op: "set-presence",
      "room-id": e,
      data: n
    });
  }
  _tryJoinRoom(e, n) {
    this._trySendAuthed(k(), { op: "join-room", "room-id": e, data: n }), delete this._roomsPendingLeave[e];
  }
  _tryLeaveRoom(e) {
    this._trySendAuthed(k(), { op: "leave-room", "room-id": e });
  }
  // TODO: look into typing again
  subscribePresence(e, n, r, s) {
    const i = this.joinRoom(n, r.initialData), o = { ...r, roomId: n, cb: s, prev: null };
    return this._presence[n] = this._presence[n] || {}, this._presence[n].handlers = this._presence[n].handlers || [], this._presence[n].handlers.push(o), this._notifyPresenceSub(n, o), () => {
      var a, c;
      this._presence[n].handlers = ((c = (a = this._presence[n]) == null ? void 0 : a.handlers) == null ? void 0 : c.filter((u) => u !== o)) ?? [], i();
    };
  }
  _notifyPresenceSubs(e) {
    var n, r;
    (r = (n = this._presence[e]) == null ? void 0 : n.handlers) == null || r.forEach((s) => {
      this._notifyPresenceSub(e, s);
    });
  }
  _notifyPresenceSub(e, n) {
    const r = this.getPresence("", e, n);
    r && (n.prev && !Oi(r, n.prev) || (n.prev = r, n.cb(r)));
  }
  _patchPresencePeers(e, n) {
    var o, a, c;
    const r = ((a = (o = this._presence[e]) == null ? void 0 : o.result) == null ? void 0 : a.peers) || {};
    let s = Object.fromEntries(
      Object.entries(r).map(([u, f]) => [u, { data: f }])
    );
    (c = this._presence[e]) == null || c.result;
    const i = Be(s, (u) => {
      for (let [f, l, d] of n)
        switch (l) {
          case "+":
            or(u, f, d);
            break;
          case "r":
            jt(u, f, d);
            break;
          case "-":
            hn(u, f);
            break;
        }
      delete u[this._sessionId];
    });
    this._setPresencePeers(e, i);
  }
  _setPresencePeers(e, n) {
    const r = { ...n };
    delete r[this._sessionId];
    const s = Object.fromEntries(
      Object.entries(r).map(([i, o]) => [i, o.data])
    );
    this._presence = Be(this._presence, (i) => {
      jt(i, [e, "result", "peers"], s);
    });
  }
  // --------
  // Broadcast
  publishTopic({ roomType: e, roomId: n, topic: r, data: s }) {
    const i = this._rooms[n];
    if (i) {
      if (!i.isConnected) {
        this._broadcastQueue[n] = this._broadcastQueue[n] ?? [], this._broadcastQueue[n].push({ topic: r, roomType: e, data: s });
        return;
      }
      this._tryBroadcast(n, e, r, s);
    }
  }
  _tryBroadcast(e, n, r, s) {
    this._trySendAuthed(k(), {
      op: "client-broadcast",
      "room-id": e,
      roomType: n,
      topic: r,
      data: s
    });
  }
  subscribeTopic(e, n, r) {
    const s = this.joinRoom(e);
    return this._broadcastSubs[e] = this._broadcastSubs[e] || {}, this._broadcastSubs[e][n] = this._broadcastSubs[e][n] || [], this._broadcastSubs[e][n].push(r), this._presence[e] = this._presence[e] || {}, () => {
      this._broadcastSubs[e][n] = this._broadcastSubs[e][n].filter((i) => i !== r), this._broadcastSubs[e][n].length || delete this._broadcastSubs[e][n], s();
    };
  }
  _notifyBroadcastSubs(e, n, r) {
    var s, i, o;
    (o = (i = (s = this._broadcastSubs) == null ? void 0 : s[e]) == null ? void 0 : i[n]) == null || o.forEach((a) => {
      var f, l, d, h, p, T;
      const c = (f = r.data) == null ? void 0 : f.data, u = r.data["peer-id"] === this._sessionId ? (d = (l = this._presence[e]) == null ? void 0 : l.result) == null ? void 0 : d.user : (T = (p = (h = this._presence[e]) == null ? void 0 : h.result) == null ? void 0 : p.peers) == null ? void 0 : T[r.data["peer-id"]];
      return a(c, u);
    });
  }
  // --------
  // Storage
  async uploadFile(e, n, r) {
    var o;
    const s = await this.getCurrentUser(), i = (o = s == null ? void 0 : s.user) == null ? void 0 : o.refresh_token;
    return Ti({
      ...r,
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      path: e,
      file: n,
      refreshToken: i
    });
  }
  async deleteFile(e) {
    var i;
    const n = await this.getCurrentUser(), r = (i = n == null ? void 0 : n.user) == null ? void 0 : i.refresh_token;
    return await wi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      path: e,
      refreshToken: r
    });
  }
  // Deprecated Storage API (Jan 2025)
  // ---------------------------------
  async upload(e, n) {
    var c;
    const r = await this.getCurrentUser(), s = (c = r == null ? void 0 : r.user) == null ? void 0 : c.refresh_token, i = e || n.name, o = await gi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      fileName: i,
      refreshToken: s
    });
    return await _i(o, n);
  }
  async getDownloadUrl(e) {
    var i;
    const n = await this.getCurrentUser(), r = (i = n == null ? void 0 : n.user) == null ? void 0 : i.refresh_token;
    return await mi({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      path: e,
      refreshToken: r
    });
  }
}
function Ki(t, e) {
  return new Re(
    Vn(t, e),
    // (XXX): LinksDef<any> stems from TypeScript’s inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    e,
    void 0
  );
}
function Qi(t) {
  return new De(t, {});
}
function Gi() {
  return new I("string", !0, !1);
}
function Bi() {
  return new I("number", !0, !1);
}
function Hi() {
  return new I("boolean", !0, !1);
}
function Ji() {
  return new I("date", !0, !1);
}
function Yi() {
  return new I("json", !0, !1);
}
function Zi() {
  return new I("json", !0, !1);
}
function Vn(t, e) {
  var s, i, o, a;
  const n = { fwd: {}, rev: {} };
  for (const c of Object.values(e))
    (s = n.fwd)[i = c.forward.on] || (s[i] = {}), (o = n.rev)[a = c.reverse.on] || (o[a] = {}), n.fwd[c.forward.on][c.forward.label] = {
      entityName: c.reverse.on,
      cardinality: c.forward.has
    }, n.rev[c.reverse.on][c.reverse.label] = {
      entityName: c.forward.on,
      cardinality: c.reverse.has
    };
  return Object.fromEntries(
    Object.entries(t).map(([c, u]) => [
      c,
      new De(u.attrs, {
        ...n.fwd[c],
        ...n.rev[c]
      })
    ])
  );
}
function Xi({
  entities: t,
  links: e,
  rooms: n
}) {
  const r = e ?? {}, s = n ?? {};
  return new Re(
    Vn(t, r),
    // (XXX): LinksDef<any> stems from TypeScript's inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    r,
    s
  );
}
const _o = {
  // constructs
  graph: Ki,
  schema: Xi,
  entity: Qi,
  // value types
  string: Gi,
  number: Bi,
  boolean: Hi,
  date: Ji,
  json: Yi,
  any: Zi
};
let Se;
function eo(t, e) {
  Se == null || Se.dispose();
  const n = oo(e), r = ro(e, a), s = no(to(t));
  function i(f) {
    var l;
    f.source === s.element.contentWindow && ((l = f.data) == null ? void 0 : l.type) === "close" && n.isVisible() && a();
  }
  function o(f) {
    const l = f.shiftKey && f.ctrlKey && f.key === "0", d = f.key === "Escape" || f.key === "Esc";
    (l || d && n.isVisible()) && a();
  }
  function a() {
    n.isVisible() ? n.element.style.display = "none" : (n.element.style.display = "block", n.element.contains(s.element) || n.element.appendChild(s.element));
  }
  function c() {
    n.element.remove(), r.element.remove(), removeEventListener("keydown", o), removeEventListener("message", i);
  }
  function u() {
    document.body.appendChild(n.element), document.body.appendChild(r.element), addEventListener("keydown", o), addEventListener("message", i), Se = {
      dispose: c
    };
  }
  return u();
}
function to(t) {
  return `${_t || jn ? "http://localhost:3000" : "https://instantdb.com"}/_devtool?appId=${t}`;
}
function no(t) {
  const e = document.createElement("iframe");
  return e.src = t, e.className = "instant-devtool-iframe", Object.assign(e.style, {
    width: "100%",
    height: "100%",
    backgroundColor: "white",
    border: "none"
  }), { element: e };
}
function ro(t, e) {
  const n = `
    <svg width="32" height="32" viewBox="0 0 512 512" fill="none" xmlns="http://www.w3.org/2000/svg">
      <rect width="512" height="512" fill="black"/>
      <rect x="97.0973" y="91.3297" width="140" height="330" fill="white"/>
    </svg>
  `, r = document.createElement("button");
  return r.innerHTML = n, r.className = "instant-devtool-toggler", Object.assign(r.style, {
    // pos
    position: "fixed",
    ...so(t.position),
    height: "32px",
    width: "32px",
    // layout
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    zIndex: "9010",
    // look
    padding: "0",
    margin: "0",
    border: "none",
    cursor: "pointer"
  }), r.addEventListener("click", e), { element: r };
}
function so(t) {
  switch (t) {
    case "bottom-left":
      return { bottom: "24px", left: "24px" };
    case "bottom-right":
      return { bottom: "24px", right: "24px" };
    case "top-right":
      return { top: "24px", right: "24px" };
    case "top-left":
      return { top: "24px", left: "24px" };
  }
}
function io(t) {
  switch (t) {
    case "bottom-left":
      return { bottom: "24px", right: "24px", left: "60px", top: "72px" };
    case "bottom-right":
      return { bottom: "24px", left: "24px", right: "60px", top: "72px" };
    case "top-right":
      return { top: "24px", left: "24px", right: "60px", bottom: "72px" };
    case "top-left":
      return { top: "24px", right: "24px", left: "60px", bottom: "72px" };
  }
}
function oo(t) {
  const e = document.createElement("div");
  Object.assign(e.style, {
    position: "fixed",
    ...io(t.position),
    display: "block",
    borderRadius: "4px",
    border: "1px #ccc solid",
    boxShadow: "0px 0px 8px #00000044",
    backgroundColor: "#eee",
    zIndex: "999990"
  }), e.style.display = "none", e.className = "instant-devtool-container";
  function n() {
    return e.style.display !== "none";
  }
  return { element: e, isVisible: n };
}
const ao = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
};
function co() {
  return globalThis.__instantDbSchemaHashStore = globalThis.__instantDbSchemaHashStore ?? /* @__PURE__ */ new WeakMap(), globalThis.__instantDbSchemaHashStore;
}
function uo() {
  return globalThis.__instantDbStore = globalThis.__instantDbStore ?? {}, globalThis.__instantDbStore;
}
function ot(t) {
  const e = t.__adminToken;
  return t.appId + "_" + (t.websocketURI || "default_ws_uri") + "_" + (t.apiURI || "default_api_uri") + "_" + (e || "client_only") + "_" + t.useDateObjects;
}
const at = uo(), Ve = co();
class fo {
  constructor(e) {
    this.db = e, this.sendMagicCode = (n) => this.db.sendMagicCode(n), this.signInWithMagicCode = (n) => this.db.signInWithMagicCode(n), this.signInWithToken = (n) => this.db.signInWithCustomToken(n), this.signInAsGuest = () => this.db.signInAsGuest(), this.createAuthorizationURL = (n) => this.db.createAuthorizationURL(n), this.signInWithIdToken = (n) => this.db.signInWithIdToken(n), this.exchangeOAuthCode = (n) => this.db.exchangeCodeForToken(n), this.issuerURI = () => this.db.issuerURI(), this.signOut = (n = { invalidateToken: !0 }) => this.db.signOut(n);
  }
}
class lo {
  constructor(e) {
    this.db = e, this.uploadFile = (n, r, s = {}) => this.db.uploadFile(n, r, s), this.delete = (n) => this.db.deleteFile(n), this.upload = (n, r) => this.db.upload(n, r), this.put = this.upload, this.getDownloadUrl = (n) => this.db.getDownloadUrl(n);
  }
}
function mo(t) {
  return JSON.parse(JSON.stringify(t));
}
class ho {
  constructor(e) {
    this.tx = Cn(), this._reactor = e, this.auth = new fo(this._reactor), this.storage = new lo(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(db.tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(db.tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(db.tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    db.tx.todos[todoId].update({ title: 'Go on a run' }),
   *    db.tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(e) {
    return this._reactor.pushTx(e);
  }
  getLocalId(e) {
    return this._reactor.getLocalId(e);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(e, n, r) {
    return this._reactor.subscribeQuery(e, n, r);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(e) {
    return this._reactor.subscribeAuth(e);
  }
  /**
   * One time query for the logged in state. This is useful
   * for scenarios where you want to know the current auth
   * state without subscribing to changes.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const user = await db.getAuth();
   *   console.log('logged in as', user.email)
   */
  getAuth() {
    return this._reactor.getAuth();
  }
  /**
   * Listen for connection status changes to Instant. This is useful
   * for building things like connectivity indicators
   *
   * @see https://www.instantdb.com/docs/patterns#connection-status
   * @example
   *   const unsub = db.subscribeConnectionStatus((status) => {
   *     const connectionState =
   *       status === 'connecting' || status === 'opened'
   *         ? 'authenticating'
   *       : status === 'authenticated'
   *         ? 'connected'
   *       : status === 'closed'
   *         ? 'closed'
   *       : status === 'errored'
   *         ? 'errored'
   *       : 'unexpected state';
   *
   *     console.log('Connection status:', connectionState);
   *   });
   */
  subscribeConnectionStatus(e) {
    return this._reactor.subscribeConnectionStatus(e);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(e = "_defaultRoomType", n = "_defaultRoomId", r) {
    return {
      leaveRoom: this._reactor.joinRoom(n, r == null ? void 0 : r.initialPresence),
      subscribeTopic: (i, o) => this._reactor.subscribeTopic(n, i, o),
      subscribePresence: (i, o) => this._reactor.subscribePresence(e, n, i, o),
      publishTopic: (i, o) => this._reactor.publishTopic({ roomType: e, roomId: n, topic: i, data: o }),
      publishPresence: (i) => this._reactor.publishPresence(e, n, i),
      getPresence: (i) => this._reactor.getPresence(e, n, i)
    };
  }
  shutdown() {
    delete at[ot(this._reactor.config)], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(e, n) {
    return this._reactor.queryOnce(e, n);
  }
}
function nn(t) {
  if (!t)
    return "0";
  if (Ve.get(t))
    return Ve.get(t);
  const e = U(t);
  return Ve.set(t, e), e;
}
function po(t, e) {
  return nn(t._reactor.config.schema) !== nn(e);
}
function yo(t, e, n, r, s) {
  const i = at[ot(t)];
  if (i)
    return po(i, t.schema) && i._reactor.updateSchema(t.schema), i;
  const o = new Wi(
    {
      ...ao,
      ...t,
      cardinalityInference: !!t.schema
    },
    e || vn,
    n || In,
    { ...r || {}, "@instantdb/core": Rn },
    s
  ), a = new ho(
    o
  );
  return at[ot(t)] = a, bo(t.appId, t.devtool), a;
}
function bo(t, e) {
  if (typeof window > "u" || typeof window.location > "u" || typeof document > "u" || typeof e == "boolean" && !e)
    return;
  const n = {
    position: "bottom-right",
    allowedHosts: ["localhost"],
    ...typeof e == "object" ? e : {}
  };
  n.allowedHosts.includes(window.location.hostname) && eo(t, n);
}
const So = yo;
export {
  fo as Auth,
  vn as IndexedDBStorage,
  je as InstantAPIError,
  ho as InstantCoreDatabase,
  be as InstantError,
  O as QueryValidationError,
  lo as Storage,
  F as TransactionValidationError,
  In as WindowNetworkListener,
  mo as coerceQuery,
  Ns as getOps,
  _o as i,
  k as id,
  yo as init,
  So as init_experimental,
  wo as lookup,
  go as tx,
  Cn as txInit,
  Jt as validateQuery,
  Di as validateTransactions,
  Rn as version,
  U as weakHash
};
