// @instantdb/react-common/dist/standalone/index.js
import hr, { useRef as dt, useCallback as ye, useSyncExternalStore as $t, useEffect as we, useMemo as Dt, useState as pr } from "react";
var xt = { exports: {} }, De = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Rn;
function Oi() {
  if (Rn) return De;
  Rn = 1;
  var e = hr, t = Symbol.for("react.element"), n = Symbol.for("react.fragment"), r = Object.prototype.hasOwnProperty, i = e.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, s = { key: !0, ref: !0, __self: !0, __source: !0 };
  function o(u, a, d) {
    var c, f = {}, h = null, p = null;
    d !== void 0 && (h = "" + d), a.key !== void 0 && (h = "" + a.key), a.ref !== void 0 && (p = a.ref);
    for (c in a) r.call(a, c) && !s.hasOwnProperty(c) && (f[c] = a[c]);
    if (u && u.defaultProps) for (c in a = u.defaultProps, a) f[c] === void 0 && (f[c] = a[c]);
    return { $$typeof: t, type: u, key: h, ref: p, props: f, _owner: i.current };
  }
  return De.Fragment = n, De.jsx = o, De.jsxs = o, De;
}
var xe = {}, Mn;
function Ei() {
  if (Mn) return xe;
  Mn = 1;
  var e = {};
  /**
   * @license React
   * react-jsx-runtime.development.js
   *
   * Copyright (c) Facebook, Inc. and its affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   */
  return e.NODE_ENV !== "production" && function() {
    var t = hr, n = Symbol.for("react.element"), r = Symbol.for("react.portal"), i = Symbol.for("react.fragment"), s = Symbol.for("react.strict_mode"), o = Symbol.for("react.profiler"), u = Symbol.for("react.provider"), a = Symbol.for("react.context"), d = Symbol.for("react.forward_ref"), c = Symbol.for("react.suspense"), f = Symbol.for("react.suspense_list"), h = Symbol.for("react.memo"), p = Symbol.for("react.lazy"), b = Symbol.for("react.offscreen"), g = Symbol.iterator, w = "@@iterator";
    function T(l) {
      if (l === null || typeof l != "object")
        return null;
      var y = g && l[g] || l[w];
      return typeof y == "function" ? y : null;
    }
    var m = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function S(l) {
      {
        for (var y = arguments.length, _ = new Array(y > 1 ? y - 1 : 0), v = 1; v < y; v++)
          _[v - 1] = arguments[v];
        q("error", l, _);
      }
    }
    function q(l, y, _) {
      {
        var v = m.ReactDebugCurrentFrame, j = v.getStackAddendum();
        j !== "" && (y += "%s", _ = _.concat([j]));
        var P = _.map(function(C) {
          return String(C);
        });
        P.unshift("Warning: " + y), Function.prototype.apply.call(console[l], console, P);
      }
    }
    var O = !1, x = !1, N = !1, de = !1, ae = !1, Me;
    Me = Symbol.for("react.module.reference");
    function Be(l) {
      return !!(typeof l == "string" || typeof l == "function" || l === i || l === o || ae || l === s || l === c || l === f || de || l === b || O || x || N || typeof l == "object" && l !== null && (l.$$typeof === p || l.$$typeof === h || l.$$typeof === u || l.$$typeof === a || l.$$typeof === d || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      l.$$typeof === Me || l.getModuleId !== void 0));
    }
    function Qe(l, y, _) {
      var v = l.displayName;
      if (v)
        return v;
      var j = y.displayName || y.name || "";
      return j !== "" ? _ + "(" + j + ")" : _;
    }
    function k(l) {
      return l.displayName || "Context";
    }
    function U(l) {
      if (l == null)
        return null;
      if (typeof l.tag == "number" && S("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof l == "function")
        return l.displayName || l.name || null;
      if (typeof l == "string")
        return l;
      switch (l) {
        case i:
          return "Fragment";
        case r:
          return "Portal";
        case o:
          return "Profiler";
        case s:
          return "StrictMode";
        case c:
          return "Suspense";
        case f:
          return "SuspenseList";
      }
      if (typeof l == "object")
        switch (l.$$typeof) {
          case a:
            var y = l;
            return k(y) + ".Consumer";
          case u:
            var _ = l;
            return k(_._context) + ".Provider";
          case d:
            return Qe(l, l.render, "ForwardRef");
          case h:
            var v = l.displayName || null;
            return v !== null ? v : U(l.type) || "Memo";
          case p: {
            var j = l, P = j._payload, C = j._init;
            try {
              return U(C(P));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var z = Object.assign, B = 0, re, ie, Z, Ge, Ye, He, Je;
    function Ze() {
    }
    Ze.__reactDisabledLog = !0;
    function Tt() {
      {
        if (B === 0) {
          re = console.log, ie = console.info, Z = console.warn, Ge = console.error, Ye = console.group, He = console.groupCollapsed, Je = console.groupEnd;
          var l = {
            configurable: !0,
            enumerable: !0,
            value: Ze,
            writable: !0
          };
          Object.defineProperties(console, {
            info: l,
            log: l,
            warn: l,
            error: l,
            group: l,
            groupCollapsed: l,
            groupEnd: l
          });
        }
        B++;
      }
    }
    function Zr() {
      {
        if (B--, B === 0) {
          var l = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: z({}, l, {
              value: re
            }),
            info: z({}, l, {
              value: ie
            }),
            warn: z({}, l, {
              value: Z
            }),
            error: z({}, l, {
              value: Ge
            }),
            group: z({}, l, {
              value: Ye
            }),
            groupCollapsed: z({}, l, {
              value: He
            }),
            groupEnd: z({}, l, {
              value: Je
            })
          });
        }
        B < 0 && S("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var wt = m.ReactCurrentDispatcher, mt;
    function Xe(l, y, _) {
      {
        if (mt === void 0)
          try {
            throw Error();
          } catch (j) {
            var v = j.stack.trim().match(/\n( *(at )?)/);
            mt = v && v[1] || "";
          }
        return `
` + mt + l;
      }
    }
    var St = !1, et;
    {
      var Xr = typeof WeakMap == "function" ? WeakMap : Map;
      et = new Xr();
    }
    function _n(l, y) {
      if (!l || St)
        return "";
      {
        var _ = et.get(l);
        if (_ !== void 0)
          return _;
      }
      var v;
      St = !0;
      var j = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var P;
      P = wt.current, wt.current = null, Tt();
      try {
        if (y) {
          var C = function() {
            throw Error();
          };
          if (Object.defineProperty(C.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(C, []);
            } catch (Q) {
              v = Q;
            }
            Reflect.construct(l, [], C);
          } else {
            try {
              C.call();
            } catch (Q) {
              v = Q;
            }
            l.call(C.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (Q) {
            v = Q;
          }
          l();
        }
      } catch (Q) {
        if (Q && v && typeof Q.stack == "string") {
          for (var E = Q.stack.split(`
`), W = v.stack.split(`
`), R = E.length - 1, $ = W.length - 1; R >= 1 && $ >= 0 && E[R] !== W[$]; )
            $--;
          for (; R >= 1 && $ >= 0; R--, $--)
            if (E[R] !== W[$]) {
              if (R !== 1 || $ !== 1)
                do
                  if (R--, $--, $ < 0 || E[R] !== W[$]) {
                    var te = `
` + E[R].replace(" at new ", " at ");
                    return l.displayName && te.includes("<anonymous>") && (te = te.replace("<anonymous>", l.displayName)), typeof l == "function" && et.set(l, te), te;
                  }
                while (R >= 1 && $ >= 0);
              break;
            }
        }
      } finally {
        St = !1, wt.current = P, Zr(), Error.prepareStackTrace = j;
      }
      var Oe = l ? l.displayName || l.name : "", be = Oe ? Xe(Oe) : "";
      return typeof l == "function" && et.set(l, be), be;
    }
    function ei(l, y, _) {
      return _n(l, !1);
    }
    function ti(l) {
      var y = l.prototype;
      return !!(y && y.isReactComponent);
    }
    function tt(l, y, _) {
      if (l == null)
        return "";
      if (typeof l == "function")
        return _n(l, ti(l));
      if (typeof l == "string")
        return Xe(l);
      switch (l) {
        case c:
          return Xe("Suspense");
        case f:
          return Xe("SuspenseList");
      }
      if (typeof l == "object")
        switch (l.$$typeof) {
          case d:
            return ei(l.render);
          case h:
            return tt(l.type, y, _);
          case p: {
            var v = l, j = v._payload, P = v._init;
            try {
              return tt(P(j), y, _);
            } catch {
            }
          }
        }
      return "";
    }
    var $e = Object.prototype.hasOwnProperty, gn = {}, Tn = m.ReactDebugCurrentFrame;
    function nt(l) {
      if (l) {
        var y = l._owner, _ = tt(l.type, l._source, y ? y.type : null);
        Tn.setExtraStackFrame(_);
      } else
        Tn.setExtraStackFrame(null);
    }
    function ni(l, y, _, v, j) {
      {
        var P = Function.call.bind($e);
        for (var C in l)
          if (P(l, C)) {
            var E = void 0;
            try {
              if (typeof l[C] != "function") {
                var W = Error((v || "React class") + ": " + _ + " type `" + C + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof l[C] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw W.name = "Invariant Violation", W;
              }
              E = l[C](y, C, v, _, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (R) {
              E = R;
            }
            E && !(E instanceof Error) && (nt(j), S("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", v || "React class", _, C, typeof E), nt(null)), E instanceof Error && !(E.message in gn) && (gn[E.message] = !0, nt(j), S("Failed %s type: %s", _, E.message), nt(null));
          }
      }
    }
    var ri = Array.isArray;
    function vt(l) {
      return ri(l);
    }
    function ii(l) {
      {
        var y = typeof Symbol == "function" && Symbol.toStringTag, _ = y && l[Symbol.toStringTag] || l.constructor.name || "Object";
        return _;
      }
    }
    function si(l) {
      try {
        return wn(l), !1;
      } catch {
        return !0;
      }
    }
    function wn(l) {
      return "" + l;
    }
    function mn(l) {
      if (si(l))
        return S("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", ii(l)), wn(l);
    }
    var Sn = m.ReactCurrentOwner, oi = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, vn, On;
    function ui(l) {
      if ($e.call(l, "ref")) {
        var y = Object.getOwnPropertyDescriptor(l, "ref").get;
        if (y && y.isReactWarning)
          return !1;
      }
      return l.ref !== void 0;
    }
    function ai(l) {
      if ($e.call(l, "key")) {
        var y = Object.getOwnPropertyDescriptor(l, "key").get;
        if (y && y.isReactWarning)
          return !1;
      }
      return l.key !== void 0;
    }
    function ci(l, y) {
      typeof l.ref == "string" && Sn.current;
    }
    function di(l, y) {
      {
        var _ = function() {
          vn || (vn = !0, S("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", y));
        };
        _.isReactWarning = !0, Object.defineProperty(l, "key", {
          get: _,
          configurable: !0
        });
      }
    }
    function fi(l, y) {
      {
        var _ = function() {
          On || (On = !0, S("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", y));
        };
        _.isReactWarning = !0, Object.defineProperty(l, "ref", {
          get: _,
          configurable: !0
        });
      }
    }
    var li = function(l, y, _, v, j, P, C) {
      var E = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: n,
        // Built-in properties that belong on the element
        type: l,
        key: y,
        ref: _,
        props: C,
        // Record the component responsible for creating this element.
        _owner: P
      };
      return E._store = {}, Object.defineProperty(E._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(E, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: v
      }), Object.defineProperty(E, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: j
      }), Object.freeze && (Object.freeze(E.props), Object.freeze(E)), E;
    };
    function hi(l, y, _, v, j) {
      {
        var P, C = {}, E = null, W = null;
        _ !== void 0 && (mn(_), E = "" + _), ai(y) && (mn(y.key), E = "" + y.key), ui(y) && (W = y.ref, ci(y, j));
        for (P in y)
          $e.call(y, P) && !oi.hasOwnProperty(P) && (C[P] = y[P]);
        if (l && l.defaultProps) {
          var R = l.defaultProps;
          for (P in R)
            C[P] === void 0 && (C[P] = R[P]);
        }
        if (E || W) {
          var $ = typeof l == "function" ? l.displayName || l.name || "Unknown" : l;
          E && di(C, $), W && fi(C, $);
        }
        return li(l, E, W, j, v, Sn.current, C);
      }
    }
    var Ot = m.ReactCurrentOwner, En = m.ReactDebugCurrentFrame;
    function ve(l) {
      if (l) {
        var y = l._owner, _ = tt(l.type, l._source, y ? y.type : null);
        En.setExtraStackFrame(_);
      } else
        En.setExtraStackFrame(null);
    }
    var Et;
    Et = !1;
    function At(l) {
      return typeof l == "object" && l !== null && l.$$typeof === n;
    }
    function An() {
      {
        if (Ot.current) {
          var l = U(Ot.current.type);
          if (l)
            return `

Check the render method of \`` + l + "`.";
        }
        return "";
      }
    }
    function pi(l) {
      return "";
    }
    var Cn = {};
    function yi(l) {
      {
        var y = An();
        if (!y) {
          var _ = typeof l == "string" ? l : l.displayName || l.name;
          _ && (y = `

Check the top-level render call using <` + _ + ">.");
        }
        return y;
      }
    }
    function jn(l, y) {
      {
        if (!l._store || l._store.validated || l.key != null)
          return;
        l._store.validated = !0;
        var _ = yi(y);
        if (Cn[_])
          return;
        Cn[_] = !0;
        var v = "";
        l && l._owner && l._owner !== Ot.current && (v = " It was passed a child from " + U(l._owner.type) + "."), ve(l), S('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', _, v), ve(null);
      }
    }
    function Pn(l, y) {
      {
        if (typeof l != "object")
          return;
        if (vt(l))
          for (var _ = 0; _ < l.length; _++) {
            var v = l[_];
            At(v) && jn(v, y);
          }
        else if (At(l))
          l._store && (l._store.validated = !0);
        else if (l) {
          var j = T(l);
          if (typeof j == "function" && j !== l.entries)
            for (var P = j.call(l), C; !(C = P.next()).done; )
              At(C.value) && jn(C.value, y);
        }
      }
    }
    function bi(l) {
      {
        var y = l.type;
        if (y == null || typeof y == "string")
          return;
        var _;
        if (typeof y == "function")
          _ = y.propTypes;
        else if (typeof y == "object" && (y.$$typeof === d || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        y.$$typeof === h))
          _ = y.propTypes;
        else
          return;
        if (_) {
          var v = U(y);
          ni(_, l.props, "prop", v, l);
        } else if (y.PropTypes !== void 0 && !Et) {
          Et = !0;
          var j = U(y);
          S("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", j || "Unknown");
        }
        typeof y.getDefaultProps == "function" && !y.getDefaultProps.isReactClassApproved && S("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function _i(l) {
      {
        for (var y = Object.keys(l.props), _ = 0; _ < y.length; _++) {
          var v = y[_];
          if (v !== "children" && v !== "key") {
            ve(l), S("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", v), ve(null);
            break;
          }
        }
        l.ref !== null && (ve(l), S("Invalid attribute `ref` supplied to `React.Fragment`."), ve(null));
      }
    }
    var kn = {};
    function In(l, y, _, v, j, P) {
      {
        var C = Be(l);
        if (!C) {
          var E = "";
          (l === void 0 || typeof l == "object" && l !== null && Object.keys(l).length === 0) && (E += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var W = pi();
          W ? E += W : E += An();
          var R;
          l === null ? R = "null" : vt(l) ? R = "array" : l !== void 0 && l.$$typeof === n ? (R = "<" + (U(l.type) || "Unknown") + " />", E = " Did you accidentally export a JSX literal instead of a component?") : R = typeof l, S("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", R, E);
        }
        var $ = hi(l, y, _, j, P);
        if ($ == null)
          return $;
        if (C) {
          var te = y.children;
          if (te !== void 0)
            if (v)
              if (vt(te)) {
                for (var Oe = 0; Oe < te.length; Oe++)
                  Pn(te[Oe], l);
                Object.freeze && Object.freeze(te);
              } else
                S("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Pn(te, l);
        }
        if ($e.call(y, "key")) {
          var be = U(l), Q = Object.keys(y).filter(function(vi) {
            return vi !== "key";
          }), Ct = Q.length > 0 ? "{key: someKey, " + Q.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!kn[be + Ct]) {
            var Si = Q.length > 0 ? "{" + Q.join(": ..., ") + ": ...}" : "{}";
            S(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, Ct, be, Si, be), kn[be + Ct] = !0;
          }
        }
        return l === i ? _i($) : bi($), $;
      }
    }
    function gi(l, y, _) {
      return In(l, y, _, !0);
    }
    function Ti(l, y, _) {
      return In(l, y, _, !1);
    }
    var wi = Ti, mi = gi;
    xe.Fragment = i, xe.jsx = wi, xe.jsxs = mi;
  }(), xe;
}
var Ai = {};
Ai.NODE_ENV === "production" ? xt.exports = Oi() : xt.exports = Ei();
var rt = xt.exports;
function G(e) {
  if (typeof e == "number")
    return (Math.abs(e * 2654435761) >>> 0).toString(16);
  if (typeof e == "boolean")
    return e ? "1" : "0";
  if (e === null)
    return "null";
  if (e === void 0)
    return "undefined";
  if (typeof e == "string") {
    let t = 2166136261;
    for (let n = 0; n < e.length; n++)
      t ^= e.charCodeAt(n), t += (t << 1) + (t << 4) + (t << 7) + (t << 8) + (t << 24), t = t >>> 0;
    return t.toString(16);
  }
  if (Array.isArray(e)) {
    let t = 2166136261;
    for (let n = 0; n < e.length; n++) {
      t ^= (n + 1) * 2654435761;
      const r = G(e[n]);
      for (let i = 0; i < r.length; i++)
        t ^= r.charCodeAt(i), t *= 16777619, t = t >>> 0;
    }
    return t.toString(16);
  }
  if (typeof e == "object") {
    let t = 2166136261;
    const n = Object.keys(e).sort();
    for (let r = 0; r < n.length; r++) {
      const i = n[r], s = G(i);
      t ^= parseInt(s, 16), t *= 16777619, t = t >>> 0;
      const o = G(e[i]);
      t ^= parseInt(o, 16), t *= 16777619, t = t >>> 0;
    }
    return t.toString(16);
  }
  return G(String(e));
}
const M = {
  Remove: "remove",
  Replace: "replace",
  Add: "add"
}, yr = Symbol.for("__MUTATIVE_PROXY_DRAFT__"), Ci = Symbol("__MUTATIVE_RAW_RETURN_SYMBOL__"), ut = Symbol.iterator, ee = {
  mutable: "mutable",
  immutable: "immutable"
}, nn = {};
function Ne(e, t) {
  return e instanceof Map ? e.has(t) : Object.prototype.hasOwnProperty.call(e, t);
}
function $n(e, t) {
  if (t in e) {
    let n = Reflect.getPrototypeOf(e);
    for (; n; ) {
      const r = Reflect.getOwnPropertyDescriptor(n, t);
      if (r)
        return r;
      n = Reflect.getPrototypeOf(n);
    }
  }
}
function rn(e) {
  return Object.getPrototypeOf(e) === Set.prototype;
}
function sn(e) {
  return Object.getPrototypeOf(e) === Map.prototype;
}
function X(e) {
  var t;
  return (t = e.copy) !== null && t !== void 0 ? t : e.original;
}
function me(e) {
  return !!A(e);
}
function A(e) {
  return typeof e != "object" ? null : e == null ? void 0 : e[yr];
}
function on(e) {
  var t;
  const n = A(e);
  return n ? (t = n.copy) !== null && t !== void 0 ? t : n.original : e;
}
function oe(e, t) {
  if (!e || typeof e != "object")
    return !1;
  let n;
  return Object.getPrototypeOf(e) === Object.prototype || Array.isArray(e) || e instanceof Map || e instanceof Set || !!(t != null && t.mark) && ((n = t.mark(e, ee)) === ee.immutable || typeof n == "function");
}
function br(e, t = []) {
  if (Object.hasOwnProperty.call(e, "key")) {
    const n = e.parent.copy, r = A(pe(n, e.key));
    if (r !== null && (r == null ? void 0 : r.original) !== e.original)
      return null;
    const i = e.parent.type === 3, s = i ? Array.from(e.parent.setMap.keys()).indexOf(e.key) : e.key;
    if (!(i && n.size > s || Ne(n, s)))
      return null;
    t.push(s);
  }
  if (e.parent)
    return br(e.parent, t);
  t.reverse();
  try {
    ji(e.copy, t);
  } catch {
    return null;
  }
  return t;
}
function Se(e) {
  return Array.isArray(e) ? 1 : e instanceof Map ? 2 : e instanceof Set ? 3 : 0;
}
function pe(e, t) {
  return Se(e) === 2 ? e.get(t) : e[t];
}
function We(e, t, n) {
  Se(e) === 2 ? e.set(t, n) : e[t] = n;
}
function jt(e, t) {
  const n = A(e);
  return (n ? X(n) : e)[t];
}
function he(e, t) {
  return e === t ? e !== 0 || 1 / e === 1 / t : e !== e && t !== t;
}
function Ut(e) {
  if (e)
    for (; e.finalities.revoke.length > 0; )
      e.finalities.revoke.pop()();
}
function ge(e, t) {
  return t ? e : [""].concat(e).map((n) => {
    const r = `${n}`;
    return r.indexOf("/") === -1 && r.indexOf("~") === -1 ? r : r.replace(/~/g, "~0").replace(/\//g, "~1");
  }).join("/");
}
function ji(e, t) {
  for (let n = 0; n < t.length - 1; n += 1) {
    const r = t[n];
    if (e = pe(Se(e) === 3 ? Array.from(e) : e, r), typeof e != "object")
      throw new Error(`Cannot resolve patch at '${t.join("/")}'.`);
  }
  return e;
}
function Pi(e) {
  const t = Object.create(Object.getPrototypeOf(e));
  return Reflect.ownKeys(e).forEach((n) => {
    let r = Reflect.getOwnPropertyDescriptor(e, n);
    if (r.enumerable && r.configurable && r.writable) {
      t[n] = e[n];
      return;
    }
    r.writable || (r.writable = !0, r.configurable = !0), (r.get || r.set) && (r = {
      configurable: !0,
      writable: !0,
      enumerable: r.enumerable,
      value: e[n]
    }), Reflect.defineProperty(t, n, r);
  }), t;
}
const ki = Object.prototype.propertyIsEnumerable;
function _r(e, t) {
  let n;
  if (Array.isArray(e))
    return Array.prototype.concat.call(e);
  if (e instanceof Set) {
    if (!rn(e)) {
      const r = Object.getPrototypeOf(e).constructor;
      return new r(e.values());
    }
    return Set.prototype.difference ? Set.prototype.difference.call(e, /* @__PURE__ */ new Set()) : new Set(e.values());
  } else if (e instanceof Map) {
    if (!sn(e)) {
      const r = Object.getPrototypeOf(e).constructor;
      return new r(e);
    }
    return new Map(e);
  } else if (t != null && t.mark && (n = t.mark(e, ee), n !== void 0) && n !== ee.mutable) {
    if (n === ee.immutable)
      return Pi(e);
    if (typeof n == "function") {
      if (t.enablePatches || t.enableAutoFreeze)
        throw new Error("You can't use mark and patches or auto freeze together.");
      return n();
    }
    throw new Error(`Unsupported mark result: ${n}`);
  } else if (typeof e == "object" && Object.getPrototypeOf(e) === Object.prototype) {
    const r = {};
    return Object.keys(e).forEach((i) => {
      r[i] = e[i];
    }), Object.getOwnPropertySymbols(e).forEach((i) => {
      ki.call(e, i) && (r[i] = e[i]);
    }), r;
  } else
    throw new Error("Please check mark() to ensure that it is a stable marker draftable function.");
}
function K(e) {
  e.copy || (e.copy = _r(e.original, e.options));
}
function Le(e) {
  if (!oe(e))
    return on(e);
  if (Array.isArray(e))
    return e.map(Le);
  if (e instanceof Map) {
    const n = Array.from(e.entries()).map(([r, i]) => [
      r,
      Le(i)
    ]);
    if (!sn(e)) {
      const r = Object.getPrototypeOf(e).constructor;
      return new r(n);
    }
    return new Map(n);
  }
  if (e instanceof Set) {
    const n = Array.from(e).map(Le);
    if (!rn(e)) {
      const r = Object.getPrototypeOf(e).constructor;
      return new r(n);
    }
    return new Set(n);
  }
  const t = Object.create(Object.getPrototypeOf(e));
  for (const n in e)
    t[n] = Le(e[n]);
  return t;
}
function at(e) {
  return me(e) ? Le(e) : e;
}
function ce(e) {
  var t;
  e.assignedMap = (t = e.assignedMap) !== null && t !== void 0 ? t : /* @__PURE__ */ new Map(), e.operated || (e.operated = !0, e.parent && ce(e.parent));
}
function Dn() {
  throw new Error("Cannot modify frozen object");
}
function Ee(e, t, n, r, i) {
  {
    n = n ?? /* @__PURE__ */ new WeakMap(), r = r ?? [], i = i ?? [];
    const o = n.has(e) ? n.get(e) : e;
    if (r.length > 0) {
      const u = r.indexOf(o);
      if (o && typeof o == "object" && u !== -1)
        throw r[0] === o ? new Error("Forbids circular reference") : new Error(`Forbids circular reference: ~/${i.slice(0, u).map((a, d) => {
          if (typeof a == "symbol")
            return `[${a.toString()}]`;
          const c = r[d];
          return typeof a == "object" && (c instanceof Map || c instanceof Set) ? Array.from(c.keys()).indexOf(a) : a;
        }).join("/")}`);
      r.push(o), i.push(t);
    } else
      r.push(o);
  }
  if (Object.isFrozen(e) || me(e)) {
    r.pop(), i.pop();
    return;
  }
  switch (Se(e)) {
    case 2:
      for (const [u, a] of e)
        Ee(u, u, n, r, i), Ee(a, u, n, r, i);
      e.set = e.clear = e.delete = Dn;
      break;
    case 3:
      for (const u of e)
        Ee(u, u, n, r, i);
      e.add = e.clear = e.delete = Dn;
      break;
    case 1:
      Object.freeze(e);
      let o = 0;
      for (const u of e)
        Ee(u, o, n, r, i), o += 1;
      break;
    default:
      Object.freeze(e), Object.keys(e).forEach((u) => {
        const a = e[u];
        Ee(a, u, n, r, i);
      });
  }
  r.pop(), i.pop();
}
function un(e, t) {
  const n = Se(e);
  if (n === 0)
    Reflect.ownKeys(e).forEach((r) => {
      t(r, e[r], e);
    });
  else if (n === 1) {
    let r = 0;
    for (const i of e)
      t(r, i, e), r += 1;
  } else
    e.forEach((r, i) => t(i, r, e));
}
function gr(e, t, n) {
  if (me(e) || !oe(e, n) || t.has(e) || Object.isFrozen(e))
    return;
  const r = e instanceof Set, i = r ? /* @__PURE__ */ new Map() : void 0;
  if (t.add(e), un(e, (s, o) => {
    var u;
    if (me(o)) {
      const a = A(o);
      K(a);
      const d = !((u = a.assignedMap) === null || u === void 0) && u.size || a.operated ? a.copy : a.original;
      We(r ? i : e, s, d);
    } else
      gr(o, t, n);
  }), i) {
    const s = e, o = Array.from(s);
    s.clear(), o.forEach((u) => {
      s.add(i.has(u) ? i.get(u) : u);
    });
  }
}
function Ii(e, t) {
  const n = e.type === 3 ? e.setMap : e.copy;
  e.finalities.revoke.length > 1 && e.assignedMap.get(t) && n && gr(pe(n, t), e.finalities.handledSet, e.options);
}
function Lt(e) {
  e.type === 3 && e.copy && (e.copy.clear(), e.setMap.forEach((t) => {
    e.copy.add(on(t));
  }));
}
function Ft(e, t, n, r) {
  if (e.operated && e.assignedMap && e.assignedMap.size > 0 && !e.finalized) {
    if (n && r) {
      const s = br(e);
      s && t(e, s, n, r);
    }
    e.finalized = !0;
  }
}
function an(e, t, n, r) {
  const i = A(n);
  i && (i.callbacks || (i.callbacks = []), i.callbacks.push((s, o) => {
    var u;
    const a = e.type === 3 ? e.setMap : e.copy;
    if (he(pe(a, t), n)) {
      let d = i.original;
      i.copy && (d = i.copy), Lt(e), Ft(e, r, s, o), e.options.enableAutoFreeze && (e.options.updatedValues = (u = e.options.updatedValues) !== null && u !== void 0 ? u : /* @__PURE__ */ new WeakMap(), e.options.updatedValues.set(d, i.original)), We(a, t, d);
    }
  }), e.options.enableAutoFreeze && i.finalities !== e.finalities && (e.options.enableAutoFreeze = !1)), oe(n, e.options) && e.finalities.draft.push(() => {
    const s = e.type === 3 ? e.setMap : e.copy;
    he(pe(s, t), n) && Ii(e, t);
  });
}
function Ri(e, t, n, r, i) {
  let { original: s, assignedMap: o, options: u } = e, a = e.copy;
  a.length < s.length && ([s, a] = [a, s], [n, r] = [r, n]);
  for (let d = 0; d < s.length; d += 1)
    if (o.get(d.toString()) && a[d] !== s[d]) {
      const c = t.concat([d]), f = ge(c, i);
      n.push({
        op: M.Replace,
        path: f,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: at(a[d])
      }), r.push({
        op: M.Replace,
        path: f,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: at(s[d])
      });
    }
  for (let d = s.length; d < a.length; d += 1) {
    const c = t.concat([d]), f = ge(c, i);
    n.push({
      op: M.Add,
      path: f,
      // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
      value: at(a[d])
    });
  }
  if (s.length < a.length) {
    const { arrayLengthAssignment: d = !0 } = u.enablePatches;
    if (d) {
      const c = t.concat(["length"]), f = ge(c, i);
      r.push({
        op: M.Replace,
        path: f,
        value: s.length
      });
    } else
      for (let c = a.length; s.length < c; c -= 1) {
        const f = t.concat([c - 1]), h = ge(f, i);
        r.push({
          op: M.Remove,
          path: h
        });
      }
  }
}
function Mi({ original: e, copy: t, assignedMap: n }, r, i, s, o) {
  n.forEach((u, a) => {
    const d = pe(e, a), c = at(pe(t, a)), f = u ? Ne(e, a) ? M.Replace : M.Add : M.Remove;
    if (he(d, c) && f === M.Replace)
      return;
    const h = r.concat(a), p = ge(h, o);
    i.push(f === M.Remove ? { op: f, path: p } : { op: f, path: p, value: c }), s.push(f === M.Add ? { op: M.Remove, path: p } : f === M.Remove ? { op: M.Add, path: p, value: d } : { op: M.Replace, path: p, value: d });
  });
}
function $i({ original: e, copy: t }, n, r, i, s) {
  let o = 0;
  e.forEach((u) => {
    if (!t.has(u)) {
      const a = n.concat([o]), d = ge(a, s);
      r.push({
        op: M.Remove,
        path: d,
        value: u
      }), i.unshift({
        op: M.Add,
        path: d,
        value: u
      });
    }
    o += 1;
  }), o = 0, t.forEach((u) => {
    if (!e.has(u)) {
      const a = n.concat([o]), d = ge(a, s);
      r.push({
        op: M.Add,
        path: d,
        value: u
      }), i.unshift({
        op: M.Remove,
        path: d,
        value: u
      });
    }
    o += 1;
  });
}
function qe(e, t, n, r) {
  const { pathAsArray: i = !0 } = e.options.enablePatches;
  switch (e.type) {
    case 0:
    case 2:
      return Mi(e, t, n, r, i);
    case 1:
      return Ri(e, t, n, r, i);
    case 3:
      return $i(e, t, n, r, i);
  }
}
const ft = (e, t, n = !1) => {
  if (typeof e == "object" && e !== null && (!oe(e, t) || n))
    throw new Error("Strict mode: Mutable data cannot be accessed directly, please use 'unsafe(callback)' wrap.");
}, Nt = {
  get size() {
    return X(A(this)).size;
  },
  has(e) {
    return X(A(this)).has(e);
  },
  set(e, t) {
    const n = A(this), r = X(n);
    return (!r.has(e) || !he(r.get(e), t)) && (K(n), ce(n), n.assignedMap.set(e, !0), n.copy.set(e, t), an(n, e, t, qe)), this;
  },
  delete(e) {
    if (!this.has(e))
      return !1;
    const t = A(this);
    return K(t), ce(t), t.original.has(e) ? t.assignedMap.set(e, !1) : t.assignedMap.delete(e), t.copy.delete(e), !0;
  },
  clear() {
    const e = A(this);
    if (this.size) {
      K(e), ce(e), e.assignedMap = /* @__PURE__ */ new Map();
      for (const [t] of e.original)
        e.assignedMap.set(t, !1);
      e.copy.clear();
    }
  },
  forEach(e, t) {
    const n = A(this);
    X(n).forEach((r, i) => {
      e.call(t, this.get(i), i, this);
    });
  },
  get(e) {
    var t, n;
    const r = A(this), i = X(r).get(e), s = ((n = (t = r.options).mark) === null || n === void 0 ? void 0 : n.call(t, i, ee)) === ee.mutable;
    if (r.options.strict && ft(i, r.options, s), s || r.finalized || !oe(i, r.options) || i !== r.original.get(e))
      return i;
    const o = nn.createDraft({
      original: i,
      parentDraft: r,
      key: e,
      finalities: r.finalities,
      options: r.options
    });
    return K(r), r.copy.set(e, o), o;
  },
  keys() {
    return X(A(this)).keys();
  },
  values() {
    const e = this.keys();
    return {
      [ut]: () => this.values(),
      next: () => {
        const t = e.next();
        return t.done ? t : {
          done: !1,
          value: this.get(t.value)
        };
      }
    };
  },
  entries() {
    const e = this.keys();
    return {
      [ut]: () => this.entries(),
      next: () => {
        const t = e.next();
        if (t.done)
          return t;
        const n = this.get(t.value);
        return {
          done: !1,
          value: [t.value, n]
        };
      }
    };
  },
  [ut]() {
    return this.entries();
  }
}, Di = Reflect.ownKeys(Nt), xn = (e, t, { isValuesIterator: n }) => () => {
  var r, i;
  const s = t.next();
  if (s.done)
    return s;
  const o = s.value;
  let u = e.setMap.get(o);
  const a = A(u), d = ((i = (r = e.options).mark) === null || i === void 0 ? void 0 : i.call(r, u, ee)) === ee.mutable;
  if (e.options.strict && ft(o, e.options, d), !d && !a && oe(o, e.options) && !e.finalized && e.original.has(o)) {
    const c = nn.createDraft({
      original: o,
      parentDraft: e,
      key: o,
      finalities: e.finalities,
      options: e.options
    });
    e.setMap.set(o, c), u = c;
  } else a && (u = a.proxy);
  return {
    done: !1,
    value: n ? u : [u, u]
  };
}, lt = {
  get size() {
    return A(this).setMap.size;
  },
  has(e) {
    const t = A(this);
    if (t.setMap.has(e))
      return !0;
    K(t);
    const n = A(e);
    return !!(n && t.setMap.has(n.original));
  },
  add(e) {
    const t = A(this);
    return this.has(e) || (K(t), ce(t), t.assignedMap.set(e, !0), t.setMap.set(e, e), an(t, e, e, qe)), this;
  },
  delete(e) {
    if (!this.has(e))
      return !1;
    const t = A(this);
    K(t), ce(t);
    const n = A(e);
    return n && t.setMap.has(n.original) ? (t.assignedMap.set(n.original, !1), t.setMap.delete(n.original)) : (!n && t.setMap.has(e) ? t.assignedMap.set(e, !1) : t.assignedMap.delete(e), t.setMap.delete(e));
  },
  clear() {
    if (!this.size)
      return;
    const e = A(this);
    K(e), ce(e);
    for (const t of e.original)
      e.assignedMap.set(t, !1);
    e.setMap.clear();
  },
  values() {
    const e = A(this);
    K(e);
    const t = e.setMap.keys();
    return {
      [Symbol.iterator]: () => this.values(),
      next: xn(e, t, { isValuesIterator: !0 })
    };
  },
  entries() {
    const e = A(this);
    K(e);
    const t = e.setMap.keys();
    return {
      [Symbol.iterator]: () => this.entries(),
      next: xn(e, t, {
        isValuesIterator: !1
      })
    };
  },
  keys() {
    return this.values();
  },
  [ut]() {
    return this.values();
  },
  forEach(e, t) {
    const n = this.values();
    let r = n.next();
    for (; !r.done; )
      e.call(t, r.value, r.value, this), r = n.next();
  }
};
Set.prototype.difference && Object.assign(lt, {
  intersection(e) {
    return Set.prototype.intersection.call(new Set(this.values()), e);
  },
  union(e) {
    return Set.prototype.union.call(new Set(this.values()), e);
  },
  difference(e) {
    return Set.prototype.difference.call(new Set(this.values()), e);
  },
  symmetricDifference(e) {
    return Set.prototype.symmetricDifference.call(new Set(this.values()), e);
  },
  isSubsetOf(e) {
    return Set.prototype.isSubsetOf.call(new Set(this.values()), e);
  },
  isSupersetOf(e) {
    return Set.prototype.isSupersetOf.call(new Set(this.values()), e);
  },
  isDisjointFrom(e) {
    return Set.prototype.isDisjointFrom.call(new Set(this.values()), e);
  }
});
const xi = Reflect.ownKeys(lt), Tr = /* @__PURE__ */ new WeakSet(), wr = {
  get(e, t, n) {
    var r, i;
    const s = (r = e.copy) === null || r === void 0 ? void 0 : r[t];
    if (s && Tr.has(s))
      return s;
    if (t === yr)
      return e;
    let o;
    if (e.options.mark) {
      const d = t === "size" && (e.original instanceof Map || e.original instanceof Set) ? Reflect.get(e.original, t) : Reflect.get(e.original, t, n);
      if (o = e.options.mark(d, ee), o === ee.mutable)
        return e.options.strict && ft(d, e.options, !0), d;
    }
    const u = X(e);
    if (u instanceof Map && Di.includes(t)) {
      if (t === "size")
        return Object.getOwnPropertyDescriptor(Nt, "size").get.call(e.proxy);
      const d = Nt[t];
      if (d)
        return d.bind(e.proxy);
    }
    if (u instanceof Set && xi.includes(t)) {
      if (t === "size")
        return Object.getOwnPropertyDescriptor(lt, "size").get.call(e.proxy);
      const d = lt[t];
      if (d)
        return d.bind(e.proxy);
    }
    if (!Ne(u, t)) {
      const d = $n(u, t);
      return d ? "value" in d ? d.value : (
        // !case: support for getter
        (i = d.get) === null || i === void 0 ? void 0 : i.call(e.proxy)
      ) : void 0;
    }
    const a = u[t];
    if (e.options.strict && ft(a, e.options), e.finalized || !oe(a, e.options))
      return a;
    if (a === jt(e.original, t)) {
      if (K(e), e.copy[t] = cn({
        original: e.original[t],
        parentDraft: e,
        key: e.type === 1 ? Number(t) : t,
        finalities: e.finalities,
        options: e.options
      }), typeof o == "function") {
        const d = A(e.copy[t]);
        return K(d), ce(d), d.copy;
      }
      return e.copy[t];
    }
    return a;
  },
  set(e, t, n) {
    var r;
    if (e.type === 3 || e.type === 2)
      throw new Error("Map/Set draft does not support any property assignment.");
    let i;
    if (e.type === 1 && t !== "length" && !(Number.isInteger(i = Number(t)) && i >= 0 && (t === 0 || i === 0 || String(i) === String(t))))
      throw new Error("Only supports setting array indices and the 'length' property.");
    const s = $n(X(e), t);
    if (s != null && s.set)
      return s.set.call(e.proxy, n), !0;
    const o = jt(X(e), t), u = A(o);
    return u && he(u.original, n) ? (e.copy[t] = n, e.assignedMap = (r = e.assignedMap) !== null && r !== void 0 ? r : /* @__PURE__ */ new Map(), e.assignedMap.set(t, !1), !0) : (he(n, o) && (n !== void 0 || Ne(e.original, t)) || (K(e), ce(e), Ne(e.original, t) && he(n, e.original[t]) ? e.assignedMap.delete(t) : e.assignedMap.set(t, !0), e.copy[t] = n, an(e, t, n, qe)), !0);
  },
  has(e, t) {
    return t in X(e);
  },
  ownKeys(e) {
    return Reflect.ownKeys(X(e));
  },
  getOwnPropertyDescriptor(e, t) {
    const n = X(e), r = Reflect.getOwnPropertyDescriptor(n, t);
    return r && {
      writable: !0,
      configurable: e.type !== 1 || t !== "length",
      enumerable: r.enumerable,
      value: n[t]
    };
  },
  getPrototypeOf(e) {
    return Reflect.getPrototypeOf(e.original);
  },
  setPrototypeOf() {
    throw new Error("Cannot call 'setPrototypeOf()' on drafts");
  },
  defineProperty() {
    throw new Error("Cannot call 'defineProperty()' on drafts");
  },
  deleteProperty(e, t) {
    var n;
    return e.type === 1 ? wr.set.call(this, e, t, void 0, e.proxy) : (jt(e.original, t) !== void 0 || t in e.original ? (K(e), ce(e), e.assignedMap.set(t, !1)) : (e.assignedMap = (n = e.assignedMap) !== null && n !== void 0 ? n : /* @__PURE__ */ new Map(), e.assignedMap.delete(t)), e.copy && delete e.copy[t], !0);
  }
};
function cn(e) {
  const { original: t, parentDraft: n, key: r, finalities: i, options: s } = e, o = Se(t), u = {
    type: o,
    finalized: !1,
    parent: n,
    original: t,
    copy: null,
    proxy: null,
    finalities: i,
    options: s,
    // Mapping of draft Set items to their corresponding draft values.
    setMap: o === 3 ? new Map(t.entries()) : void 0
  };
  (r || "key" in e) && (u.key = r);
  const { proxy: a, revoke: d } = Proxy.revocable(o === 1 ? Object.assign([], u) : u, wr);
  if (i.revoke.push(d), Tr.add(a), u.proxy = a, n) {
    const c = n;
    c.finalities.draft.push((f, h) => {
      var p, b;
      const g = A(a);
      let w = c.type === 3 ? c.setMap : c.copy;
      const T = pe(w, r), m = A(T);
      if (m) {
        let S = m.original;
        m.operated && (S = on(T)), Lt(m), Ft(m, qe, f, h), c.options.enableAutoFreeze && (c.options.updatedValues = (p = c.options.updatedValues) !== null && p !== void 0 ? p : /* @__PURE__ */ new WeakMap(), c.options.updatedValues.set(S, m.original)), We(w, r, S);
      }
      (b = g.callbacks) === null || b === void 0 || b.forEach((S) => {
        S(f, h);
      });
    });
  } else {
    const c = A(a);
    c.finalities.draft.push((f, h) => {
      Lt(c), Ft(c, qe, f, h);
    });
  }
  return a;
}
nn.createDraft = cn;
function Ui(e, t, n, r, i) {
  var s;
  const o = A(e), u = (s = o == null ? void 0 : o.original) !== null && s !== void 0 ? s : e, a = !!t.length;
  if (o != null && o.operated)
    for (; o.finalities.draft.length > 0; )
      o.finalities.draft.pop()(n, r);
  const d = a ? t[0] : o ? o.operated ? o.copy : o.original : e;
  return o && Ut(o), i && Ee(d, d, o == null ? void 0 : o.options.updatedValues), [
    d,
    n && a ? [{ op: M.Replace, path: [], value: t[0] }] : n,
    r && a ? [{ op: M.Replace, path: [], value: u }] : r
  ];
}
function Li(e, t) {
  var n;
  const r = {
    draft: [],
    revoke: [],
    handledSet: /* @__PURE__ */ new WeakSet()
  };
  let i, s;
  t.enablePatches && (i = [], s = []);
  const u = ((n = t.mark) === null || n === void 0 ? void 0 : n.call(t, e, ee)) === ee.mutable || !oe(e, t) ? e : cn({
    original: e,
    parentDraft: null,
    finalities: r,
    options: t
  });
  return [
    u,
    (a = []) => {
      const [d, c, f] = Ui(u, a, i, s, t.enableAutoFreeze);
      return t.enablePatches ? [d, c, f] : d;
    }
  ];
}
function Vt(e) {
  const { rootDraft: t, value: n, useRawReturn: r = !1, isRoot: i = !0 } = e;
  un(n, (s, o, u) => {
    const a = A(o);
    if (a && t && a.finalities === t.finalities) {
      e.isContainDraft = !0;
      const d = a.original;
      if (u instanceof Set) {
        const c = Array.from(u);
        u.clear(), c.forEach((f) => u.add(s === f ? d : f));
      } else
        We(u, s, d);
    } else typeof o == "object" && o !== null && (e.value = o, e.isRoot = !1, Vt(e));
  }), i && (e.isContainDraft || console.warn("The return value does not contain any draft, please use 'rawReturn()' to wrap the return value to improve performance."), r && console.warn("The return value contains drafts, please don't use 'rawReturn()' to wrap the return value."));
}
function mr(e) {
  var t;
  const n = A(e);
  if (!oe(e, n == null ? void 0 : n.options))
    return e;
  const r = Se(e);
  if (n && !n.operated)
    return n.original;
  let i;
  function s() {
    i = r === 2 ? sn(e) ? new Map(e) : new (Object.getPrototypeOf(e)).constructor(e) : r === 3 ? Array.from(n.setMap.values()) : _r(e, n == null ? void 0 : n.options);
  }
  if (n) {
    n.finalized = !0;
    try {
      s();
    } finally {
      n.finalized = !1;
    }
  } else
    i = e;
  if (un(i, (o, u) => {
    if (n && he(pe(n.original, o), u))
      return;
    const a = mr(u);
    a !== u && (i === e && s(), We(i, o, a));
  }), r === 3) {
    const o = (t = n == null ? void 0 : n.original) !== null && t !== void 0 ? t : i;
    return rn(o) ? new Set(i) : new (Object.getPrototypeOf(o)).constructor(i);
  }
  return i;
}
function Un(e) {
  if (!me(e))
    throw new Error(`current() is only used for Draft, parameter: ${e}`);
  return mr(e);
}
const Fi = (e) => function t(n, r, i) {
  var s, o, u;
  if (typeof n == "function" && typeof r != "function")
    return function(O, ...x) {
      return t(O, (N) => n.call(this, N, ...x), r);
    };
  const a = n, d = r;
  let c = i;
  if (typeof r != "function" && (c = r), c !== void 0 && Object.prototype.toString.call(c) !== "[object Object]")
    throw new Error(`Invalid options: ${c}, 'options' should be an object.`);
  c = Object.assign(Object.assign({}, e), c);
  const f = me(a) ? Un(a) : a, h = Array.isArray(c.mark) ? (O, x) => {
    for (const N of c.mark) {
      if (typeof N != "function")
        throw new Error(`Invalid mark: ${N}, 'mark' should be a function.`);
      const de = N(O, x);
      if (de)
        return de;
    }
  } : c.mark, p = (s = c.enablePatches) !== null && s !== void 0 ? s : !1, b = (o = c.strict) !== null && o !== void 0 ? o : !1, w = {
    enableAutoFreeze: (u = c.enableAutoFreeze) !== null && u !== void 0 ? u : !1,
    mark: h,
    strict: b,
    enablePatches: p
  };
  if (!oe(f, w) && typeof f == "object" && f !== null)
    throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
  const [T, m] = Li(f, w);
  if (typeof r != "function") {
    if (!oe(f, w))
      throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
    return [T, m];
  }
  let S;
  try {
    S = d(T);
  } catch (O) {
    throw Ut(A(T)), O;
  }
  const q = (O) => {
    const x = A(T);
    if (!me(O)) {
      if (O !== void 0 && !he(O, T) && (x != null && x.operated))
        throw new Error("Either the value is returned as a new non-draft value, or only the draft is modified without returning any value.");
      const de = O == null ? void 0 : O[Ci];
      if (de) {
        const ae = de[0];
        return w.strict && typeof O == "object" && O !== null && Vt({
          rootDraft: x,
          value: O,
          useRawReturn: !0
        }), m([ae]);
      }
      if (O !== void 0)
        return typeof O == "object" && O !== null && Vt({ rootDraft: x, value: O }), m([O]);
    }
    if (O === T || O === void 0)
      return m([]);
    const N = A(O);
    if (w === N.options) {
      if (N.operated)
        throw new Error("Cannot return a modified child draft.");
      return m([Un(O)]);
    }
    return m([O]);
  };
  return S instanceof Promise ? S.then(q, (O) => {
    throw Ut(A(T)), O;
  }) : q(S);
}, qt = Fi();
Object.prototype.constructor.toString();
function Sr(e, t) {
  const n = Object.keys(e), r = Object.keys(t);
  return n.length === r.length && Object.keys(e).every((i) => t.hasOwnProperty(i));
}
function Ln(e, t) {
  return Object.keys(e).length === Object.keys(t).length && Object.keys(e).every((n) => t.hasOwnProperty(n) && e[n] === t[n]);
}
function ct(e, t) {
  return typeof e != "object" || typeof t != "object" || e === null || t === null ? e === t : Sr(e, t) ? Object.keys(e).every((n) => ct(e[n], t[n])) : !1;
}
function dn(e) {
  if (!Fe(e))
    return e;
  const t = {};
  for (const [n, r] of Object.entries(e))
    r !== void 0 && (t[n] = r);
  return t;
}
function vr(e, t) {
  if (!Fe(e) || !Fe(t))
    return t;
  const n = Object.assign({}, e);
  for (const r of Object.keys(t)) {
    if (t[r] === void 0)
      continue;
    if (t[r] === null) {
      delete n[r];
      continue;
    }
    const i = Fe(e[r]) && Fe(t[r]);
    n[r] = i ? vr(e[r], t[r]) : t[r];
  }
  return n;
}
function Fe(e) {
  return typeof e == "object" && e !== null && !Array.isArray(e);
}
function Ni(e, t, n) {
  if (!e || t.length === 0)
    return;
  let r = e || {};
  for (let s = 0; s < t.length - 1; s++) {
    const o = t[s];
    (!(o in r) || typeof r[o] != "object") && (r[o] = typeof t[s + 1] == "number" ? [] : {}), r = r[o];
  }
  const i = t[t.length - 1];
  Array.isArray(r) && typeof i == "number" ? r.splice(i, 0, n) : r[i] = n;
}
function Fn(e, t, n) {
  if (!e || t.length === 0)
    return;
  let r = e || {};
  for (let i = 0; i < t.length - 1; i++) {
    const s = t[i];
    (!(s in r) || typeof r[s] != "object") && (r[s] = typeof t[i + 1] == "number" ? [] : {}), r = r[s];
  }
  r[t[t.length - 1]] = n;
}
function Or(e, t) {
  if (!e || t.length === 0)
    return;
  const [n, ...r] = t;
  if (n in e) {
    if (r.length === 0) {
      Array.isArray(e) ? e.splice(n, 1) : delete e[n];
      return;
    }
    Or(e[n], r), Vi(e[n]) && delete e[n];
  }
}
function Vi(e) {
  return e && Object.keys(e).length === 0;
}
const Nn = /ZULU|YEKT|YEKST|YAPT|YAKT|YAKST|XJT|WGT|WGST|WFT|WETDST|WET|WDT|WAT|WAST|WAKT|WADT|VUT|VOLT|VLAT|VLAST|VET|UZT|UZST|UYT|UYST|UTC|UT|ULAT|ULAST|UCT|TVT|TRUT|TOT|TMT|TKT|TJT|TFT|TAHT|SGT|SCT|SAST|SADT|RET|PYT|PYST|PWT|PST|PONT|PMST|PMDT|PKT|PKST|PHT|PGT|PETT|PETST|PET|PDT|OMST|OMSST|NZT|NZST|NZDT|NUT|NST|NPT|NOVT|NOVST|NFT|NDT|MYT|MVT|MUT|MUST|MST|MSK|MSD|MPT|MMT|MHT|MEZ|METDST|MET|MESZ|MEST|MDT|MAWT|MART|MAGT|MAGST|LKT|LINT|LIGT|LHST|LHDT|KST|KRAT|KRAST|KOST|KGT|KGST|KDT|JST|JAYT|IST|IRT|IRKT|IRKST|IOT|IDT|ICT|HST|HKT|GYT|GMT|GILT|GFT|GET|GEST|GAMT|GALT|FNT|FNST|FKT|FKST|FJT|FJST|FET|EST|EGT|EGST|EETDST|EET|EEST|EDT|EAT|EAST|EASST|DDUT|DAVT|CXT|CST|COT|CLT|CLST|CKT|CHUT|CHAST|CHADT|CETDST|CET|CEST|CDT|CCT|CAST|CADT|BTT|BST|BRT|BRST|BRA|BOT|BORT|BNT|BDT|BDST|AZT|AZST|AZOT|AZOST|AWST|AWSST|AST|ART|ARST|ANAT|ANAST|AMT|AMST|ALMT|ALMST|AKST|AKDT|AFT|AEST|AESST|AEDT|ADT|ACWST|ACT|ACST|ACSST|ACDT$/, qi = {
  ZULU: 0,
  YEKT: 18e3,
  YEKST: 21600,
  YAPT: 36e3,
  YAKT: 32400,
  YAKST: 32400,
  XJT: 21600,
  WGT: -10800,
  WGST: -7200,
  WFT: 43200,
  WETDST: 3600,
  WET: 0,
  WDT: 32400,
  WAT: 3600,
  WAST: 25200,
  WAKT: 43200,
  WADT: 28800,
  VUT: 39600,
  VOLT: 10800,
  VLAT: 36e3,
  VLAST: 36e3,
  VET: -14400,
  UZT: 18e3,
  UZST: 21600,
  UYT: -10800,
  UYST: -7200,
  UTC: 0,
  UT: 0,
  ULAT: 28800,
  ULAST: 32400,
  UCT: 0,
  TVT: 43200,
  TRUT: 36e3,
  TOT: 46800,
  TMT: 18e3,
  TKT: 46800,
  TJT: 18e3,
  TFT: 18e3,
  TAHT: -36e3,
  SGT: 28800,
  SCT: 14400,
  SAST: 7200,
  SADT: 37800,
  RET: 14400,
  PYT: -14400,
  PYST: -10800,
  PWT: 32400,
  PST: -28800,
  PONT: 39600,
  PMST: -10800,
  PMDT: -7200,
  PKT: 18e3,
  PKST: 21600,
  PHT: 28800,
  PGT: 36e3,
  PETT: 43200,
  PETST: 43200,
  PET: -18e3,
  PDT: -25200,
  OMST: 21600,
  OMSST: 21600,
  NZT: 43200,
  NZST: 43200,
  NZDT: 46800,
  NUT: -39600,
  NST: -12600,
  NPT: 20700,
  NOVT: 25200,
  NOVST: 25200,
  NFT: -12600,
  NDT: -9e3,
  MYT: 28800,
  MVT: 18e3,
  MUT: 14400,
  MUST: 18e3,
  MST: -25200,
  MSK: 10800,
  MSD: 14400,
  MPT: 36e3,
  MMT: 23400,
  MHT: 43200,
  MEZ: 3600,
  METDST: 7200,
  MET: 3600,
  MESZ: 7200,
  MEST: 7200,
  MDT: -21600,
  MAWT: 18e3,
  MART: -34200,
  MAGT: 39600,
  MAGST: 39600,
  LKT: 19800,
  LINT: 50400,
  LIGT: 36e3,
  LHST: 37800,
  LHDT: 37800,
  KST: 32400,
  KRAT: 25200,
  KRAST: 25200,
  KOST: 39600,
  KGT: 21600,
  KGST: 21600,
  KDT: 36e3,
  JST: 32400,
  JAYT: 32400,
  IST: 7200,
  IRT: 12600,
  IRKT: 28800,
  IRKST: 28800,
  IOT: 21600,
  IDT: 10800,
  ICT: 25200,
  HST: -36e3,
  HKT: 28800,
  GYT: -14400,
  GMT: 0,
  GILT: 43200,
  GFT: -10800,
  GET: 14400,
  GEST: 14400,
  GAMT: -32400,
  GALT: -21600,
  FNT: -7200,
  FNST: -3600,
  FKT: -10800,
  FKST: -10800,
  FJT: 43200,
  FJST: 46800,
  FET: 10800,
  EST: -18e3,
  EGT: -3600,
  EGST: 0,
  EETDST: 10800,
  EET: 7200,
  EEST: 10800,
  EDT: -14400,
  EAT: 10800,
  EAST: -21600,
  EASST: -21600,
  DDUT: 36e3,
  DAVT: 25200,
  CXT: 25200,
  CST: -21600,
  COT: -18e3,
  CLT: -14400,
  CLST: -10800,
  CKT: -36e3,
  CHUT: 36e3,
  CHAST: 45900,
  CHADT: 49500,
  CETDST: 7200,
  CET: 3600,
  CEST: 7200,
  CDT: -18e3,
  CCT: 28800,
  CAST: 34200,
  CADT: 37800,
  BTT: 21600,
  BST: 3600,
  BRT: -10800,
  BRST: -7200,
  BRA: -10800,
  BOT: -14400,
  BORT: 28800,
  BNT: 28800,
  BDT: 21600,
  BDST: 7200,
  AZT: 14400,
  AZST: 14400,
  AZOT: -3600,
  AZOST: 0,
  AWST: 28800,
  AWSST: 32400,
  AST: -14400,
  ART: -10800,
  ARST: -10800,
  ANAT: 43200,
  ANAST: 43200,
  AMT: -14400,
  AMST: 14400,
  ALMT: 21600,
  ALMST: 25200,
  AKST: -32400,
  AKDT: -28800,
  AFT: 16200,
  AEST: 36e3,
  AESST: 39600,
  AEDT: 39600,
  ADT: -10800,
  ACWST: 31500,
  ACT: -18e3,
  ACST: 34200,
  ACSST: 37800,
  ACDT: 37800
};
function Wi(e) {
  return new Date(e);
}
function Ki(e) {
  return /* @__PURE__ */ new Date(e + "Z");
}
const zi = /^(\d+)[\./-](\d+)[\./-](\d+)$/;
function Bi(e) {
  const t = e.match(zi);
  if (!t)
    return null;
  const [n, r, i, s] = t;
  return r <= 0 || i <= 0 || s <= 0 ? null : r > 999 ? new Date(Date.UTC(r, i - 1, s, 0, 0, 0, 0)) : new Date(Date.UTC(s, r - 1, i, 0, 0, 0, 0));
}
function Qi(e) {
  const [t, n] = e.split(" ");
  return /* @__PURE__ */ new Date(t + "T" + n + "Z");
}
function Gi(e) {
  const [t, n] = e.split(" ");
  return /* @__PURE__ */ new Date(t + "T" + n + "Z");
}
function Yi(e) {
  return new Date(e);
}
function Hi(e) {
  const t = /^(\w{3}) (\w{3}) (\d{2}) (\d{4})$/;
  if (!e.match(t))
    throw new Error(`Unable to parse \`${e}\` as a date.`);
  const r = /* @__PURE__ */ new Date(e + " UTC");
  return new Date(Date.UTC(r.getUTCFullYear(), r.getUTCMonth(), r.getUTCDate(), 0, 0, 0, 0));
}
function Ji(e) {
  const t = /^(.+T.+)([+-])(\d{2})$/, n = e.match(t);
  if (n) {
    const [, r, i, s] = n, o = `${r}${i}${s}:00`;
    return new Date(o);
  }
  return null;
}
function Zi(e) {
  const [t, n] = e.split(", "), [r, i, s] = t.split("/").map(Number), o = n.match(/(\d{1,2}):(\d{2}):(\d{2}) (AM|PM)/);
  if (!o)
    throw new Error(`Unable to parse time from: ${e}`);
  let [, u, a, d, c] = o;
  return u = Number(u), a = Number(a), d = Number(d), c === "PM" && u !== 12 ? u += 12 : c === "AM" && u === 12 && (u = 0), new Date(Date.UTC(s, r - 1, i, u, a, d));
}
function Xi(e) {
  switch (e) {
    case "epoch":
      return /* @__PURE__ */ new Date(0);
    case "infinity":
    case "-infinity":
    case "today":
    case "tomorrow":
    case "yesterday":
      return null;
  }
}
function es(e) {
  const t = e.match(Nn);
  if (!t)
    return null;
  const [n] = t, r = qi[n], i = new Date(e.replace(Nn, "Z"));
  return new Date(i.getTime() - r * 1e3);
}
const ts = [
  Bi,
  Gi,
  Hi,
  Zi,
  Yi,
  Ki,
  Ji,
  Qi,
  Wi,
  Xi,
  es
];
function ns(e, t) {
  try {
    const n = e(t);
    return n instanceof Date && !isNaN(n.getTime()) ? n : null;
  } catch {
    return null;
  }
}
function Wt(e) {
  for (const t of ts) {
    const n = ns(t, e);
    if (n)
      return n;
  }
  return null;
}
function rs(e) {
  try {
    const t = JSON.parse(e);
    return typeof t == "string" ? Wt(t) : null;
  } catch {
    return null;
  }
}
function yt(e) {
  if (e !== void 0) {
    if (e === null)
      return null;
    if (e instanceof Date)
      return e;
    if (typeof e == "string") {
      const t = Wt(e) || rs(e) || Wt(e.trim());
      if (!t)
        throw new Error(`Unable to parse \`${e}\` as a date.`);
      return t;
    } else if (typeof e == "number")
      return new Date(e);
    throw new Error(`Invalid date value \`${e}\`. Expected a date, number, or string, got type ${typeof e}.`);
  }
}
function is(e) {
  return e.cardinality === "one";
}
function fn(e) {
  return e["value-type"] === "ref";
}
function ln(e) {
  return e["value-type"] === "blob";
}
function Ie(e, t) {
  return e[t];
}
function bt(e, t) {
  return t.reduce((n, r) => n && n.get(r), e);
}
function se(e, t) {
  if (t.length === 0)
    throw new Error("path must have at least one element");
  if (t.length === 1) {
    e.delete(t[0]);
    return;
  }
  const [n, ...r] = t;
  e.has(n) && se(e.get(n), r);
}
function Y(e, t, n) {
  if (t.length === 0)
    throw new Error("path must have at least one element");
  if (t.length === 1) {
    e.set(t[0], n);
    return;
  }
  const [r, ...i] = t;
  let s = e.get(r);
  s || (s = /* @__PURE__ */ new Map(), e.set(r, s)), Y(s, i, n);
}
function Er(e, t, n) {
  const r = /* @__PURE__ */ new Map(), i = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map();
  for (const o of t) {
    let [u, a, d, c] = o;
    const f = Ie(e, a);
    if (!f) {
      console.warn("no such attr", u, e);
      continue;
    }
    f["checked-data-type"] === "date" && n && (d = yt(d), o[2] = d), fn(f) && Y(s, [d, a, u], o), Y(r, [u, a, d], o), Y(i, [a, u, d], o);
  }
  return { eav: r, aev: i, vae: s };
}
function Ar(e) {
  const t = /* @__PURE__ */ new Map(), n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), i = /* @__PURE__ */ new Map();
  for (const s of Object.values(e)) {
    const o = s["forward-identity"], [u, a, d] = o, c = s["reverse-identity"];
    if (Y(r, [a, d], s), ln(s) && Y(t, [a, d], s), s["primary?"] && Y(n, [a], s), c) {
      const [f, h, p] = c;
      Y(i, [h, p], s);
    }
  }
  return { blobAttrs: t, primaryKeys: n, forwardIdents: r, revIdents: i };
}
function ss(e) {
  return {
    __type: e.__type,
    attrs: e.attrs,
    triples: H(e.eav, 3),
    cardinalityInference: e.cardinalityInference,
    linkIndex: e.linkIndex,
    useDateObjects: e.useDateObjects
  };
}
function os(e) {
  return Kt(e.attrs, e.triples, e.cardinalityInference, e.linkIndex, e.useDateObjects);
}
function hn(e) {
  e.attrIndexes = Ar(e.attrs);
}
function Kt(e, t, n, r, i) {
  const s = Er(e, t, i);
  return s.useDateObjects = i, s.attrs = e, s.attrIndexes = Ar(e), s.cardinalityInference = n, s.linkIndex = r, s.__type = "store", s;
}
function Ke(e, t) {
  var n, r;
  let i;
  if (Array.isArray(t[0])) {
    const [o, u] = t[0], a = e.aev.get(o);
    if (!a)
      return null;
    i = (n = H(a, 2).find((c) => c[2] === u)) === null || n === void 0 ? void 0 : n[0];
  } else
    i = t[0];
  if (!i)
    return null;
  const s = t[2];
  if (Array.isArray(s) && s.length === 2 && e.aev.get(s[0])) {
    const [o, u] = s, a = e.aev.get(o);
    if (!a)
      return null;
    const c = (r = H(a, 2).find((g) => g[2] === u)) === null || r === void 0 ? void 0 : r[0];
    if (!c)
      return null;
    const [f, h, p, ...b] = t;
    return [i, h, c, ...b];
  } else {
    const [o, ...u] = t;
    return [i, ...u];
  }
}
function us(e, t) {
  const n = Ke(e, t);
  if (!n)
    return;
  const [r, i, s] = n, o = Ie(e.attrs, i);
  o && (se(e.eav, [r, i, s]), se(e.aev, [i, r, s]), fn(o) && se(e.vae, [s, i, r]));
}
let as = 0;
function Cr(e, t, n) {
  const [r, i, s] = n;
  let o;
  const u = bt(e.ea, [r, i, s]);
  return u && (o = u[3]), o || Date.now() * 10 + as++;
}
function cs(e, t) {
  var n;
  const r = Ke(e, t);
  if (!r)
    return;
  let [i, s, o] = r;
  const u = Ie(e.attrs, s);
  if (!u)
    return;
  u["checked-data-type"] === "date" && e.useDateObjects && (o = yt(o));
  const a = bt(e.eav, [i, s, o]), d = (n = a == null ? void 0 : a[3]) !== null && n !== void 0 ? n : Cr(e, u, r), c = [i, s, o, d];
  is(u) ? (Y(e.eav, [i, s], /* @__PURE__ */ new Map([[o, c]])), Y(e.aev, [s, i], /* @__PURE__ */ new Map([[o, c]]))) : (Y(e.eav, [i, s, o], c), Y(e.aev, [s, i, o], c)), fn(u) && Y(e.vae, [o, s, i], c);
}
function ds(e, t) {
  var n;
  const r = Ke(e, t);
  if (!r)
    return;
  const [i, s, o] = r, u = Ie(e.attrs, s);
  if (!u)
    return;
  if (!ln(u))
    throw new Error("merge operation is not supported for links");
  const a = bt(e.eav, [i, s]);
  if (!a)
    return;
  const d = (n = a.values().next()) === null || n === void 0 ? void 0 : n.value;
  if (!d)
    return;
  const c = d[2], f = vr(c, o), h = [
    i,
    s,
    f,
    Cr(e, u, d)
  ];
  Y(e.eav, [i, s], /* @__PURE__ */ new Map([[f, h]]));
}
function zt(e, t) {
  var n, r;
  const [i, s] = t, o = Ke(e, [i]);
  if (!o)
    return;
  const [u] = o, a = e.eav.get(u);
  if (a) {
    for (const c of a.keys()) {
      const f = e.attrs[c];
      f && f["on-delete-reverse"] === "cascade" && H(a.get(c), 1).forEach(([h, p, b]) => {
        var g;
        return zt(e, [b, (g = f["reverse-identity"]) === null || g === void 0 ? void 0 : g[1]]);
      }), // Fall back to deleting everything if we've rehydrated tx-steps from
      // the store that didn't set `etype` in deleteEntity
      (!s || // If we don't know about the attr, let's just get rid of it
      !f || // Make sure it matches the etype
      ((n = f["forward-identity"]) === null || n === void 0 ? void 0 : n[1]) === s) && (se(e.aev, [c, u]), se(e.eav, [u, c]));
    }
    a.size === 0 && se(e.eav, [u]);
  }
  const d = e.vae.get(u) && H(e.vae.get(u), 2);
  d && d.forEach((c) => {
    var f, h, p;
    const [b, g, w] = c, T = e.attrs[g];
    (!s || !T || ((f = T["reverse-identity"]) === null || f === void 0 ? void 0 : f[1]) === s) && (se(e.eav, [b, g, w]), se(e.aev, [g, b, w]), se(e.vae, [w, g, b])), T && T["on-delete"] === "cascade" && ((h = T["reverse-identity"]) === null || h === void 0 ? void 0 : h[1]) === s && zt(e, [b, (p = T["forward-identity"]) === null || p === void 0 ? void 0 : p[1]]);
  }), ((r = e.vae.get(u)) === null || r === void 0 ? void 0 : r.size) === 0 && se(e.vae, [u]);
}
function jr(e, t) {
  const n = Er(e.attrs, t, e.useDateObjects);
  Object.keys(n).forEach((r) => {
    e[r] = n[r];
  });
}
function fs(e, [t]) {
  e.attrs[t.id] = t, hn(e);
}
function Pr(e) {
  return H(e.eav, 3);
}
function ls(e, [t]) {
  if (!e.attrs[t])
    return;
  const n = Pr(e).filter(([r, i]) => i !== t);
  delete e.attrs[t], hn(e), jr(e, n);
}
function hs(e, [t]) {
  const n = e.attrs[t.id];
  n && (e.attrs[t.id] = Object.assign(Object.assign({}, n), t), hn(e), jr(e, Pr(e)));
}
function ps(e, t) {
  const [n, ...r] = t;
  switch (n) {
    case "add-triple":
      cs(e, r);
      break;
    case "deep-merge-triple":
      ds(e, r);
      break;
    case "retract-triple":
      us(e, r);
      break;
    case "delete-entity":
      zt(e, r);
      break;
    case "add-attr":
      fs(e, r);
      break;
    case "delete-attr":
      ls(e, r);
      break;
    case "update-attr":
      hs(e, r);
      break;
    case "restore-attr":
      break;
    case "rule-params":
      break;
    default:
      throw new Error(`unhandled transaction action: ${n}`);
  }
}
function H(e, t, n = []) {
  if (!e || t === 0)
    return n;
  if (t === 1) {
    for (const r of e.values())
      n.push(r);
    return n;
  }
  for (const r of e.values())
    H(r, t - 1, n);
  return n;
}
function it(e, t, n) {
  var r, i;
  const s = [];
  if (n != null && n.hasOwnProperty("$not")) {
    for (const u of t.keys())
      n.$not !== u && s.push(t.get(u));
    return s;
  }
  if (n != null && n.hasOwnProperty("$isNull")) {
    const { attrId: u, isNull: a, reverse: d } = n.$isNull;
    if (d)
      for (const c of t.keys()) {
        const f = e.vae.get(c), h = !f || ((r = f.get(u)) === null || r === void 0 ? void 0 : r.get(null)) || !f.get(u);
        (a ? h : !h) && s.push(t.get(c));
      }
    else {
      const c = e.aev.get(u);
      for (const f of t.keys()) {
        const h = !c || ((i = c.get(f)) === null || i === void 0 ? void 0 : i.get(null)) || !c.get(f);
        (a ? h : !h) && s.push(t.get(f));
      }
    }
    return s;
  }
  if (n != null && n.$comparator)
    return H(t, 1).filter(n.$op);
  const o = n.in || n.$in || [n];
  for (const u of o) {
    const a = t.get(u);
    a && s.push(a);
  }
  return s;
}
function ys(e, t, n) {
  let r = "";
  return e !== void 0 && (r += "e"), t !== void 0 && (r += "a"), n !== void 0 && (r += "v"), r;
}
function bs(e, [t, n, r]) {
  var i, s;
  switch (ys(t, n, r)) {
    case "e": {
      const u = e.eav.get(t);
      return H(u, 2);
    }
    case "ea": {
      const u = (i = e.eav.get(t)) === null || i === void 0 ? void 0 : i.get(n);
      return H(u, 1);
    }
    case "eav": {
      const u = (s = e.eav.get(t)) === null || s === void 0 ? void 0 : s.get(n);
      return u ? it(e, u, r) : [];
    }
    case "ev": {
      const u = e.eav.get(t);
      if (!u)
        return [];
      const a = [];
      for (const d of u.values())
        a.push(...it(e, d, r));
      return a;
    }
    case "a": {
      const u = e.aev.get(n);
      return H(u, 2);
    }
    case "av": {
      const u = e.aev.get(n);
      if (!u)
        return [];
      const a = [];
      for (const d of u.values())
        a.push(...it(e, d, r));
      return a;
    }
    case "v": {
      const u = [];
      for (const a of e.eav.values())
        for (const d of a.values())
          u.push(...it(e, d, r));
      return u;
    }
    default:
      return H(e.eav, 3);
  }
}
function _s(e, t, n) {
  var r;
  const i = {};
  for (const [s, o] of t.entries()) {
    const u = (r = e.eav.get(n)) === null || r === void 0 ? void 0 : r.get(o.id), a = H(u, 1);
    for (const d of a)
      i[s] = d[2];
  }
  return i;
}
function Ce(e, t, n) {
  var r;
  return (r = e.attrIndexes.forwardIdents.get(t)) === null || r === void 0 ? void 0 : r.get(n);
}
function kr(e, t, n) {
  var r;
  return (r = e.attrIndexes.revIdents.get(t)) === null || r === void 0 ? void 0 : r.get(n);
}
function gs(e, t) {
  return e.attrIndexes.blobAttrs.get(t);
}
function Ir(e, t) {
  var n;
  const r = e.attrIndexes.primaryKeys.get(t);
  return r || ((n = e.attrIndexes.forwardIdents.get(t)) === null || n === void 0 ? void 0 : n.get("id"));
}
function Ts(e, t) {
  const n = Ke(e, t);
  if (!n)
    return;
  const [r, i, s] = n;
  if (Ie(e.attrs, i))
    return bt(e.eav, [r, i]);
}
function ws(e, t) {
  const n = t.filter(([r, i, s, o, u]) => {
    if (r !== "add-triple" && r !== "deep-merge-triple")
      return !0;
    const a = u == null ? void 0 : u.mode;
    if (a !== "create" && a !== "update")
      return !0;
    let d = !1;
    const c = Ie(e.attrs, s);
    if (c) {
      const f = Ir(e, c["forward-identity"][1]);
      d = !!Ts(e, [i, f.id, i]);
    }
    return !(a === "create" && d || a === "update" && !d);
  });
  return qt(e, (r) => {
    n.forEach((i) => {
      ps(r, i);
    });
  });
}
function ms(e) {
  return typeof e == "string" && e.startsWith("?");
}
function Ss(e, t, n) {
  if (n.hasOwnProperty(e)) {
    const r = n[e];
    return Rr(r, t, n);
  }
  return Object.assign(Object.assign({}, n), { [e]: t });
}
function Vn(e, t, n) {
  return e === t ? n : null;
}
function vs(e) {
  switch (typeof e) {
    case "string":
      return e.startsWith("?") ? Ss : Vn;
    default:
      return Vn;
  }
}
const Os = [
  "in",
  "$in",
  "$not",
  "$isNull",
  "$comparator"
  // covers all of $gt, $lt, etc.
];
function Es(e) {
  for (const t of Os)
    if (e.hasOwnProperty(t))
      return !0;
  return !1;
}
function Rr(e, t, n) {
  return n ? typeof e == "object" ? Es(e) ? n : null : vs(e)(e, t, n) : null;
}
function As(e, t, n) {
  return e.reduce((r, i, s) => {
    const o = t[s];
    return Rr(i, o, r);
  }, n);
}
function Cs(e, t, n) {
  return ks(e, t, n).map((r) => As(t, r, n)).filter((r) => r);
}
function js(e, t, n) {
  return t.or ? t.or.patterns.flatMap((r) => Bt(e, r, n)) : t.and ? t.and.patterns.reduce((r, i) => Bt(e, i, r), n) : n.flatMap((r) => Cs(e, t, r));
}
function Bt(e, t, n = [{}]) {
  return t.reduce((r, i) => js(e, i, r), n);
}
function pn(e, t) {
  return Array.isArray(t) ? t.map((n) => pn(e, n)) : ms(t) ? e[t] : t;
}
function Ps(e, { find: t, where: n }) {
  return Bt(e, n).map((i) => pn(i, t));
}
function ks(e, t, n) {
  return bs(e, pn(n, t));
}
const Is = /^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000|ffffffff-ffff-ffff-ffff-ffffffffffff)$/i;
function je(e) {
  return typeof e == "string" && Is.test(e);
}
const L = [];
for (let e = 0; e < 256; ++e)
  L.push((e + 256).toString(16).slice(1));
function Rs(e, t = 0) {
  return (L[e[t + 0]] + L[e[t + 1]] + L[e[t + 2]] + L[e[t + 3]] + "-" + L[e[t + 4]] + L[e[t + 5]] + "-" + L[e[t + 6]] + L[e[t + 7]] + "-" + L[e[t + 8]] + L[e[t + 9]] + "-" + L[e[t + 10]] + L[e[t + 11]] + L[e[t + 12]] + L[e[t + 13]] + L[e[t + 14]] + L[e[t + 15]]).toLowerCase();
}
let Pt;
const Ms = new Uint8Array(16);
function $s() {
  if (!Pt) {
    if (typeof crypto > "u" || !crypto.getRandomValues)
      throw new Error("crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported");
    Pt = crypto.getRandomValues.bind(crypto);
  }
  return Pt(Ms);
}
const Ds = typeof crypto < "u" && crypto.randomUUID && crypto.randomUUID.bind(crypto), qn = { randomUUID: Ds };
function xs(e, t, n) {
  var i;
  if (qn.randomUUID && !e)
    return qn.randomUUID();
  e = e || {};
  const r = e.random ?? ((i = e.rng) == null ? void 0 : i.call(e)) ?? $s();
  if (r.length < 16)
    throw new Error("Random bytes length must be >= 16");
  return r[6] = r[6] & 15 | 64, r[8] = r[8] & 63 | 128, Rs(r);
}
function Wn(e) {
  const t = e.replace(/-/g, ""), n = [];
  for (let r = 0; r < t.length; r += 2)
    n.push(parseInt(t.substring(r, r + 2), 16));
  return n;
}
function Us(e, t) {
  for (let n = 0; n < e.length; n++) {
    if (e[n] < t[n])
      return -1;
    if (e[n] > t[n])
      return 1;
  }
  return 0;
}
function Ls(e, t) {
  return Us(Wn(e), Wn(t));
}
function V() {
  return xs();
}
function Fs(e, t) {
  return e.localeCompare(t);
}
function Ns() {
  let e = Fs;
  if (typeof Intl == "object" && Intl.hasOwnProperty("Collator"))
    try {
      e = Intl.Collator("en-US").compare;
    } catch {
    }
  return e;
}
const Vs = Ns();
let qs = 0;
function Ve(e) {
  return _t(`_${e}`, qs++);
}
function _t(e, t) {
  return `?${e}-${t}`;
}
class Pe extends Error {
  constructor(t) {
    super(t), this.name = "AttrNotFoundError";
  }
}
function Ws(e, t) {
  const n = Ir(e, t);
  if (!n)
    throw new Pe(`Could not find id attr for ${t}`);
  return n;
}
function Kn(e, t, n, r) {
  return [Ks(e, t, n, r)];
}
function Ks(e, t, n, r) {
  return [
    e(n, r),
    Ws(t, n).id,
    e(n, r),
    e("time", r)
  ];
}
function zs(e, t, n) {
  return e.map((r) => r === t ? n : r);
}
function Mr(e, t, n, r, i) {
  const s = Ce(t, n, i), o = kr(t, n, i), u = s || o;
  if (!u)
    throw new Pe(`Could not find attr for ${[n, i]}`);
  if (u["value-type"] !== "ref")
    throw new Error(`Attr ${u.id} is not a ref`);
  const [a, d] = u["forward-identity"], [c, f] = u["reverse-identity"], h = r + 1, p = s ? [
    e(d, r),
    u.id,
    e(f, h),
    Ve("time")
  ] : [
    e(d, h),
    u.id,
    e(f, r),
    Ve("time")
  ];
  return [s ? f : d, h, p, u, !!s];
}
function zn(e, t) {
  if (typeof t != "string")
    return function(o) {
      return !1;
    };
  const r = t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&").replace(/%/g, ".*").replace(/_/g, "."), i = new RegExp(`^${r}$`, e ? void 0 : "i");
  return function(o) {
    return typeof o != "string" ? !1 : i.test(o);
  };
}
function Bs(e, t) {
  if (typeof t != "object" || t.hasOwnProperty("$in") || t.hasOwnProperty("in"))
    return t;
  const n = e["checked-data-type"] === "date";
  if (t.hasOwnProperty("$gt"))
    return {
      $comparator: !0,
      $op: n ? function(i) {
        return new Date(i[2]) > new Date(t.$gt);
      } : function(i) {
        return i[2] > t.$gt;
      }
    };
  if (t.hasOwnProperty("$gte"))
    return {
      $comparator: !0,
      $op: n ? function(i) {
        return new Date(i[2]) >= new Date(t.$gte);
      } : function(i) {
        return i[2] >= t.$gte;
      }
    };
  if (t.hasOwnProperty("$lt"))
    return {
      $comparator: !0,
      $op: n ? function(i) {
        return new Date(i[2]) < new Date(t.$lt);
      } : function(i) {
        return i[2] < t.$lt;
      }
    };
  if (t.hasOwnProperty("$lte"))
    return {
      $comparator: !0,
      $op: n ? function(i) {
        return new Date(i[2]) <= new Date(t.$lte);
      } : function(i) {
        return i[2] <= t.$lte;
      }
    };
  if (t.hasOwnProperty("$like")) {
    const r = zn(!0, t.$like);
    return {
      $comparator: !0,
      $op: function(s) {
        return r(s[2]);
      }
    };
  }
  if (t.hasOwnProperty("$ilike")) {
    const r = zn(!1, t.$ilike);
    return {
      $comparator: !0,
      $op: function(s) {
        return r(s[2]);
      }
    };
  }
  return t;
}
function Qs(e, t, n, r, i, s) {
  const o = Ce(t, n, i), u = kr(t, n, i), a = o || u;
  if (!a)
    throw new Pe(`No attr for etype = ${n} label = ${i}`);
  if (s != null && s.hasOwnProperty("$isNull")) {
    const d = Ce(t, n, "id");
    if (!d)
      throw new Pe(`No attr for etype = ${n} label = id`);
    return [
      e(n, r),
      d.id,
      { $isNull: { attrId: a.id, isNull: s.$isNull, reverse: !o } },
      Ve("time")
    ];
  }
  return o ? [
    e(n, r),
    a.id,
    Bs(a, s),
    Ve("time")
  ] : [s, a.id, e(n, r), Ve("time")];
}
function Gs(e, t, n, r, i) {
  const [s, o, u] = i.reduce((a, d) => {
    const [c, f, h] = a, [p, b, g] = Mr(e, t, c, f, d);
    return [p, b, [...h, g]];
  }, [n, r, []]);
  return [s, o, u];
}
function Qt(e, t, n, r, i, s) {
  const o = i.slice(0, i.length - 1), u = i[i.length - 1], [a, d, c] = Gs(e, t, n, r, o), f = Qs(e, t, a, d, u, s);
  return c.concat([f]);
}
function Ys(e, t) {
  return t ? [t].concat(e) : e;
}
function Hs([e, t]) {
  return e === "or" && Array.isArray(t);
}
function Js([e, t]) {
  return e === "and" && Array.isArray(t);
}
function Zs(e, t, n) {
  return (r, i) => {
    const s = e(r, i);
    return t == s ? s : `${s}-${n}`;
  };
}
function Bn(e, t, n, r, i, s) {
  const o = e(r, i), u = s.map((a, d) => {
    const c = Zs(e, o, d);
    return $r(c, n, r, i, a);
  });
  return { [t]: { patterns: u, joinSym: o } };
}
function Xs(e) {
  const t = [];
  for (let n = 1; n <= e.length; n++)
    t.push(e.slice(0, n));
  return t;
}
function Qn(e, t, n, r, i) {
  return Xs(i).map((s) => Qt(e, t, n, r, s, { $isNull: !0 }));
}
function $r(e, t, n, r, i) {
  return Object.entries(i).flatMap(([s, o]) => {
    if (Hs([s, o]))
      return Bn(e, "or", t, n, r, o);
    if (Js([s, o]))
      return Bn(e, "and", t, n, r, o);
    if (s === "$entityIdStartsWith")
      return [];
    const u = s.split(".");
    if (o != null && o.hasOwnProperty("$ne") && (o = Object.assign(Object.assign({}, o), { $not: o.$ne }), delete o.$ne), o != null && o.hasOwnProperty("$not")) {
      const a = Qt(e, t, n, r, u, o), d = Qn(e, t, n, r, u);
      return [
        {
          or: {
            patterns: [a, ...d],
            joinSym: e(n, r)
          }
        }
      ];
    }
    return o != null && o.hasOwnProperty("$isNull") && o.$isNull === !0 && u.length > 1 ? [
      {
        or: {
          patterns: Qn(e, t, n, r, u),
          joinSym: e(n, r)
        }
      }
    ] : Qt(e, t, n, r, u, o);
  });
}
function eo(e, t, n, r) {
  const i = _t;
  return r ? $r(i, e, t, n, r).concat(Kn(i, e, t, n)) : Kn(i, e, t, n);
}
function to(e, t, n) {
  return [e(t, n), e("time", n)];
}
function no(e, t, n, r, i, s) {
  const [o, u, a, d, c] = Mr(e, t, n, r, i), f = zs(a, e(n, r), s);
  return [o, u, f, d, c];
}
function ro(e, t, { etype: n, level: r, form: i }, s) {
  const o = Object.keys(i).filter((u) => u !== "$");
  return o.length ? Object.entries(s).map(function([a, d]) {
    return o.map(function(h) {
      var p, b, g;
      const w = !!(t.cardinalityInference && (!((g = (b = (p = t.linkIndex) === null || p === void 0 ? void 0 : p[n]) === null || b === void 0 ? void 0 : b[h]) === null || g === void 0) && g.isSingular));
      try {
        const [T, m, S] = no(e, t, n, r, h, a), q = xr(t, {
          etype: T,
          level: m,
          form: i[h],
          join: S
        }), O = w ? q[0] : q;
        return { [h]: O };
      } catch (T) {
        if (T instanceof Pe)
          return { [h]: w ? void 0 : [] };
        throw T;
      }
    }).reduce(function(h, p) {
      return Object.assign(Object.assign({}, h), p);
    }, d);
  }) : Object.values(s);
}
function io(e, t, n) {
  return n === "string" ? Vs(e, t) : e > t ? 1 : -1;
}
function ht([e, t], [n, r], i) {
  return t === r || t == null && r == null ? Ls(e, n) : r == null ? 1 : t == null ? -1 : io(t, r, i);
}
function Gt(e) {
  return e == null ? e : new Date(e).getTime();
}
function so(e, t, n, r) {
  var i;
  const [s, o, u, a] = e, d = n === "desc" ? 1 : -1;
  if (((i = t["forward-identity"]) === null || i === void 0 ? void 0 : i[2]) === "id")
    return ht(r, [s, a], null) === d;
  const [c, f] = r, h = t["checked-data-type"], p = h === "date" ? Gt(f) : f, b = h === "date" ? Gt(u) : u;
  return ht([c, p], [s, b], h) === d;
}
function oo(e, t) {
  const n = t[1];
  return e.attrs[n];
}
function uo(e, t, n) {
  const r = Object.keys(n)[0];
  return Ce(e, t, r);
}
function ao(e, t, n, r) {
  if (n)
    return oo(e, n);
  if (r)
    return uo(e, t, r);
}
function co(e, t, n) {
  var r, i;
  if (!Array.isArray(n.fields))
    return gs(e, t);
  const s = /* @__PURE__ */ new Map();
  for (const o of n.fields) {
    const u = Ce(e, t, o), a = (r = u == null ? void 0 : u["forward-identity"]) === null || r === void 0 ? void 0 : r[2];
    a && ln(u) && s.set(a, u);
  }
  if (!s.has("id")) {
    const o = Ce(e, t, "id"), u = (i = o == null ? void 0 : o["forward-identity"]) === null || i === void 0 ? void 0 : i[2];
    u && s.set(u, o);
  }
  return s;
}
function fo(e, { etype: t, pageInfo: n, dq: r, form: i }) {
  var s, o;
  const u = (s = i == null ? void 0 : i.$) === null || s === void 0 ? void 0 : s.order, a = Dr(i), d = lo(i);
  let c = Ps(e, r);
  const f = n == null ? void 0 : n["start-cursor"], h = ao(e, t, f, u);
  if (h && ((o = h == null ? void 0 : h["forward-identity"]) === null || o === void 0 ? void 0 : o[2]) !== "id") {
    const g = h["checked-data-type"] === "date", w = h.id;
    c = c.map(([T]) => {
      var m, S, q, O, x;
      let N = (x = (O = (q = (S = (m = e.eav.get(T)) === null || m === void 0 ? void 0 : m.get(w)) === null || S === void 0 ? void 0 : S.values()) === null || q === void 0 ? void 0 : q.next()) === null || O === void 0 ? void 0 : O.value) === null || x === void 0 ? void 0 : x[2];
      return g && (N = Gt(N)), [T, N];
    });
  }
  c.sort(d === "asc" ? function(w, T) {
    return ht(w, T, h == null ? void 0 : h["checked-data-type"]);
  } : function(w, T) {
    return ht(T, w, h == null ? void 0 : h["checked-data-type"]);
  });
  let p = {};
  const b = co(e, t, r);
  for (const g of c) {
    const [w] = g;
    if (p[w] || !a && f && h && so(f, h, d, g))
      continue;
    const T = _s(e, b, w);
    T && (p[w] = T);
  }
  return p;
}
function lo(e) {
  var t;
  const n = (t = e.$) === null || t === void 0 ? void 0 : t.order;
  return n && n[Object.keys(n)[0]] || "asc";
}
function Dr(e) {
  var t, n, r;
  const i = (t = e.$) === null || t === void 0 ? void 0 : t.offset, s = (n = e.$) === null || n === void 0 ? void 0 : n.before, o = (r = e.$) === null || r === void 0 ? void 0 : r.after;
  return !i && !s && !o;
}
function ho(e, { etype: t, level: n, form: r, join: i, pageInfo: s }) {
  var o, u, a, d, c;
  if (!Dr(r) && (!s || !s["start-cursor"]))
    return [];
  const f = Ys(eo(e, t, n, (o = r.$) === null || o === void 0 ? void 0 : o.where), i), h = to(_t, t, n), p = (u = r.$) === null || u === void 0 ? void 0 : u.fields, b = fo(e, {
    etype: t,
    pageInfo: s,
    form: r,
    dq: { where: f, find: h, fields: p }
  }), g = ((a = r.$) === null || a === void 0 ? void 0 : a.limit) || ((d = r.$) === null || d === void 0 ? void 0 : d.first) || ((c = r.$) === null || c === void 0 ? void 0 : c.last);
  if (g != null) {
    const w = Object.entries(b);
    return w.length <= g ? b : Object.fromEntries(w.slice(0, g));
  }
  return b;
}
function po(e, t) {
  try {
    return ho(e, t);
  } catch (n) {
    if (n instanceof Pe)
      return {};
    throw n;
  }
}
function xr(e, t) {
  const n = po(e, t);
  return ro(_t, e, t, n);
}
function yo(e) {
  const t = {};
  for (const [n, r] of Object.entries(e))
    t[n] = {
      startCursor: r["start-cursor"],
      endCursor: r["end-cursor"],
      hasNextPage: r["has-next-page?"],
      hasPreviousPage: r["has-previous-page?"]
    };
  return t;
}
function bo({ store: e, pageInfo: t, aggregate: n }, r) {
  const s = { data: Object.keys(r).reduce(function(u, a) {
    return n != null && n[a] || a === "$$ruleParams" || (u[a] = xr(e, {
      etype: a,
      form: r[a],
      level: 0,
      pageInfo: t == null ? void 0 : t[a]
    })), u;
  }, {}) };
  return t && (s.pageInfo = yo(t)), n && (s.aggregate = n), s;
}
function _o() {
  const t = {
    __etype: 1,
    __ops: 1,
    create: 1,
    update: 1,
    link: 1,
    unlink: 1,
    delete: 1,
    merge: 1,
    ruleParams: 1
  };
  return new Set(Object.keys(t));
}
const go = _o();
function Yt(e, t, n) {
  const r = {
    __etype: e,
    __ops: n
  };
  return new Proxy(r, {
    get: (i, s) => {
      if (s === "__ops")
        return n;
      if (s === "__etype")
        return e;
      if (go.has(s))
        return (o, u) => Yt(e, t, [
          ...n,
          u ? [s, e, t, o, u] : [s, e, t, o]
        ]);
    }
  });
}
function pt(e) {
  return e.startsWith("lookup__");
}
function Ur(e) {
  const [t, n, ...r] = e.split("__");
  return [n, JSON.parse(r.join("__"))];
}
function To(e) {
  return new Proxy({
    __etype: e
  }, {
    get(t, n) {
      if (n === "__etype")
        return e;
      const r = n;
      return pt(r) ? Yt(e, Ur(r), []) : Yt(e, r, []);
    }
  });
}
function yn() {
  return new Proxy({}, {
    get(e, t) {
      return To(t);
    }
  });
}
yn();
function wo(e) {
  return e.__ops;
}
function mo(e, t) {
  const { attrIdMap: n, refSwapAttrIds: r } = e, i = [];
  for (const o of t) {
    const u = n[o];
    if (u)
      i.push(u);
    else if (Array.isArray(o) && o.length == 2 && n[o[0]]) {
      const [a, d] = o;
      i.push([n[a], d]);
    } else
      i.push(o);
  }
  const [s] = t;
  if ((s === "add-triple" || s === "retract-triple") && r.has(t[2])) {
    const o = i[1];
    i[1] = i[3], i[3] = o;
  }
  return i;
}
function F(e, t, n) {
  return Object.values(e).find((r) => {
    const [i, s, o] = r["forward-identity"];
    return s === t && o === n;
  });
}
function Te(e, t, n) {
  return Object.values(e).find((r) => {
    const i = r["reverse-identity"];
    if (!i)
      return !1;
    const [s, o, u] = i;
    return o === t && u === n;
  });
}
function So(e) {
  if (Array.isArray(e))
    return e;
  const t = Object.entries(e);
  if (t.length !== 1)
    throw new Error("lookup must be an object with a single unique attr and value.");
  return t[0];
}
function Ht(e, t, n) {
  return n.indexOf(".") !== -1 && // attr names can have `.` in them, so use the attr we find with a `.`
  // before assuming it's a ref lookup.
  !F(e, t, n);
}
function Jt(e) {
  const [t, n, ...r] = e.split(".");
  if (r.length > 0 || n !== "id")
    throw new Error(`${e} is not a valid lookup attribute.`);
  return t;
}
function vo(e, t, n) {
  if (!Ht(e, t, n))
    return F(e, t, n);
  const r = Jt(n), i = F(e, t, r) || Te(e, t, r);
  if (i && i["value-type"] !== "ref")
    throw new Error(`${n} does not reference a valid link attribute.`);
  return i;
}
function Zt(e) {
  return typeof e == "string" && !pt(e) ? null : typeof e == "string" && pt(e) ? Ur(e) : So(e);
}
function J(e, t, n) {
  const r = Zt(n);
  if (r === null)
    return n;
  const [i, s] = r, o = vo(e, t, i);
  if (!o || !o["unique?"])
    throw new Error(`${i} is not a unique attribute.`);
  return [o.id, s];
}
function Lr(e, t, n, r) {
  const i = J(e, t, n);
  return Array.isArray(i) ? [[
    "add-triple",
    i,
    F(e, t, "id").id,
    i
  ]].concat(r) : r;
}
function Oo({ attrs: e }, [t, n, r]) {
  const i = Object.entries(r).flatMap(([s, o]) => {
    const u = Array.isArray(o) ? o : [o], a = F(e, t, s), d = Te(e, t, s);
    return u.map((c) => a ? [
      "add-triple",
      J(e, t, n),
      a.id,
      J(e, a["reverse-identity"][1], c)
    ] : [
      "add-triple",
      J(e, d["forward-identity"][1], c),
      d.id,
      J(e, t, n)
    ]);
  });
  return Lr(e, t, n, i);
}
function Eo({ attrs: e }, [t, n, r]) {
  const i = Object.entries(r).flatMap(([s, o]) => {
    const u = Array.isArray(o) ? o : [o], a = F(e, t, s), d = Te(e, t, s);
    return u.map((c) => a ? [
      "retract-triple",
      J(e, t, n),
      a.id,
      J(e, a["reverse-identity"][1], c)
    ] : [
      "retract-triple",
      J(e, d["forward-identity"][1], c),
      d.id,
      J(e, t, n)
    ]);
  });
  return Lr(e, t, n, i);
}
function Ao(e, t, n) {
  if (Array.isArray(n)) {
    const [r, i] = n;
    for (const s of e || []) {
      const o = s == null ? void 0 : s.aev.get(r);
      if (o) {
        for (const [u, a, d] of H(o, 2))
          if (d === i)
            return !0;
      }
    }
  } else
    for (const r of e || []) {
      const i = r == null ? void 0 : r.eav.get(n);
      if (i) {
        for (const s of i.keys())
          if (r.attrs[s]["forward-identity"][1] == t)
            return !0;
      }
    }
  return !1;
}
function Fr({ stores: e, attrs: t }, [n, r, i, s]) {
  return (s == null ? void 0 : s.upsert) === !1 ? { mode: "update" } : (s == null ? void 0 : s.upsert) === !0 ? null : Ao(e, n, r) ? { mode: "update" } : null;
}
function Co(e, t) {
  const { stores: n, attrs: r } = e, [i, s, o, u] = t, a = dn(o), d = J(r, i, s);
  return [["id", d]].concat(Object.entries(a)).map(([f, h]) => {
    const p = F(r, i, f);
    return p["checked-data-type"] === "date" && e.useDateObjects && (h = yt(h)), ["add-triple", d, p.id, h, { mode: "create" }];
  });
}
function jo(e, t) {
  const { stores: n, attrs: r } = e, [i, s, o, u] = t, a = dn(o), d = J(r, i, s), c = Fr(e, [i, d, o, u]);
  return [["id", d]].concat(Object.entries(a)).map(([h, p]) => {
    const b = F(r, i, h);
    return b["checked-data-type"] === "date" && e.useDateObjects && (p = yt(p)), [
      "add-triple",
      d,
      b.id,
      p,
      ...c ? [c] : []
    ];
  });
}
function Po({ attrs: e }, [t, n]) {
  return [["delete-entity", J(e, t, n), t]];
}
function ko(e, t) {
  const { stores: n, attrs: r } = e, [i, s, o, u] = t, a = dn(o), d = J(r, i, s), c = Fr(e, [i, d, o, u]), f = Object.entries(a).map(([p, b]) => {
    const g = F(r, i, p);
    return [
      "deep-merge-triple",
      d,
      g.id,
      b,
      ...c ? [c] : []
    ];
  });
  return [[
    "add-triple",
    d,
    F(r, i, "id").id,
    d,
    ...c ? [c] : []
  ]].concat(f);
}
function Io({ attrs: e }, [t, n, r]) {
  return [["rule-params", J(e, t, n), t, r]];
}
function Ro(e) {
  const [t, n, r, i, s] = e;
  if (!i)
    return e;
  const o = Object.assign({}, i);
  return delete o.id, [t, n, r, o, ...s ? [s] : []];
}
function Mo(e, t) {
  const [n, ...r] = Ro(t);
  switch (n) {
    case "merge":
      return ko(e, r);
    case "create":
      return Co(e, r);
    case "update":
      return jo(e, r);
    case "link":
      return Oo(e, r);
    case "unlink":
      return Eo(e, r);
    case "delete":
      return Po(e, r);
    case "ruleParams":
      return Io(e, r);
    default:
      throw new Error(`unsupported action ${n}`);
  }
}
function $o(e) {
  switch (e) {
    case "string":
    case "date":
    case "boolean":
    case "number":
      return e;
    default:
      return;
  }
}
function Do(e, t, n) {
  var r, i;
  const s = (i = (r = e.entities[t]) === null || r === void 0 ? void 0 : r.attrs) === null || i === void 0 ? void 0 : i[n];
  if (n === "id")
    return null;
  if (!s)
    throw new Error(`${t}.${n} does not exist in your schema`);
  const { unique: o, indexed: u } = s == null ? void 0 : s.config, a = $o(s == null ? void 0 : s.valueType);
  return {
    "index?": u,
    "unique?": o,
    "checked-data-type": a
  };
}
function st(e, t, n, r) {
  const i = e ? Do(e, t, n) : null, s = V(), u = [V(), t, n];
  return Object.assign(Object.assign({ id: s, "forward-identity": u, "value-type": "blob", cardinality: "one", "unique?": !1, "index?": !1, isUnsynced: !0 }, i || {}), r || {});
}
function xo(e, t, n) {
  return Object.values(e.links).find((i) => i.forward.on === t && i.forward.label === n || i.reverse.on === t && i.reverse.label === n);
}
function Uo(e, t, n) {
  const r = xo(e, t, n);
  if (!r)
    throw new Error(`Couldn't find the link ${t}.${n} in your schema`);
  const { forward: i, reverse: s } = r;
  return {
    "forward-identity": [V(), i.on, i.label],
    "reverse-identity": [V(), s.on, s.label],
    cardinality: i.has === "one" ? "one" : "many",
    "unique?": s.has === "one",
    "on-delete": i.onDelete,
    "on-delete-reverse": s.onDelete
  };
}
function Gn(e, t, n, r) {
  const i = e ? Uo(e, t, n) : null, s = V(), o = [V(), t, n], u = [V(), n, t];
  return Object.assign(Object.assign({ id: s, "forward-identity": o, "reverse-identity": u, "value-type": "ref", cardinality: "many", "unique?": !1, "index?": !1, isUnsynced: !0 }, i || {}), r || {});
}
const Lo = /* @__PURE__ */ new Set(["create", "update", "merge", "link", "unlink"]), Fo = /* @__PURE__ */ new Set(["link", "unlink"]), No = /* @__PURE__ */ new Set(["create", "update", "merge"]), Vo = /* @__PURE__ */ new Set([
  "link",
  "unlink",
  "create",
  "update",
  "merge",
  "delete",
  "ruleParams"
]), Xt = { "unique?": !0, "index?": !0 }, qo = Object.assign(Object.assign({}, Xt), { cardinality: "one" });
function Wo(e) {
  const t = [], [n, r, i, s] = e;
  if (!Vo.has(n))
    return t;
  const o = Zt(i);
  if (o && t.push({ etype: r, lookupPair: o }), n === "link")
    for (const [u, a] of Object.entries(s)) {
      const d = Array.isArray(a) ? a : [a];
      for (const c of d) {
        const f = Zt(c);
        f && t.push({
          etype: r,
          lookupPair: f,
          linkLabel: u
        });
      }
    }
  return t;
}
function Ko({ attrs: e, schema: t }, n) {
  var r, i;
  const [s, o, u] = [/* @__PURE__ */ new Set(), Object.assign({}, e), []];
  function a(f) {
    o[f.id] = f, u.push(["add-attr", f]), s.add(f.id);
  }
  function d(f) {
    f != null && f.isUnsynced && !s.has(f.id) && (u.push(["add-attr", f]), s.add(f.id));
  }
  function c(f, h) {
    const p = F(o, f, h), b = Te(o, f, h);
    d(p), d(b), !p && !b && a(Gn(t, f, h, qo));
  }
  for (const f of n)
    for (const { etype: h, lookupPair: p, linkLabel: b } of Wo(f)) {
      const g = p[0];
      if (b) {
        c(h, b);
        const w = F(o, h, b), T = Te(o, h, b);
        d(w), d(T);
        const m = ((r = w == null ? void 0 : w["reverse-identity"]) === null || r === void 0 ? void 0 : r[1]) || ((i = T == null ? void 0 : T["forward-identity"]) === null || i === void 0 ? void 0 : i[1]) || b;
        if (Ht(o, m, g))
          c(m, Jt(g));
        else {
          const S = F(o, m, g);
          S || a(st(t, m, g, Xt)), d(S);
        }
      } else if (Ht(o, h, g))
        c(h, Jt(g));
      else {
        const w = F(o, h, g);
        w || a(st(t, h, g, Xt)), d(w);
      }
    }
  for (const f of n) {
    const [h, p, b, g] = f;
    if (Lo.has(h)) {
      const w = F(o, p, "id");
      d(w), w || a(st(t, p, "id", { "unique?": !0 }));
      for (const T of Object.keys(g)) {
        const m = F(o, p, T);
        if (d(m), No.has(h) && (m || a(st(t, p, T, T === "id" ? { "unique?": !0 } : null))), Fo.has(h)) {
          const S = Te(o, p, T);
          !m && !S && a(Gn(t, p, T)), d(S);
        }
      }
    }
  }
  return [o, u];
}
function zo(e, t) {
  const r = (Array.isArray(t) ? t : [t]).flatMap((a) => wo(a)), [i, s] = Ko(e, r), o = Object.assign(Object.assign({}, e), { attrs: i }), u = r.flatMap((a) => Mo(o, a));
  return [...s, ...u];
}
var Yn = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
class Nr {
  constructor(t) {
    this.dbName = t, this._storeName = "kv", this._dbPromise = this._init();
  }
  _init() {
    return new Promise((t, n) => {
      const r = indexedDB.open(this.dbName, 1);
      r.onerror = (i) => {
        n(i);
      }, r.onsuccess = (i) => {
        const s = i.target;
        t(s.result);
      }, r.onupgradeneeded = (i) => {
        i.target.result.createObjectStore(this._storeName);
      };
    });
  }
  getItem(t) {
    return Yn(this, void 0, void 0, function* () {
      const n = yield this._dbPromise;
      return new Promise((r, i) => {
        const u = n.transaction([this._storeName], "readonly").objectStore(this._storeName).get(t);
        u.onerror = (a) => {
          i(a);
        }, u.onsuccess = (a) => {
          u.result ? r(u.result) : r(null);
        };
      });
    });
  }
  setItem(t, n) {
    return Yn(this, void 0, void 0, function* () {
      const r = yield this._dbPromise;
      return new Promise((i, s) => {
        const a = r.transaction([this._storeName], "readwrite").objectStore(this._storeName).put(n, t);
        a.onerror = (d) => {
          s(d);
        }, a.onsuccess = (d) => {
          i();
        };
      });
    });
  }
}
var Bo = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
class Vr {
  static getIsOnline() {
    return Bo(this, void 0, void 0, function* () {
      return navigator.onLine;
    });
  }
  static listen(t) {
    const n = () => {
      t(!0);
    }, r = () => {
      t(!1);
    };
    return addEventListener("online", n), addEventListener("offline", r), () => {
      removeEventListener("online", n), removeEventListener("offline", r);
    };
  }
}
class ke extends Error {
  constructor(t, n) {
    super(t), this.hint = n;
    const r = new.target.prototype;
    Object.setPrototypeOf && Object.setPrototypeOf(this, r), Error.captureStackTrace && Error.captureStackTrace(this, ke), this.name = "InstantError";
  }
  get [Symbol.toStringTag]() {
    return "InstantError";
  }
}
var Qo = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
class gt extends ke {
  constructor(t) {
    var n;
    const r = ((n = t.body) === null || n === void 0 ? void 0 : n.message) || `API Error (${t.status})`;
    super(r, t.body.hint);
    const i = new.target.prototype;
    Object.setPrototypeOf && Object.setPrototypeOf(this, i), Error.captureStackTrace && Error.captureStackTrace(this, gt), this.name = "InstantAPIError", this.status = t.status, this.body = t.body;
  }
  get [Symbol.toStringTag]() {
    return "InstantAPIError";
  }
}
function ue(e, t) {
  return Qo(this, void 0, void 0, function* () {
    const n = yield fetch(e, t), r = yield n.json();
    return n.status === 200 ? Promise.resolve(r) : Promise.reject(new gt({ status: n.status, body: r }));
  });
}
var Re = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
function Go({ apiURI: e, appId: t, email: n }) {
  return ue(`${e}/runtime/auth/send_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ "app-id": t, email: n })
  });
}
function Yo(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n, email: r, code: i, refreshToken: s }) {
    return yield ue(`${t}/runtime/auth/verify_magic_code`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(Object.assign({
        "app-id": n,
        email: r,
        code: i
      }, s ? { "refresh-token": s } : {}))
    });
  });
}
function Ho(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n, refreshToken: r }) {
    return yield ue(`${t}/runtime/auth/verify_refresh_token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        "app-id": n,
        "refresh-token": r
      })
    });
  });
}
function Jo(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n }) {
    return yield ue(`${t}/runtime/auth/sign_in_guest`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        "app-id": n
      })
    });
  });
}
function Hn(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n, code: r, codeVerifier: i, refreshToken: s }) {
    return yield ue(`${t}/runtime/oauth/token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        app_id: n,
        code: r,
        code_verifier: i,
        refresh_token: s
      })
    });
  });
}
function Zo(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n, nonce: r, idToken: i, clientName: s, refreshToken: o }) {
    return yield ue(`${t}/runtime/oauth/id_token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        app_id: n,
        nonce: r,
        id_token: i,
        client_name: s,
        refresh_token: o
      })
    });
  });
}
function Xo(e) {
  return Re(this, arguments, void 0, function* ({ apiURI: t, appId: n, refreshToken: r }) {
    return yield ue(`${t}/runtime/signout`, {
      method: "POST",
      headers: {
        "content-type": "application/json"
      },
      body: JSON.stringify({
        app_id: n,
        refresh_token: r
      })
    });
  });
}
var ze = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
function eu(e) {
  return ze(this, arguments, void 0, function* ({ apiURI: t, appId: n, path: r, file: i, refreshToken: s, contentType: o, contentDisposition: u }) {
    const a = {
      app_id: n,
      path: r,
      authorization: `Bearer ${s}`,
      "content-type": o || i.type
    };
    return u && (a["content-disposition"] = u), yield ue(`${t}/storage/upload`, {
      method: "PUT",
      headers: a,
      body: i
    });
  });
}
function tu(e) {
  return ze(this, arguments, void 0, function* ({ apiURI: t, appId: n, path: r, refreshToken: i }) {
    const { data: s } = yield ue(`${t}/storage/files?app_id=${n}&filename=${encodeURIComponent(r)}`, {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      }
    });
    return s;
  });
}
function nu(e) {
  return ze(this, arguments, void 0, function* ({ apiURI: t, appId: n, fileName: r, refreshToken: i, metadata: s = {} }) {
    const { data: o } = yield ue(`${t}/storage/signed-upload-url`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      },
      body: JSON.stringify({
        app_id: n,
        filename: r
      })
    });
    return o;
  });
}
function ru(e, t) {
  return ze(this, void 0, void 0, function* () {
    return (yield fetch(e, {
      method: "PUT",
      body: t,
      headers: {
        "Content-Type": t.type
      }
    })).ok;
  });
}
function iu(e) {
  return ze(this, arguments, void 0, function* ({ apiURI: t, appId: n, path: r, refreshToken: i }) {
    const { data: s } = yield ue(`${t}/storage/signed-download-url?app_id=${n}&filename=${encodeURIComponent(r)}`, {
      method: "GET",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      }
    });
    return s;
  });
}
let bn = !1, qr = !1, Wr = !1;
typeof window < "u" && typeof window.localStorage < "u" && (bn = !!window.localStorage.getItem("devBackend"), qr = !!window.localStorage.getItem("__instantLogging"), Wr = !!window.localStorage.getItem("__devtoolLocalDash"));
function Jn(e, t) {
  if (!t)
    return e;
  const n = {};
  return t.forEach((r) => {
    n[r] = e[r];
  }), n;
}
function su(e, t, n) {
  var r, i;
  const s = {
    peers: {}
  };
  if (t && "user" in t ? t.user : !0) {
    const u = Jn((r = e.user) !== null && r !== void 0 ? r : {}, t == null ? void 0 : t.keys);
    s.user = Object.assign(Object.assign({}, u), { peerId: n });
  }
  for (const u of Object.keys((i = e.peers) !== null && i !== void 0 ? i : {})) {
    const a = (t == null ? void 0 : t.peers) === void 0, d = Array.isArray(t == null ? void 0 : t.peers) && (t == null ? void 0 : t.peers.includes(u));
    if (a || d) {
      const c = Jn(e.peers[u], t == null ? void 0 : t.keys);
      s.peers[u] = Object.assign(Object.assign({}, c), { peerId: u });
    }
  }
  return s;
}
function ou(e, t) {
  if (e.isLoading !== t.isLoading || e.error !== t.error || (e.user || t.user) && (!e.user || !t.user || !Ln(e.user, t.user)) || !Sr(e.peers, t.peers))
    return !0;
  for (const r of Object.keys(e.peers))
    if (!Ln(e.peers[r], t.peers[r]))
      return !0;
  return !1;
}
class Zn {
  constructor() {
    this.promise = new Promise((t, n) => {
      this._resolve = t, this._reject = n;
    });
  }
  resolve(t) {
    this._resolve(t);
  }
  reject(t) {
    this._reject(t);
  }
}
var Ue = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
function uu(e, t) {
  typeof requestIdleCallback > "u" ? e() : requestIdleCallback(e, { timeout: t });
}
class Xn {
  constructor(t, n, r, i, s = (d) => d, o = (d) => d, u = 100, a = 1e3) {
    this._subs = [], this._nextSave = null, this._persister = t, this._key = n, this._onMerge = i, this._loadedCbs = [], this._isLoading = !0, this.currentValue = r, this.serialize = s, this.parse = o, this._saveThrottleMs = u, this._pendingSaveCbs = [], this._version = 0, this._idleCallbackMaxWaitMs = a, this._load();
  }
  _getFromStorage() {
    return Ue(this, void 0, void 0, function* () {
      try {
        return this.parse(yield this._persister.getItem(this._key));
      } catch (t) {
        return console.error(`Unable to read from storage for key=${this._key}`, t), null;
      }
    });
  }
  _load() {
    return Ue(this, void 0, void 0, function* () {
      const t = yield this._getFromStorage();
      this._isLoading = !1, this._onMerge(t, this.currentValue);
      for (const n of this._loadedCbs)
        n();
    });
  }
  waitForLoaded() {
    return Ue(this, void 0, void 0, function* () {
      if (!this._isLoading)
        return;
      yield new Promise((n) => {
        this._loadedCbs.push(n);
      });
    });
  }
  isLoading() {
    return this._isLoading;
  }
  version() {
    return this._version;
  }
  waitForSync() {
    return Ue(this, void 0, void 0, function* () {
      if (!this._nextSave)
        return;
      yield new Promise((n) => {
        this._pendingSaveCbs.push(n);
      });
    });
  }
  _writeToStorage() {
    this._persister.setItem(this._key, this.serialize(this.currentValue));
    for (const t of this._pendingSaveCbs)
      t();
    this._pendingSaveCbs.length = 0;
  }
  flush() {
    return Ue(this, void 0, void 0, function* () {
      this._nextSave && (clearTimeout(this._nextSave), this._writeToStorage());
    });
  }
  _enqueuePersist(t) {
    if (this._nextSave) {
      t && this._pendingSaveCbs.push(t);
      return;
    }
    this._nextSave = setTimeout(() => {
      uu(() => {
        this._nextSave = null, this._writeToStorage();
      }, this._idleCallbackMaxWaitMs);
    }, this._saveThrottleMs);
  }
  set(t, n) {
    this._version++, this.currentValue = t(this.currentValue), this._isLoading ? this._loadedCbs.push(() => this._enqueuePersist(n)) : this._enqueuePersist(n);
    for (const r of this._subs)
      r(this.currentValue);
  }
  subscribe(t) {
    return this._subs.push(t), t(this.currentValue), () => {
      this._subs = this._subs.filter((n) => n !== t);
    };
  }
}
function Kr(e, t = []) {
  e.forEach((n) => {
    const { data: r } = n, { "datalog-result": i } = r, { "join-rows": s } = i;
    for (const o of s)
      for (const u of o)
        t.push(u);
    Kr(n["child-nodes"], t);
  });
}
function er(e) {
  const t = [];
  return Kr(e, t), t;
}
function tr(e) {
  return Object.values(e.links).reduce((t, n) => {
    var r, i, s, o;
    return (r = t[s = n.forward.on]) !== null && r !== void 0 || (t[s] = {}), t[n.forward.on][n.forward.label] = {
      isForward: !0,
      isSingular: n.forward.has === "one",
      link: n
    }, (i = t[o = n.reverse.on]) !== null && i !== void 0 || (t[o] = {}), t[n.reverse.on][n.reverse.label] = {
      isForward: !1,
      isSingular: n.reverse.has === "one",
      link: n
    }, t;
  }, {});
}
const zr = "v0.22.34";
function au(e) {
  return {
    info: e ? console.info.bind(console) : () => {
    },
    debug: e ? console.debug.bind(console) : () => {
    },
    error: e ? console.error.bind(console) : () => {
    }
  };
}
class le {
  constructor(t, n, r, i = { indexed: !1, unique: !1 }) {
    this.valueType = t, this.required = n, this.isIndexed = r, this.config = i, this.metadata = {};
  }
  /**
   * @deprecated Only use this temporarily for attributes that you want
   * to treat as required in frontend code but can’t yet mark as required
   * and enforced for backend
   */
  clientRequired() {
    return new le(this.valueType, !1, this.isIndexed, this.config);
  }
  optional() {
    return new le(this.valueType, !1, this.isIndexed, this.config);
  }
  unique() {
    return new le(this.valueType, this.required, this.isIndexed, Object.assign(Object.assign({}, this.config), { unique: !0 }));
  }
  indexed() {
    return new le(this.valueType, this.required, !0, Object.assign(Object.assign({}, this.config), { indexed: !0 }));
  }
}
class I extends Error {
  constructor(t, n) {
    const r = n ? `At path '${n}': ${t}` : t;
    super(r), this.name = "QueryValidationError";
  }
}
const nr = [
  "where",
  "order",
  "limit",
  "last",
  "first",
  "offset",
  "after",
  "before",
  "fields",
  "aggregate"
], cu = (e) => e.valueType || "unknown", Br = (e, t, n = !1) => {
  if (n || e == null)
    return !0;
  switch (t) {
    case "string":
      return typeof e == "string";
    case "number":
      return typeof e == "number" && !isNaN(e);
    case "boolean":
      return typeof e == "boolean";
    case "date":
      return e instanceof Date || typeof e == "string" || typeof e == "number";
    default:
      return !0;
  }
}, du = (e, t, n, r, i, s, o) => {
  const u = s.valueType === "json", a = (d, c, f) => {
    if (!Br(f, c, u))
      throw new I(`Invalid value for operator '${d}' on attribute '${r}' in entity '${i}'. Expected ${c}, but received: ${typeof f}`, o);
  };
  switch (e) {
    case "in":
    case "$in":
      if (!Array.isArray(t))
        throw new I(`Operator '${e}' for attribute '${r}' in entity '${i}' must be an array, but received: ${typeof t}`, o);
      for (const d of t)
        a(e, n, d);
      break;
    case "$not":
    case "$ne":
    case "$gt":
    case "$lt":
    case "$gte":
    case "$lte":
      a(e, n, t);
      break;
    case "$like":
    case "$ilike":
      if (a(e, "string", t), e === "$ilike" && !s.isIndexed)
        throw new I(`Operator '${e}' can only be used with indexed attributes, but '${r}' in entity '${i}' is not indexed`, o);
      break;
    case "$isNull":
      a(e, "boolean", t);
      break;
    default:
      throw new I(`Unknown operator '${e}' for attribute '${r}' in entity '${i}'`, o);
  }
}, Ae = (e, t, n, r, i) => {
  const s = cu(n), o = n.valueType === "json";
  if (typeof e == "object" && e !== null && !Array.isArray(e)) {
    if (o)
      return;
    const a = e;
    for (const [d, c] of Object.entries(a))
      du(d, c, s, t, r, n, `${i}.${d}`);
  } else if (!Br(e, s, o))
    throw new I(`Invalid value for attribute '${t}' in entity '${r}'. Expected ${s}, but received: ${typeof e}`, i);
}, fu = (e, t, n, r, i) => {
  const s = e.split(".");
  if (s.length < 2)
    throw new I(`Invalid dot notation path '${e}'. Must contain at least one dot.`, i);
  let o = n;
  for (let c = 0; c < s.length - 1; c++) {
    const f = s[c], h = r.entities[o];
    if (!h)
      throw new I(`Entity '${o}' does not exist in schema while traversing dot notation path '${e}'.`, i);
    const p = h.links[f];
    if (!p) {
      const b = Object.keys(h.links);
      throw new I(`Link '${f}' does not exist on entity '${o}' in dot notation path '${e}'. Available links: ${b.length > 0 ? b.join(", ") : "none"}`, i);
    }
    o = p.entityName;
  }
  const u = s[s.length - 1], a = r.entities[o];
  if (!a)
    throw new I(`Target entity '${o}' does not exist in schema for dot notation path '${e}'.`, i);
  if (u === "id") {
    if (typeof t == "string" && !je(t))
      throw new I(`Invalid value for id field in entity '${o}'. Expected a UUID, but received: ${t}`, i);
    Ae(t, e, new le("string", !1, !0), n, i);
    return;
  }
  const d = a.attrs[u];
  if (Object.keys(a.links).includes(u)) {
    if (typeof t == "string" && !je(t))
      throw new I(`Invalid value for link '${u}' in entity '${o}'. Expected a UUID, but received: ${t}`, i);
    Ae(t, e, new le("string", !1, !0), n, i);
    return;
  }
  if (!d) {
    const c = Object.keys(a.attrs);
    throw new I(`Attribute '${u}' does not exist on entity '${o}' in dot notation path '${e}'. Available attributes: ${c.length > 0 ? c.join(", ") + ", id" : "id"}`, i);
  }
  Ae(t, e, d, n, i);
}, Qr = (e, t, n, r) => {
  for (const [i, s] of Object.entries(e)) {
    if (i === "or" || i === "and") {
      if (Array.isArray(s))
        for (const d of s)
          typeof d == "object" && d !== null && Qr(d, t, n, `${r}.${i}[${d}]`);
      continue;
    }
    if (i === "id") {
      Ae(s, "id", new le("string", !1, !0), t, `${r}.id`);
      continue;
    }
    if (i.includes(".")) {
      fu(i, s, t, n, `${r}.${i}`);
      continue;
    }
    const o = n.entities[t];
    if (!o)
      continue;
    const u = o.attrs[i], a = o.links[i];
    if (!u && !a) {
      const d = Object.keys(o.attrs), c = Object.keys(o.links);
      throw new I(`Attribute or link '${i}' does not exist on entity '${t}'. Available attributes: ${d.length > 0 ? d.join(", ") : "none"}. Available links: ${c.length > 0 ? c.join(", ") : "none"}`, `${r}.${i}`);
    }
    if (u)
      Ae(s, i, u, t, `${r}.${i}`);
    else if (a) {
      if (typeof s == "string" && !je(s))
        throw new I(`Invalid value for link '${i}' in entity '${t}'. Expected a UUID, but received: ${s}`, `${r}.${i}`);
      const d = new le("string", !1, !0);
      Ae(s, i, d, t, `${r}.${i}`);
    }
  }
}, lu = (e, t, n, r, i = 0) => {
  for (const o of Object.keys(e))
    if (!nr.includes(o))
      throw new I(`Invalid query parameter '${o}' in $ object. Valid parameters are: ${nr.join(", ")}. Found: ${o}`, r);
  const s = [
    "limit",
    "offset",
    "before",
    "after",
    "first",
    "last"
  ];
  for (const o of s)
    if (e[o] !== void 0 && i > 0)
      throw new I(`'${o}' can only be used on top-level namespaces. It cannot be used in nested queries.`, r);
  if (e.where && n) {
    if (typeof e.where != "object" || e.where === null)
      throw new I(`'where' clause must be an object in entity '${t}', but received: ${typeof e.where}`, r ? `${r}.where` : void 0);
    Qr(e.where, t, n, r ? `${r}.where` : "where");
  }
}, Gr = (e, t, n, r, i = 0) => {
  var s;
  if (!e || typeof e != "object")
    throw new I(`Query part for entity '${t}' must be an object, but received: ${typeof e}`, r);
  for (const o of Object.keys(e))
    if (o !== "$") {
      if (n && !(o in n.entities[t].links)) {
        const a = Object.keys(n.entities[t].links);
        throw new I(`Link '${o}' does not exist on entity '${t}'. Available links: ${a.length > 0 ? a.join(", ") : "none"}`, `${r}.${o}`);
      }
      const u = e[o];
      if (typeof u == "object" && u !== null) {
        const a = (s = n == null ? void 0 : n.entities[t].links[o]) === null || s === void 0 ? void 0 : s.entityName;
        a && Gr(u, a, n, `${r}.${o}`, i + 1);
      }
    } else {
      const u = e[o];
      if (typeof u != "object" || u === null)
        throw new I(`Query parameter '$' must be an object in entity '${t}', but received: ${typeof u}`, `${r}.$`);
      lu(u, t, n, `${r}.$`, i);
    }
}, rr = (e, t) => {
  if (typeof e != "object" || e === null)
    throw new I(`Query must be an object, but received: ${typeof e}${e === null ? " (null)" : ""}`);
  if (Array.isArray(e))
    throw new I(`Query must be an object, but received: ${typeof e}`);
  const n = e;
  for (const r of Object.keys(n)) {
    if (Array.isArray(e[r]))
      throw new I(`Query keys must be strings, but found key of type: ${typeof r}`, r);
    if (typeof r != "string")
      throw new I(`Query keys must be strings, but found key of type: ${typeof r}`, r);
    if (r !== "$$ruleParams") {
      if (t && !t.entities[r]) {
        const i = Object.keys(t.entities);
        throw new I(`Entity '${r}' does not exist in schema. Available entities: ${i.length > 0 ? i.join(", ") : "none"}`, r);
      }
      Gr(n[r], r, t, r, 0);
    }
  }
}, ir = (e) => typeof e != "string" ? !1 : pt(e) ? !0 : je(e);
class ne extends Error {
  constructor(t) {
    super(t), this.name = "TransactionValidationError";
  }
}
const Yr = (e) => e.length > 0 ? e.join(", ") : "none", hu = (e, t) => new ne(`Entity '${e}' does not exist in schema. Available entities: ${Yr(t)}`), sr = {
  string: (e) => typeof e == "string",
  number: (e) => typeof e == "number" && !isNaN(e),
  boolean: (e) => typeof e == "boolean",
  date: (e) => e instanceof Date || typeof e == "string" || typeof e == "number",
  json: () => !0
}, pu = (e, t) => {
  var n, r;
  return e == null ? !0 : (r = (n = sr[t.valueType]) === null || n === void 0 ? void 0 : n.call(sr, e)) !== null && r !== void 0 ? r : !1;
}, Hr = (e, t) => {
  const n = t.entities[e];
  if (!n)
    throw hu(e, Object.keys(t.entities));
  return n;
}, kt = (e, t, n) => {
  const r = Hr(e, n);
  if (typeof t != "object" || t === null)
    throw new ne(`Arguments for data operation on entity '${e}' must be an object, but received: ${typeof t}`);
  for (const [i, s] of Object.entries(t)) {
    if (i === "id")
      continue;
    const o = r.attrs[i];
    if (o && !pu(s, o))
      throw new ne(`Invalid value for attribute '${i}' in entity '${e}'. Expected ${o.valueType}, but received: ${typeof s}`);
  }
}, or = (e, t, n) => {
  const r = Hr(e, n);
  if (typeof t != "object" || t === null)
    throw new ne(`Arguments for link operation on entity '${e}' must be an object, but received: ${typeof t}`);
  for (const [i, s] of Object.entries(t)) {
    if (!r.links[i]) {
      const u = Object.keys(r.links);
      throw new ne(`Link '${i}' does not exist on entity '${e}'. Available links: ${Yr(u)}`);
    }
    if (s != null) {
      if (Array.isArray(s)) {
        for (const u of s)
          if (!ir(u))
            throw new ne(`Invalid entity ID in link '${i}' for entity '${e}'. Expected a UUID or a lookup, but received: ${u}`);
      } else if (!ir(s))
        throw new ne(`Invalid UUID in link '${i}' for entity '${e}'. Expected a UUID, but received: ${s}`);
    }
  }
}, yu = {
  create: kt,
  update: kt,
  merge: kt,
  link: or,
  unlink: or,
  delete: () => {
  }
}, bu = (e, t) => {
  if (!t)
    return;
  const [n, r, i, s] = e;
  if (!Array.isArray(i) && !je(i))
    throw new ne(`Invalid id for entity '${r}'. Expected a UUID, but received: ${i}`);
  if (typeof r != "string")
    throw new ne(`Entity name must be a string, but received: ${typeof r}`);
  const o = yu[n];
  o && s !== void 0 && o(r, s, t);
}, _u = (e, t) => {
  const n = Array.isArray(e) ? e : [e];
  for (const r of n) {
    if (!r || typeof r != "object")
      throw new ne(`Transaction chunk must be an object, but received: ${typeof r}`);
    if (!Array.isArray(r.__ops))
      throw new ne(`Transaction chunk must have __ops array, but received: ${typeof r.__ops}`);
    for (const i of r.__ops) {
      if (!Array.isArray(i))
        throw new ne(`Transaction operation must be an array, but received: ${typeof i}`);
      bu(i, t);
    }
  }
};
var ur = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
};
let Jr = 0;
class ar {
  constructor(t) {
    this.type = "ws", this.id = `${this.type}_${Jr++}`, this.conn = new WebSocket(t), this.conn.onopen = (n) => {
      this.onopen && this.onopen({ target: this });
    }, this.conn.onmessage = (n) => {
      this.onmessage && this.onmessage({
        target: this,
        message: JSON.parse(n.data.toString())
      });
    }, this.conn.onclose = (n) => {
      this.onclose && this.onclose({ target: this });
    }, this.conn.onerror = (n) => {
      this.onerror && this.onerror({ target: this });
    };
  }
  close() {
    this.conn.close();
  }
  isOpen() {
    return this.conn.readyState === WebSocket.OPEN;
  }
  isConnecting() {
    return this.conn.readyState === WebSocket.CONNECTING;
  }
  send(t) {
    return this.conn.send(JSON.stringify(t));
  }
}
class gu {
  constructor(t, n) {
    this.type = "sse", this.initParams = null, this.sendQueue = [], this.closeFired = !1, this.sseInitTimeout = null, this.id = `${this.type}_${Jr++}`, this.url = n, this.ES = t, this.conn = new t(n), this.sseInitTimeout = setTimeout(() => {
      this.initParams || this.handleError();
    }, 1e4), this.conn.onmessage = (r) => {
      const i = JSON.parse(r.data);
      if (i.op === "sse-init") {
        this.initParams = {
          machineId: i["machine-id"],
          sessionId: i["session-id"],
          sseToken: i["sse-token"]
        }, this.onopen && this.onopen({ target: this }), clearTimeout(this.sseInitTimeout);
        return;
      }
      this.onmessage && this.onmessage({
        target: this,
        message: i
      });
    }, this.conn.onerror = (r) => {
      this.handleError();
    };
  }
  // Runs the onerror and closes the connection
  handleError() {
    try {
      this.onerror && this.onerror({ target: this });
    } finally {
      this.handleClose();
    }
  }
  handleClose() {
    this.conn.close(), this.onclose && !this.closeFired && (this.closeFired = !0, this.onclose({ target: this }));
  }
  postMessages(t) {
    return ur(this, void 0, void 0, function* () {
      try {
        (yield fetch(this.url, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            machine_id: this.initParams.machineId,
            session_id: this.initParams.sessionId,
            sse_token: this.initParams.sseToken,
            messages: t
          })
        })).ok || this.handleError();
      } catch {
        this.handleError();
      }
    });
  }
  flushQueue() {
    return ur(this, void 0, void 0, function* () {
      if (this.sendPromise || !this.sendQueue.length)
        return;
      const t = this.sendQueue;
      this.sendQueue = [];
      const n = this.postMessages(t);
      this.sendPromise = n, n.then(() => {
        this.sendPromise = null, this.flushQueue();
      });
    });
  }
  send(t) {
    if (!this.isOpen() || !this.initParams)
      throw this.isConnecting() ? new Error("Failed to execute 'send' on 'EventSource': Still in CONNECTING state.") : this.conn.readyState === this.ES.CLOSED ? new Error("EventSource is already in CLOSING or CLOSED state.") : new Error("EventSource is in invalid state.");
    this.sendQueue.push(t), this.flushQueue();
  }
  isOpen() {
    return this.conn.readyState === this.ES.OPEN && this.initParams !== null;
  }
  isConnecting() {
    return this.conn.readyState === this.ES.CONNECTING || this.conn.readyState === this.ES.OPEN && this.initParams === null;
  }
  close() {
    this.handleClose();
  }
}
var D = function(e, t, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function u(c) {
      try {
        d(r.next(c));
      } catch (f) {
        o(f);
      }
    }
    function a(c) {
      try {
        d(r.throw(c));
      } catch (f) {
        o(f);
      }
    }
    function d(c) {
      c.done ? s(c.value) : i(c.value).then(u, a);
    }
    d((r = r.apply(e, t || [])).next());
  });
}, Tu = function(e, t) {
  var n = {};
  for (var r in e) Object.prototype.hasOwnProperty.call(e, r) && t.indexOf(r) < 0 && (n[r] = e[r]);
  if (e != null && typeof Object.getOwnPropertySymbols == "function")
    for (var i = 0, r = Object.getOwnPropertySymbols(e); i < r.length; i++)
      t.indexOf(r[i]) < 0 && Object.prototype.propertyIsEnumerable.call(e, r[i]) && (n[r[i]] = e[r[i]]);
  return n;
};
const fe = {
  CONNECTING: "connecting",
  OPENED: "opened",
  AUTHENTICATED: "authenticated",
  CLOSED: "closed",
  ERRORED: "errored"
}, wu = 3e4, mu = 3e4, Su = 200, vu = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
}, It = "_instant_oauth_redirect", Rt = "currentUser";
function Ou({ transportType: e, appId: t, apiURI: n, wsURI: r, EventSourceImpl: i }) {
  if (!i)
    return new ar(`${r}?app_id=${t}`);
  switch (e) {
    case "ws":
      return new ar(`${r}?app_id=${t}`);
    case "sse":
      return new gu(i, `${n}/runtime/sse?app_id=${t}`);
    default:
      throw new Error("Unknown transport type " + e);
  }
}
function Eu() {
  return typeof window < "u" || typeof chrome < "u";
}
const cr = {
  "set-presence": !0,
  "set-presence-ok": !0,
  "refresh-presence": !0,
  "patch-presence": !0
};
function Au(e, t) {
  var n;
  const r = typeof e == "string" ? JSON.parse(e) : e;
  for (const i in r) {
    const s = r[i];
    if (!((n = s == null ? void 0 : s.result) === null || n === void 0) && n.store) {
      const o = s.result.store;
      s.result.store = os(Object.assign(Object.assign({}, o), { useDateObjects: t }));
    }
  }
  return r;
}
function Cu(e) {
  var t;
  const n = {};
  for (const r in e) {
    const i = e[r], s = Object.assign({}, i);
    !((t = i.result) === null || t === void 0) && t.store && (s.result = Object.assign(Object.assign({}, i.result), { store: ss(i.result.store) })), n[r] = s;
  }
  return n;
}
function dr(e) {
  return [...e].sort((t, n) => {
    const [r, i] = t, [s, o] = n, u = i.order || 0, a = o.order || 0;
    return u == a ? r < s ? -1 : r > s ? 1 : 0 : u - a;
  });
}
class ju {
  constructor(t, n = Nr, r = Vr, i, s) {
    var o, u, a;
    if (this._isOnline = !0, this._isShutdown = !1, this.status = fe.CONNECTING, this.queryCbs = {}, this.queryOnceDfds = {}, this.authCbs = [], this.attrsCbs = [], this.mutationErrorCbs = [], this.connectionStatusCbs = [], this.mutationDeferredStore = /* @__PURE__ */ new Map(), this._reconnectTimeoutId = null, this._reconnectTimeoutMs = 0, this._transportType = "ws", this._wsOk = null, this._localIdPromises = {}, this._errorMessage = null, this._oauthCallbackResponse = null, this._linkIndex = null, this._rooms = {}, this._roomsPendingLeave = {}, this._presence = {}, this._broadcastQueue = [], this._broadcastSubs = {}, this._currentUserCached = { isLoading: !0, error: void 0, user: void 0 }, this._beforeUnloadCbs = [], this._dataForQueryCache = {}, this._onMergeQuerySubs = (d, c) => {
      const f = d || {}, h = Object.assign({}, c);
      Object.entries(c).forEach(([b, g]) => {
        var w;
        const T = (w = f == null ? void 0 : f[b]) === null || w === void 0 ? void 0 : w.result, m = g.result;
        T && !m && (h[b].result = T);
      }), Object.keys(f).filter((b) => !c[b]).sort((b, g) => {
        var w, T;
        const m = ((w = f[b]) === null || w === void 0 ? void 0 : w.lastAccessed) || 0;
        return (((T = f[g]) === null || T === void 0 ? void 0 : T.lastAccessed) || 0) - m;
      }).slice(0, this.queryCacheLimit).forEach((b) => {
        h[b] = f[b];
      }), this.querySubs.set((b) => h), this.loadedNotifyAll();
    }, this._onMergePendingMutations = (d, c) => {
      const f = new Map([...d.entries(), ...c.entries()]);
      this.pendingMutations.set((p) => f), this.loadedNotifyAll(), this._rewriteMutationsSorted(this.attrs, d).forEach(([p, b]) => {
        !c.has(p) && !b["tx-id"] && this._sendMutation(p, b);
      });
    }, this.getPreviousResult = (d) => {
      const c = G(d);
      return this.dataForQuery(c);
    }, this.notifyOne = (d) => {
      var c, f;
      const h = (c = this.queryCbs[d]) !== null && c !== void 0 ? c : [], p = (f = this._dataForQueryCache[d]) === null || f === void 0 ? void 0 : f.data, b = this.dataForQuery(d);
      b && (ct(b, p) || h.forEach((g) => g.cb(b)));
    }, this.notifyOneQueryOnce = (d) => {
      var c;
      const f = (c = this.queryOnceDfds[d]) !== null && c !== void 0 ? c : [], h = this.dataForQuery(d);
      f.forEach((p) => {
        this._completeQueryOnce(p.q, d, p.dfd), p.dfd.resolve(h);
      });
    }, this.notifyQueryError = (d, c) => {
      (this.queryCbs[d] || []).forEach((h) => h.cb({ error: c }));
    }, this.pushTx = (d) => {
      this.config.disableValidation || _u(d, this.config.schema);
      try {
        const c = zo({
          attrs: this.optimisticAttrs(),
          schema: this.config.schema,
          stores: Object.values(this.querySubs.currentValue).map((f) => {
            var h;
            return (h = f == null ? void 0 : f.result) === null || h === void 0 ? void 0 : h.store;
          }),
          useDateObjects: this.config.useDateObjects
        }, d);
        return this.pushOps(c);
      } catch (c) {
        return this.pushOps([], c);
      }
    }, this.pushOps = (d, c) => {
      const f = V(), h = [...this.pendingMutations.currentValue.values()], p = Math.max(0, ...h.map((w) => w.order || 0)) + 1, b = {
        op: "transact",
        "tx-steps": d,
        created: Date.now(),
        error: c,
        order: p
      };
      this.pendingMutations.set((w) => (w.set(f, b), w));
      const g = new Zn();
      return this.mutationDeferredStore.set(f, g), this._sendMutation(f, b), this.notifyAll(), g.promise;
    }, this._transportOnOpen = (d) => {
      const c = d.target;
      if (this._transport !== c) {
        this._log.info("[socket][open]", c.id, "skip; this is no longer the current transport");
        return;
      }
      this._log.info("[socket][open]", this._transport.id), this._setStatus(fe.OPENED), this.getCurrentUser().then((f) => {
        var h;
        this._trySend(V(), {
          op: "init",
          "app-id": this.config.appId,
          "refresh-token": (h = f.user) === null || h === void 0 ? void 0 : h.refresh_token,
          versions: this.versions,
          // If an admin token is provided for an app, we will
          // skip all permission checks. This is an advanced feature,
          // to let users write internal tools
          // This option is not exposed in `Config`, as it's
          // not ready for prime time
          "__admin-token": this.config.__adminToken
        });
      }).catch((f) => {
        this._log.error("[socket][error]", c.id, f);
      });
    }, this._transportOnMessage = (d) => {
      const c = d.target, f = d.message;
      if (this._transport !== c) {
        this._log.info("[socket][message]", c.id, f, "skip; this is no longer the current transport");
        return;
      }
      !this._wsOk && c.type === "ws" && (this._wsOk = !0), this._transportType = "ws", this._handleReceive(c.id, d.message);
    }, this._transportOnError = (d) => {
      const c = d.target;
      if (this._transport !== c) {
        this._log.info("[socket][error]", c.id, "skip; this is no longer the current transport");
        return;
      }
      this._log.error("[socket][error]", c.id, d);
    }, this._scheduleReconnect = () => {
      !this._wsOk && this._transportType !== "sse" && (this._transportType = "sse", this._reconnectTimeoutMs = 0), setTimeout(() => {
        if (this._reconnectTimeoutMs = Math.min(this._reconnectTimeoutMs + 1e3, 1e4), !this._isOnline) {
          this._log.info("[socket][close]", this._transport.id, "we are offline, no need to start socket");
          return;
        }
        this._startSocket();
      }, this._reconnectTimeoutMs);
    }, this._transportOnClose = (d) => {
      const c = d.target;
      if (this._transport !== c) {
        this._log.info("[socket][close]", c.id, "skip; this is no longer the current transport");
        return;
      }
      this._setStatus(fe.CLOSED);
      for (const f of Object.values(this._rooms))
        f.isConnected = !1;
      if (this._isShutdown) {
        this._log.info("[socket][close]", c.id, "Reactor has been shut down and will not reconnect");
        return;
      }
      this._log.info("[socket][close]", c.id, "schedule reconnect, ms =", this._reconnectTimeoutMs), this._scheduleReconnect();
    }, this._EventSource = s, this.config = Object.assign(Object.assign({}, vu), t), this.queryCacheLimit = (o = this.config.queryCacheLimit) !== null && o !== void 0 ? o : 10, this._pendingTxCleanupTimeout = (u = this.config.pendingTxCleanupTimeout) !== null && u !== void 0 ? u : mu, this._pendingMutationCleanupThreshold = (a = this.config.pendingMutationCleanupThreshold) !== null && a !== void 0 ? a : Su, this._log = au(t.verbose || bn || qr), this.versions = Object.assign(Object.assign({}, i || {}), { "@instantdb/core": zr }), this.config.schema && (this._linkIndex = tr(this.config.schema)), !!Eu()) {
      if (!t.appId)
        throw new Error("Instant must be initialized with an appId.");
      if (!je(t.appId))
        throw new Error(`Instant must be initialized with a valid appId. \`${t.appId}\` is not a valid uuid.`);
      typeof BroadcastChannel == "function" && (this._broadcastChannel = new BroadcastChannel("@instantdb"), this._broadcastChannel.addEventListener("message", (d) => D(this, void 0, void 0, function* () {
        var c;
        try {
          if (((c = d.data) === null || c === void 0 ? void 0 : c.type) === "auth") {
            const f = yield this.getCurrentUser();
            this.updateUser(f.user);
          }
        } catch (f) {
          this._log.error("[error] handle broadcast channel", f);
        }
      }))), this._initStorage(n), this._oauthCallbackResponse = this._oauthLoginInit(), this.getCurrentUser(), r.getIsOnline().then((d) => {
        this._isOnline = d, this._startSocket(), r.listen((c) => {
          c !== this._isOnline && (this._log.info("[network] online =", c), this._isOnline = c, this._isOnline ? this._startSocket() : (this._log.info("Changing status from", this.status, "to", fe.CLOSED), this._setStatus(fe.CLOSED)));
        });
      }), typeof addEventListener < "u" && (this._beforeUnload = this._beforeUnload.bind(this), addEventListener("beforeunload", this._beforeUnload));
    }
  }
  updateSchema(t) {
    this.config = Object.assign(Object.assign({}, this.config), { schema: t, cardinalityInference: !!t }), this._linkIndex = t ? tr(this.config.schema) : null;
  }
  _initStorage(t) {
    this._persister = new t(`instant_${this.config.appId}_5`), this.querySubs = new Xn(this._persister, "querySubs", {}, this._onMergeQuerySubs, Cu, (n) => Au(n, this.config.useDateObjects)), this.pendingMutations = new Xn(
      this._persister,
      "pendingMutations",
      /* @__PURE__ */ new Map(),
      this._onMergePendingMutations,
      (n) => [...n.entries()],
      (n) => new Map(typeof n == "string" ? JSON.parse(n) : n),
      100,
      // throttle
      100
    ), this._beforeUnloadCbs.push(() => {
      this.pendingMutations.flush(), this.querySubs.flush();
    });
  }
  _beforeUnload() {
    for (const t of this._beforeUnloadCbs)
      t();
  }
  /**
   * @param {'enqueued' | 'pending' | 'synced' | 'timeout' |  'error' } status
   * @param {string} eventId
   * @param {{message?: string, type?: string, status?: number, hint?: unknown}} [errorMsg]
   */
  _finishTransaction(t, n, r) {
    const i = this.mutationDeferredStore.get(n);
    this.mutationDeferredStore.delete(n);
    const s = t !== "error" && t !== "timeout";
    if (!i && !s && console.error("Mutation failed", Object.assign({ status: t, eventId: n }, r)), !!i)
      if (s)
        i.resolve({ status: t, eventId: n });
      else if (r.type) {
        const { status: o } = r, u = Tu(r, ["status"]);
        i.reject(new gt({
          // @ts-expect-error body.type is not constant typed
          body: u,
          status: o
        }));
      } else
        i.reject(new ke(r.message, r.hint));
  }
  _setStatus(t, n) {
    this.status = t, this._errorMessage = n, this.notifyConnectionStatusSubs(t);
  }
  _flushEnqueuedRoomData(t) {
    var n, r;
    const i = (r = (n = this._presence[t]) === null || n === void 0 ? void 0 : n.result) === null || r === void 0 ? void 0 : r.user, s = this._broadcastQueue[t];
    if (this._broadcastQueue[t] = [], i && this._trySetPresence(t, i), s)
      for (const o of s) {
        const { topic: u, roomType: a, data: d } = o;
        this._tryBroadcast(t, a, u, d);
      }
  }
  _handleReceive(t, n) {
    var r, i, s, o, u, a;
    const d = !!this.config.schema && ("cardinalityInference" in this.config ? !!this.config.cardinalityInference : !0);
    switch (cr[n.op] || this._log.info("[receive]", t, n.op, n), n.op) {
      case "init-ok":
        this._setStatus(fe.AUTHENTICATED), this._reconnectTimeoutMs = 0, this._setAttrs(n.attrs), this._flushPendingMessages(), this._sessionId = n["session-id"];
        for (const k of Object.keys(this._rooms)) {
          const U = (i = (r = this._presence[k]) === null || r === void 0 ? void 0 : r.result) === null || i === void 0 ? void 0 : i.user;
          this._tryJoinRoom(k, U);
        }
        break;
      case "add-query-exists":
        this.notifyOneQueryOnce(G(n.q));
        break;
      case "add-query-ok":
        const { q: c, result: f } = n, h = G(c), p = (o = (s = f == null ? void 0 : f[0]) === null || s === void 0 ? void 0 : s.data) === null || o === void 0 ? void 0 : o["page-info"], b = (a = (u = f == null ? void 0 : f[0]) === null || u === void 0 ? void 0 : u.data) === null || a === void 0 ? void 0 : a.aggregate, g = er(f), w = Kt(this.attrs, g, d, this._linkIndex, this.config.useDateObjects);
        this.querySubs.set((k) => (k[h].result = {
          store: w,
          pageInfo: p,
          aggregate: b,
          processedTxId: n["processed-tx-id"]
        }, k)), this._cleanupPendingMutationsQueries(), this.notifyOne(h), this.notifyOneQueryOnce(h), this._cleanupPendingMutationsTimeout();
        break;
      case "refresh-ok":
        const { computations: T, attrs: m } = n, S = n["processed-tx-id"];
        m && this._setAttrs(m), this._cleanupPendingMutationsTimeout();
        const q = this._rewriteMutations(this.attrs, this.pendingMutations.currentValue, S);
        q !== this.pendingMutations.currentValue && this.pendingMutations.set(() => q);
        const O = dr(q.entries()), x = T.map((k) => {
          var U, z, B, re;
          const ie = k["instaql-query"], Z = k["instaql-result"], Ge = G(ie), Ye = er(Z), He = Kt(this.attrs, Ye, d, this._linkIndex, this.config.useDateObjects), Je = this._applyOptimisticUpdates(He, O, S), Ze = (z = (U = Z == null ? void 0 : Z[0]) === null || U === void 0 ? void 0 : U.data) === null || z === void 0 ? void 0 : z["page-info"], Tt = (re = (B = Z == null ? void 0 : Z[0]) === null || B === void 0 ? void 0 : B.data) === null || re === void 0 ? void 0 : re.aggregate;
          return { hash: Ge, store: Je, pageInfo: Ze, aggregate: Tt };
        });
        x.forEach(({ hash: k, store: U, pageInfo: z, aggregate: B }) => {
          this.querySubs.set((re) => (re[k].result = { store: U, pageInfo: z, aggregate: B, processedTxId: S }, re));
        }), this._cleanupPendingMutationsQueries(), x.forEach(({ hash: k }) => {
          this.notifyOne(k);
        });
        break;
      case "transact-ok": {
        const { "client-event-id": k, "tx-id": U } = n, B = this._rewriteMutations(this.attrs, this.pendingMutations.currentValue).get(k);
        if (!B)
          break;
        this.pendingMutations.set((ie) => (ie.set(k, Object.assign(Object.assign({}, ie.get(k)), { "tx-id": U, confirmed: Date.now() })), ie));
        const re = B["tx-steps"].filter(([ie, ...Z]) => ie === "add-attr").map(([ie, Z]) => Z).concat(Object.values(this.attrs));
        this._setAttrs(re), this._finishTransaction("synced", k), this._cleanupPendingMutationsTimeout();
        break;
      }
      case "patch-presence": {
        const k = n["room-id"];
        this._patchPresencePeers(k, n.edits), this._notifyPresenceSubs(k);
        break;
      }
      case "refresh-presence": {
        const k = n["room-id"];
        this._setPresencePeers(k, n.data), this._notifyPresenceSubs(k);
        break;
      }
      case "server-broadcast":
        const N = n["room-id"], de = n.topic;
        this._notifyBroadcastSubs(N, de, n);
        break;
      case "join-room-ok":
        const ae = n["room-id"], Me = this._rooms[ae];
        if (!Me) {
          this._roomsPendingLeave[ae] && (this._tryLeaveRoom(ae), delete this._roomsPendingLeave[ae]);
          break;
        }
        Me.isConnected = !0, this._notifyPresenceSubs(ae), this._flushEnqueuedRoomData(ae);
        break;
      case "join-room-error":
        const Be = n["room-id"], Qe = this._rooms[Be];
        Qe && (Qe.error = n.error), this._notifyPresenceSubs(Be);
        break;
      case "error":
        this._handleReceiveError(n);
        break;
    }
  }
  /**
   * @param {'timeout' | 'error'} status
   * @param {string} eventId
   * @param {{message: string, type?: string, status?: number, hint?: unknown}} errorMsg
   */
  _handleMutationError(t, n, r) {
    const i = this.pendingMutations.currentValue.get(n);
    if (i && (t !== "timeout" || !i["tx-id"])) {
      this.pendingMutations.set((o) => (o.delete(n), o));
      const s = {
        message: r.message,
        hint: r.hint
      };
      this.notifyAll(), this.notifyAttrsSubs(), this.notifyMutationErrorSubs(s), this._finishTransaction(t, n, r);
    }
  }
  _handleReceiveError(t) {
    var n, r, i, s, o;
    console.log("error", t);
    const u = t["client-event-id"], a = this.pendingMutations.currentValue.get(u), d = {
      message: t.message || "Uh-oh, something went wrong. Ping Joe & Stopa."
    };
    if (t.hint && (d.hint = t.hint), a) {
      this._handleMutationError("error", u, t);
      return;
    }
    if (!((n = t["original-event"]) === null || n === void 0) && n.hasOwnProperty("q") && ((r = t["original-event"]) === null || r === void 0 ? void 0 : r.op) === "add-query") {
      const h = (i = t["original-event"]) === null || i === void 0 ? void 0 : i.q, p = G(h);
      this.notifyQueryError(G(h), d), this.notifyQueryOnceError(h, p, u, d);
      return;
    }
    if (((s = t["original-event"]) === null || s === void 0 ? void 0 : s.op) === "init") {
      if (t.type === "record-not-found" && ((o = t.hint) === null || o === void 0 ? void 0 : o["record-type"]) === "app-user") {
        this.changeCurrentUser(null);
        return;
      }
      this._setStatus(fe.ERRORED, d), this.notifyAll();
      return;
    }
    const f = Object.assign({}, t);
    delete f.message, delete f.hint, console.error(t.message, f), t.hint && console.error(`This error comes with some debugging information. Here it is: 
`, t.hint);
  }
  notifyQueryOnceError(t, n, r, i) {
    var s;
    const o = (s = this.queryOnceDfds[n]) === null || s === void 0 ? void 0 : s.find((u) => u.eventId === r);
    o && (o.dfd.reject(i), this._completeQueryOnce(t, n, o.dfd));
  }
  _setAttrs(t) {
    this.attrs = t.reduce((n, r) => (n[r.id] = r, n), {}), this.notifyAttrsSubs();
  }
  _startQuerySub(t, n) {
    const r = V();
    return this.querySubs.set((i) => (i[n] = i[n] || { q: t, result: null, eventId: r }, i[n].lastAccessed = Date.now(), i)), this._trySendAuthed(r, { op: "add-query", q: t }), r;
  }
  /**
   *  When a user subscribes to a query the following side effects occur:
   *
   *  - We update querySubs to include the new query
   *  - We update queryCbs to include the new cb
   *  - If we already have a result for the query we call cb immediately
   *  - We send the server an `add-query` message
   *
   *  Returns an unsubscribe function
   */
  subscribeQuery(t, n, r) {
    var i;
    this.config.disableValidation || rr(t, this.config.schema), r && "ruleParams" in r && (t = Object.assign({ $$ruleParams: r.ruleParams }, t));
    const s = G(t), o = this.getPreviousResult(t);
    return o && n(o), this.queryCbs[s] = (i = this.queryCbs[s]) !== null && i !== void 0 ? i : [], this.queryCbs[s].push({ q: t, cb: n }), this._startQuerySub(t, s), () => {
      this._unsubQuery(t, s, n);
    };
  }
  queryOnce(t, n) {
    var r;
    this.config.disableValidation || rr(t, this.config.schema), n && "ruleParams" in n && (t = Object.assign({ $$ruleParams: n.ruleParams }, t));
    const i = new Zn();
    if (!this._isOnline)
      return i.reject(new Error("We can't run `queryOnce`, because the device is offline.")), i.promise;
    if (!this.querySubs)
      return i.reject(new Error("We can't run `queryOnce` on the backend. Use adminAPI.query instead: https://www.instantdb.com/docs/backend#query")), i.promise;
    const s = G(t), o = this._startQuerySub(t, s);
    return this.queryOnceDfds[s] = (r = this.queryOnceDfds[s]) !== null && r !== void 0 ? r : [], this.queryOnceDfds[s].push({ q: t, dfd: i, eventId: o }), setTimeout(() => i.reject(new Error("Query timed out")), wu), i.promise;
  }
  _completeQueryOnce(t, n, r) {
    this.queryOnceDfds[n] && (this.queryOnceDfds[n] = this.queryOnceDfds[n].filter((i) => i.dfd !== r), this._cleanupQuery(t, n));
  }
  _unsubQuery(t, n, r) {
    this.queryCbs[n] && (this.queryCbs[n] = this.queryCbs[n].filter((i) => i.cb !== r), this._cleanupQuery(t, n));
  }
  _cleanupQuery(t, n) {
    var r, i;
    !((r = this.queryCbs[n]) === null || r === void 0) && r.length || !((i = this.queryOnceDfds[n]) === null || i === void 0) && i.length || (delete this.queryCbs[n], delete this.queryOnceDfds[n], this._trySendAuthed(V(), { op: "remove-query", q: t }));
  }
  // When we `pushTx`, it's possible that we don't yet have `this.attrs`
  // This means that `tx-steps` in `pendingMutations` will include `add-attr`
  // commands for attrs that already exist.
  //
  // This will also affect `add-triple` and `retract-triple` which
  // reference attr-ids that do not match the server.
  //
  // We fix this by rewriting `tx-steps` in each `pendingMutation`.
  // We remove `add-attr` commands for attrs that already exist.
  // We update `add-triple` and `retract-triple` commands to use the
  // server attr-ids.
  _rewriteMutations(t, n, r) {
    if (!t)
      return n;
    const i = (c) => {
      const [f, h, p] = c["forward-identity"];
      return F(t, h, p);
    }, s = (c) => {
      const [f, h, p] = c["forward-identity"];
      return Te(t, h, p);
    }, o = { attrIdMap: {}, refSwapAttrIds: /* @__PURE__ */ new Set() };
    let u = !1;
    const a = (c, f) => {
      const h = [];
      for (const p of c) {
        const [b] = p;
        if (b === "add-attr") {
          const [w, T] = p, m = i(T);
          if (m && T.id !== m.id) {
            o.attrIdMap[T.id] = m.id, u = !0;
            continue;
          }
          if (T["value-type"] === "ref") {
            const S = s(T);
            if (S) {
              o.attrIdMap[T.id] = S.id, o.refSwapAttrIds.add(T.id), u = !0;
              continue;
            }
          }
        }
        if (r && f && r >= f && b === "add-attr" || b === "update-attr" || b === "delete-attr") {
          u = !0;
          continue;
        }
        const g = u ? mo(o, p) : p;
        h.push(g);
      }
      return u ? h : c;
    }, d = /* @__PURE__ */ new Map();
    for (const [c, f] of n.entries())
      d.set(c, Object.assign(Object.assign({}, f), { "tx-steps": a(f["tx-steps"], f["tx-id"]) }));
    return u ? d : n;
  }
  _rewriteMutationsSorted(t, n) {
    return dr(this._rewriteMutations(t, n).entries());
  }
  // ---------------------------
  // Transact
  optimisticAttrs() {
    var t;
    const n = [
      ...this.pendingMutations.currentValue.values()
    ].flatMap((u) => u["tx-steps"]), r = new Set(n.filter(([u, a]) => u === "delete-attr").map(([u, a]) => a)), i = [];
    for (const [u, a] of n)
      if (u === "add-attr")
        i.push(a);
      else if (u === "update-attr" && a.id && (!((t = this.attrs) === null || t === void 0) && t[a.id])) {
        const d = Object.assign(Object.assign({}, this.attrs[a.id]), a);
        i.push(d);
      }
    const s = [
      ...Object.values(this.attrs || {}),
      ...i
    ].filter((u) => !r.has(u.id));
    return Object.fromEntries(s.map((u) => [u.id, u]));
  }
  /** Runs instaql on a query and a store */
  dataForQuery(t) {
    const n = this._errorMessage;
    if (n)
      return { error: n };
    if (!this.querySubs || !this.pendingMutations)
      return;
    const r = this.querySubs.version(), i = this.querySubs.currentValue, s = this.pendingMutations.version(), o = this.pendingMutations.currentValue, { q: u, result: a } = i[t] || {};
    if (!a)
      return;
    const d = this._dataForQueryCache[t];
    if (d && r === d.querySubVersion && s === d.pendingMutationsVersion)
      return d.data;
    const { store: c, pageInfo: f, aggregate: h, processedTxId: p } = a, b = this._rewriteMutationsSorted(c.attrs, o), g = this._applyOptimisticUpdates(c, b, p), w = bo({ store: g, pageInfo: f, aggregate: h }, u);
    return this._dataForQueryCache[t] = {
      querySubVersion: r,
      pendingMutationsVersion: s,
      data: w
    }, w;
  }
  _applyOptimisticUpdates(t, n, r) {
    for (const [i, s] of n)
      (!s["tx-id"] || r && s["tx-id"] > r) && (t = ws(t, s["tx-steps"]));
    return t;
  }
  /** Re-compute all subscriptions */
  notifyAll() {
    Object.keys(this.queryCbs).forEach((t) => {
      this.notifyOne(t);
    });
  }
  loadedNotifyAll() {
    this.pendingMutations.isLoading() || this.querySubs.isLoading() || this.notifyAll();
  }
  shutdown() {
    var t;
    this._log.info("[shutdown]", this.config.appId), this._isShutdown = !0, (t = this._transport) === null || t === void 0 || t.close();
  }
  /**
   * Sends mutation to server and schedules a timeout to cancel it if
   * we don't hear back in time.
   * Note: If we're offline we don't schedule a timeout, we'll schedule it
   * later once we're back online and send the mutation again
   *
   */
  _sendMutation(t, n) {
    if (n.error) {
      this._handleMutationError("error", t, {
        message: n.error.message
      });
      return;
    }
    if (this.status !== fe.AUTHENTICATED) {
      this._finishTransaction("enqueued", t);
      return;
    }
    const r = Math.max(5e3, this.pendingMutations.currentValue.size * 5e3);
    this._isOnline ? (this._trySend(t, n), setTimeout(() => {
      this._isOnline && this._handleMutationError("timeout", t, {
        message: "transaction timed out"
      });
    }, r)) : this._finishTransaction("enqueued", t);
  }
  // ---------------------------
  // Websocket
  /** Send messages we accumulated while we were connecting */
  _flushPendingMessages() {
    Object.keys(this.queryCbs).map((i) => this.querySubs.currentValue[i]).filter((i) => i).forEach(({ eventId: i, q: s }) => {
      this._trySendAuthed(i, { op: "add-query", q: s });
    }), Object.values(this.queryOnceDfds).flat().forEach(({ eventId: i, q: s }) => {
      this._trySendAuthed(i, { op: "add-query", q: s });
    }), this._rewriteMutationsSorted(this.attrs, this.pendingMutations.currentValue).forEach(([i, s]) => {
      s["tx-id"] || this._sendMutation(i, s);
    });
  }
  /**
   * Clean up pendingMutations that all queries have seen
   */
  _cleanupPendingMutationsQueries() {
    let t = Number.MAX_SAFE_INTEGER;
    for (const { result: n } of Object.values(this.querySubs.currentValue))
      n != null && n.processedTxId && (t = Math.min(t, n == null ? void 0 : n.processedTxId));
    this.pendingMutations.set((n) => {
      for (const [r, i] of Array.from(n.entries()))
        i["tx-id"] && i["tx-id"] <= t && n.delete(r);
      return n;
    });
  }
  /**
   * After mutations is confirmed by server, we give each query 30 sec
   * to update its results. If that doesn't happen, we assume query is
   * unaffected by this mutation and it’s safe to delete it from local queue
   */
  _cleanupPendingMutationsTimeout() {
    if (this.pendingMutations.currentValue.size < this._pendingMutationCleanupThreshold)
      return;
    const t = Date.now();
    this.pendingMutations.set((n) => {
      for (const [r, i] of Array.from(n.entries()))
        i.confirmed && i.confirmed + this._pendingTxCleanupTimeout < t && n.delete(r);
      return n;
    });
  }
  _trySendAuthed(...t) {
    this.status === fe.AUTHENTICATED && this._trySend(...t);
  }
  _trySend(t, n, r) {
    this._transport.isOpen() && (cr[n.op] || this._log.info("[send]", this._transport.id, n.op, n), this._transport.send(Object.assign({ "client-event-id": t }, n)));
  }
  _startSocket() {
    if (this._wsOk = null, this._isShutdown) {
      this._log.info("[socket][start]", this.config.appId, "Reactor has been shut down and will not start a new socket");
      return;
    }
    if (this._transport && this._transport.isConnecting()) {
      this._log.info("[socket][start]", this._transport.id, "maintained as current transport, we were still in a connecting state");
      return;
    }
    const t = this._transport;
    this._transport = Ou({
      transportType: this._transportType,
      appId: this.config.appId,
      apiURI: this.config.apiURI,
      wsURI: this.config.websocketURI,
      EventSourceImpl: this._EventSource
    }), this._transport.onopen = this._transportOnOpen, this._transport.onmessage = this._transportOnMessage, this._transport.onclose = this._transportOnClose, this._transport.onerror = this._transportOnError, this._log.info("[socket][start]", this._transport.id), t != null && t.isOpen() && (this._log.info("[socket][start]", this._transport.id, "close previous transport id = ", t.id), t.close());
  }
  /**
   * Given a key, returns a stable local id, unique to this device and app.
   *
   * This can be useful if you want to create guest ids for example.
   *
   * Note: If the user deletes their local storage, this id will change.
   *
   * We use this._localIdPromises to ensure that we only generate a local
   * id once, even if multiple callers call this function concurrently.
   */
  getLocalId(t) {
    return D(this, void 0, void 0, function* () {
      const n = `localToken_${t}`, r = yield this._persister.getItem(n);
      if (r)
        return r;
      if (this._localIdPromises[n])
        return this._localIdPromises[n];
      const i = V();
      return this._localIdPromises[n] = this._persister.setItem(n, i).then(() => i), this._localIdPromises[n];
    });
  }
  // ----
  // Auth
  _replaceUrlAfterOAuth() {
    if (typeof URL > "u")
      return;
    const t = new URL(window.location.href);
    if (t.searchParams.get(It)) {
      const n = t.toString();
      t.searchParams.delete(It), t.searchParams.delete("code"), t.searchParams.delete("error");
      const r = t.pathname + (t.searchParams.size ? "?" + t.searchParams : "") + t.hash;
      if (history.replaceState(history.state, "", r), // @ts-ignore (waiting for ts support)
      typeof navigation == "object" && // @ts-ignore (waiting for ts support)
      typeof navigation.addEventListener == "function" && // @ts-ignore (waiting for ts support)
      typeof navigation.removeEventListener == "function") {
        let i = !1;
        const s = (o) => {
          var u;
          i || (i = !0, navigation.removeEventListener("navigate", s), !o.userInitiated && o.navigationType === "replace" && ((u = o.destination) === null || u === void 0 ? void 0 : u.url) === n && history.replaceState(history.state, "", r));
        };
        navigation.addEventListener("navigate", s);
      }
    }
  }
  /**
   *
   * @returns Promise<null | {error: {message: string}}>
   */
  _oauthLoginInit() {
    return D(this, void 0, void 0, function* () {
      var t, n, r, i;
      if (typeof window > "u" || typeof window.location > "u" || typeof URLSearchParams > "u")
        return null;
      const s = new URLSearchParams(window.location.search);
      if (!s.get(It))
        return null;
      const o = s.get("error");
      if (o)
        return this._replaceUrlAfterOAuth(), { error: { message: o } };
      const u = s.get("code");
      if (!u)
        return null;
      this._replaceUrlAfterOAuth();
      try {
        const a = yield this._getCurrentUser(), d = (a == null ? void 0 : a.type) === "guest", { user: c } = yield Hn({
          apiURI: this.config.apiURI,
          appId: this.config.appId,
          code: u,
          refreshToken: d ? a.refresh_token : void 0
        });
        return this.setCurrentUser(c), null;
      } catch (a) {
        return ((t = a == null ? void 0 : a.body) === null || t === void 0 ? void 0 : t.type) === "record-not-found" && ((r = (n = a == null ? void 0 : a.body) === null || n === void 0 ? void 0 : n.hint) === null || r === void 0 ? void 0 : r["record-type"]) === "app-oauth-code" && (yield this._hasCurrentUser()) ? null : { error: { message: ((i = a == null ? void 0 : a.body) === null || i === void 0 ? void 0 : i.message) || "Error logging in." } };
      }
    });
  }
  _waitForOAuthCallbackResponse() {
    return D(this, void 0, void 0, function* () {
      return yield this._oauthCallbackResponse;
    });
  }
  __subscribeMutationErrors(t) {
    return this.mutationErrorCbs.push(t), () => {
      this.mutationErrorCbs = this.mutationErrorCbs.filter((n) => n !== t);
    };
  }
  subscribeAuth(t) {
    this.authCbs.push(t);
    const n = this._currentUserCached;
    n.isLoading || t(this._currentUserCached);
    let r = !1;
    return this.getCurrentUser().then((i) => {
      r || ct(i, n) || t(i);
    }), () => {
      r = !0, this.authCbs = this.authCbs.filter((i) => i !== t);
    };
  }
  getAuth() {
    return D(this, void 0, void 0, function* () {
      const { user: t, error: n } = yield this.getCurrentUser();
      if (n)
        throw new ke("Could not get current user: " + n.message);
      return t;
    });
  }
  subscribeConnectionStatus(t) {
    return this.connectionStatusCbs.push(t), () => {
      this.connectionStatusCbs = this.connectionStatusCbs.filter((n) => n !== t);
    };
  }
  subscribeAttrs(t) {
    return this.attrsCbs.push(t), this.attrs && t(this.attrs), () => {
      this.attrsCbs = this.attrsCbs.filter((n) => n !== t);
    };
  }
  notifyAuthSubs(t) {
    this.authCbs.forEach((n) => n(t));
  }
  notifyMutationErrorSubs(t) {
    this.mutationErrorCbs.forEach((n) => n(t));
  }
  notifyAttrsSubs() {
    if (!this.attrs)
      return;
    const t = this.optimisticAttrs();
    this.attrsCbs.forEach((n) => n(t));
  }
  notifyConnectionStatusSubs(t) {
    this.connectionStatusCbs.forEach((n) => n(t));
  }
  setCurrentUser(t) {
    return D(this, void 0, void 0, function* () {
      yield this._persister.setItem(Rt, t);
    });
  }
  getCurrentUserCached() {
    return this._currentUserCached;
  }
  _getCurrentUser() {
    return D(this, void 0, void 0, function* () {
      const t = yield this._persister.getItem(Rt);
      return typeof t == "string" ? JSON.parse(t) : t;
    });
  }
  getCurrentUser() {
    return D(this, void 0, void 0, function* () {
      const t = yield this._waitForOAuthCallbackResponse();
      if (t != null && t.error) {
        const i = { error: t.error, user: void 0 };
        return this._currentUserCached = Object.assign({ isLoading: !1 }, i), i;
      }
      const r = { user: yield this._getCurrentUser(), error: void 0 };
      return this._currentUserCached = Object.assign({ isLoading: !1 }, r), r;
    });
  }
  _hasCurrentUser() {
    return D(this, void 0, void 0, function* () {
      const t = yield this._persister.getItem(Rt);
      return JSON.parse(t) != null;
    });
  }
  changeCurrentUser(t) {
    return D(this, void 0, void 0, function* () {
      var n;
      const { user: r } = yield this.getCurrentUser();
      if (!ct(r, t)) {
        yield this.setCurrentUser(t), this.updateUser(t);
        try {
          (n = this._broadcastChannel) === null || n === void 0 || n.postMessage({ type: "auth" });
        } catch (i) {
          console.error("Error posting message to broadcast channel", i);
        }
      }
    });
  }
  updateUser(t) {
    const n = { error: void 0, user: t };
    this._currentUserCached = Object.assign({ isLoading: !1 }, n), this._dataForQueryCache = {}, this.querySubs.set((r) => (Object.keys(r).forEach((i) => {
      delete r[i].result;
    }), r)), this._reconnectTimeoutMs = 0, this._transport.close(), this._oauthCallbackResponse = null, this.notifyAuthSubs(n);
  }
  sendMagicCode({ email: t }) {
    return Go({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: t
    });
  }
  signInWithMagicCode(t) {
    return D(this, arguments, void 0, function* ({ email: n, code: r }) {
      var i;
      const s = yield this.getCurrentUser(), o = ((i = s == null ? void 0 : s.user) === null || i === void 0 ? void 0 : i.type) === "guest", u = yield Yo({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        email: n,
        code: r,
        refreshToken: o ? s.user.refresh_token : void 0
      });
      return yield this.changeCurrentUser(u.user), u;
    });
  }
  signInWithCustomToken(t) {
    return D(this, void 0, void 0, function* () {
      const n = yield Ho({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        refreshToken: t
      });
      return yield this.changeCurrentUser(n.user), n;
    });
  }
  signInAsGuest() {
    return D(this, void 0, void 0, function* () {
      const t = yield Jo({
        apiURI: this.config.apiURI,
        appId: this.config.appId
      });
      return yield this.changeCurrentUser(t.user), t;
    });
  }
  potentiallyInvalidateToken(t, n) {
    var r;
    const i = (r = t == null ? void 0 : t.user) === null || r === void 0 ? void 0 : r.refresh_token;
    if (!i)
      return;
    if (n.invalidateToken === !1) {
      this._log.info("[auth-invalidate] skipped invalidateToken");
      return;
    }
    Xo({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      refreshToken: i
    }).then(() => {
      this._log.info("[auth-invalidate] completed invalidateToken");
    }).catch((o) => {
    });
  }
  signOut(t) {
    return D(this, void 0, void 0, function* () {
      const n = yield this.getCurrentUser();
      this.potentiallyInvalidateToken(n, t), yield this.changeCurrentUser(null);
    });
  }
  /**
   * Creates an OAuth authorization URL.
   *
   * @param {Object} params - The parameters to create the authorization URL.
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.redirectURL - The URL to redirect users to after authorization.
   * @returns {string} The created authorization URL.
   */
  createAuthorizationURL({ clientName: t, redirectURL: n }) {
    const { apiURI: r, appId: i } = this.config;
    return `${r}/runtime/oauth/start?app_id=${i}&client_name=${t}&redirect_uri=${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.code - The code received from the OAuth service.
   * @param {string} [params.codeVerifier] - The code verifier used to generate the code challenge.
   */
  exchangeCodeForToken(t) {
    return D(this, arguments, void 0, function* ({ code: n, codeVerifier: r }) {
      var i;
      const s = yield this.getCurrentUser(), o = ((i = s == null ? void 0 : s.user) === null || i === void 0 ? void 0 : i.type) === "guest", u = yield Hn({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        code: n,
        codeVerifier: r,
        refreshToken: o ? s.user.refresh_token : void 0
      });
      return yield this.changeCurrentUser(u.user), u;
    });
  }
  issuerURI() {
    const { apiURI: t, appId: n } = this.config;
    return `${t}/runtime/${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.idToken - The id_token from the external service
   * @param {string | null | undefined} [params.nonce] - The nonce used when requesting the id_token from the external service
   */
  signInWithIdToken(t) {
    return D(this, arguments, void 0, function* ({ idToken: n, clientName: r, nonce: i }) {
      var s;
      const o = yield this.getCurrentUser(), u = (s = o == null ? void 0 : o.user) === null || s === void 0 ? void 0 : s.refresh_token, a = yield Zo({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        idToken: n,
        clientName: r,
        nonce: i,
        refreshToken: u
      });
      return yield this.changeCurrentUser(a.user), a;
    });
  }
  // --------
  // Rooms
  /**
   * @param {string} roomId
   * @param {any | null | undefined} [initialData] -- initial presence data to send when joining the room
   * @returns () => void
   */
  joinRoom(t, n) {
    return this._rooms[t] || (this._rooms[t] = {
      isConnected: !1,
      error: void 0
    }), this._presence[t] = this._presence[t] || {}, n && (this._presence[t].result = this._presence[t].result || {}, this._presence[t].result.user = n, this._notifyPresenceSubs(t)), this._tryJoinRoom(t, n), () => {
      this._cleanupRoom(t);
    };
  }
  _cleanupRoom(t) {
    var n, r, i, s;
    if (!(!((r = (n = this._presence[t]) === null || n === void 0 ? void 0 : n.handlers) === null || r === void 0) && r.length) && !Object.keys((i = this._broadcastSubs[t]) !== null && i !== void 0 ? i : {}).length) {
      const o = (s = this._rooms[t]) === null || s === void 0 ? void 0 : s.isConnected;
      delete this._rooms[t], delete this._presence[t], delete this._broadcastSubs[t], o ? this._tryLeaveRoom(t) : this._roomsPendingLeave[t] = !0;
    }
  }
  // --------
  // Presence
  // TODO: look into typing again
  getPresence(t, n, r = {}) {
    const i = this._rooms[n], s = this._presence[n];
    return !i || !s || !s.result ? null : Object.assign(Object.assign({}, su(s.result, r, this._sessionId)), { isLoading: !i.isConnected, error: i.error });
  }
  // TODO: look into typing again
  publishPresence(t, n, r) {
    const i = this._rooms[n], s = this._presence[n];
    if (!i || !s)
      return;
    s.result = s.result || {};
    const o = Object.assign(Object.assign({}, s.result.user), r);
    s.result.user = o, i.isConnected && (this._trySetPresence(n, o), this._notifyPresenceSubs(n));
  }
  _trySetPresence(t, n) {
    this._trySendAuthed(V(), {
      op: "set-presence",
      "room-id": t,
      data: n
    });
  }
  _tryJoinRoom(t, n) {
    this._trySendAuthed(V(), { op: "join-room", "room-id": t, data: n }), delete this._roomsPendingLeave[t];
  }
  _tryLeaveRoom(t) {
    this._trySendAuthed(V(), { op: "leave-room", "room-id": t });
  }
  // TODO: look into typing again
  subscribePresence(t, n, r, i) {
    const s = this.joinRoom(n, r.initialData), o = Object.assign(Object.assign({}, r), { roomId: n, cb: i, prev: null });
    return this._presence[n] = this._presence[n] || {}, this._presence[n].handlers = this._presence[n].handlers || [], this._presence[n].handlers.push(o), this._notifyPresenceSub(n, o), () => {
      var u, a, d;
      this._presence[n].handlers = (d = (a = (u = this._presence[n]) === null || u === void 0 ? void 0 : u.handlers) === null || a === void 0 ? void 0 : a.filter((c) => c !== o)) !== null && d !== void 0 ? d : [], s();
    };
  }
  _notifyPresenceSubs(t) {
    var n, r;
    (r = (n = this._presence[t]) === null || n === void 0 ? void 0 : n.handlers) === null || r === void 0 || r.forEach((i) => {
      this._notifyPresenceSub(t, i);
    });
  }
  _notifyPresenceSub(t, n) {
    const r = this.getPresence("", t, n);
    r && (n.prev && !ou(r, n.prev) || (n.prev = r, n.cb(r)));
  }
  _patchPresencePeers(t, n) {
    var r, i, s;
    const o = ((i = (r = this._presence[t]) === null || r === void 0 ? void 0 : r.result) === null || i === void 0 ? void 0 : i.peers) || {};
    let u = Object.fromEntries(Object.entries(o).map(([d, c]) => [d, { data: c }]));
    (s = this._presence[t]) === null || s === void 0 || s.result;
    const a = qt(u, (d) => {
      for (let [c, f, h] of n)
        switch (f) {
          case "+":
            Ni(d, c, h);
            break;
          case "r":
            Fn(d, c, h);
            break;
          case "-":
            Or(d, c);
            break;
        }
      delete d[this._sessionId];
    });
    this._setPresencePeers(t, a);
  }
  _setPresencePeers(t, n) {
    const r = Object.assign({}, n);
    delete r[this._sessionId];
    const i = Object.fromEntries(Object.entries(r).map(([s, o]) => [s, o.data]));
    this._presence = qt(this._presence, (s) => {
      Fn(s, [t, "result", "peers"], i);
    });
  }
  // --------
  // Broadcast
  publishTopic({ roomType: t, roomId: n, topic: r, data: i }) {
    var s;
    const o = this._rooms[n];
    if (o) {
      if (!o.isConnected) {
        this._broadcastQueue[n] = (s = this._broadcastQueue[n]) !== null && s !== void 0 ? s : [], this._broadcastQueue[n].push({ topic: r, roomType: t, data: i });
        return;
      }
      this._tryBroadcast(n, t, r, i);
    }
  }
  _tryBroadcast(t, n, r, i) {
    this._trySendAuthed(V(), {
      op: "client-broadcast",
      "room-id": t,
      roomType: n,
      topic: r,
      data: i
    });
  }
  subscribeTopic(t, n, r) {
    const i = this.joinRoom(t);
    return this._broadcastSubs[t] = this._broadcastSubs[t] || {}, this._broadcastSubs[t][n] = this._broadcastSubs[t][n] || [], this._broadcastSubs[t][n].push(r), this._presence[t] = this._presence[t] || {}, () => {
      this._broadcastSubs[t][n] = this._broadcastSubs[t][n].filter((s) => s !== r), this._broadcastSubs[t][n].length || delete this._broadcastSubs[t][n], i();
    };
  }
  _notifyBroadcastSubs(t, n, r) {
    var i, s, o;
    (o = (s = (i = this._broadcastSubs) === null || i === void 0 ? void 0 : i[t]) === null || s === void 0 ? void 0 : s[n]) === null || o === void 0 || o.forEach((u) => {
      var a, d, c, f, h, p;
      const b = (a = r.data) === null || a === void 0 ? void 0 : a.data, g = r.data["peer-id"] === this._sessionId ? (c = (d = this._presence[t]) === null || d === void 0 ? void 0 : d.result) === null || c === void 0 ? void 0 : c.user : (p = (h = (f = this._presence[t]) === null || f === void 0 ? void 0 : f.result) === null || h === void 0 ? void 0 : h.peers) === null || p === void 0 ? void 0 : p[r.data["peer-id"]];
      return u(b, g);
    });
  }
  // --------
  // Storage
  uploadFile(t, n, r) {
    return D(this, void 0, void 0, function* () {
      var i;
      const s = yield this.getCurrentUser(), o = (i = s == null ? void 0 : s.user) === null || i === void 0 ? void 0 : i.refresh_token;
      return eu(Object.assign(Object.assign({}, r), { apiURI: this.config.apiURI, appId: this.config.appId, path: t, file: n, refreshToken: o }));
    });
  }
  deleteFile(t) {
    return D(this, void 0, void 0, function* () {
      var n;
      const r = yield this.getCurrentUser(), i = (n = r == null ? void 0 : r.user) === null || n === void 0 ? void 0 : n.refresh_token;
      return yield tu({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        path: t,
        refreshToken: i
      });
    });
  }
  // Deprecated Storage API (Jan 2025)
  // ---------------------------------
  upload(t, n) {
    return D(this, void 0, void 0, function* () {
      var r;
      const i = yield this.getCurrentUser(), s = (r = i == null ? void 0 : i.user) === null || r === void 0 ? void 0 : r.refresh_token, o = t || n.name, u = yield nu({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        fileName: o,
        refreshToken: s
      });
      return yield ru(u, n);
    });
  }
  getDownloadUrl(t) {
    return D(this, void 0, void 0, function* () {
      var n;
      const r = yield this.getCurrentUser(), i = (n = r == null ? void 0 : r.user) === null || n === void 0 ? void 0 : n.refresh_token;
      return yield iu({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        path: t,
        refreshToken: i
      });
    });
  }
}
let ot;
function Pu(e, t) {
  ot == null || ot.dispose();
  const n = Du(t), r = Ru(t, u), i = Iu(ku(e));
  function s(c) {
    var f;
    c.source === i.element.contentWindow && ((f = c.data) === null || f === void 0 ? void 0 : f.type) === "close" && n.isVisible() && u();
  }
  function o(c) {
    const f = c.shiftKey && c.ctrlKey && c.key === "0", h = c.key === "Escape" || c.key === "Esc";
    (f || h && n.isVisible()) && u();
  }
  function u() {
    n.isVisible() ? n.element.style.display = "none" : (n.element.style.display = "block", n.element.contains(i.element) || n.element.appendChild(i.element));
  }
  function a() {
    n.element.remove(), r.element.remove(), removeEventListener("keydown", o), removeEventListener("message", s);
  }
  function d() {
    document.body.appendChild(n.element), document.body.appendChild(r.element), addEventListener("keydown", o), addEventListener("message", s), ot = {
      dispose: a
    };
  }
  return d();
}
function ku(e) {
  return `${bn || Wr ? "http://localhost:3000" : "https://instantdb.com"}/_devtool?appId=${e}`;
}
function Iu(e) {
  const t = document.createElement("iframe");
  return t.src = e, t.className = "instant-devtool-iframe", Object.assign(t.style, {
    width: "100%",
    height: "100%",
    backgroundColor: "white",
    border: "none"
  }), { element: t };
}
function Ru(e, t) {
  const n = `
    <svg width="32" height="32" viewBox="0 0 512 512" fill="none" xmlns="http://www.w3.org/2000/svg">
      <rect width="512" height="512" fill="black"/>
      <rect x="97.0973" y="91.3297" width="140" height="330" fill="white"/>
    </svg>
  `, r = document.createElement("button");
  return r.innerHTML = n, r.className = "instant-devtool-toggler", Object.assign(r.style, Object.assign(Object.assign({
    // pos
    position: "fixed"
  }, Mu(e.position)), {
    height: "32px",
    width: "32px",
    // layout
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    zIndex: "9010",
    // look
    padding: "0",
    margin: "0",
    border: "none",
    cursor: "pointer"
  })), r.addEventListener("click", t), { element: r };
}
function Mu(e) {
  switch (e) {
    case "bottom-left":
      return { bottom: "24px", left: "24px" };
    case "bottom-right":
      return { bottom: "24px", right: "24px" };
    case "top-right":
      return { top: "24px", right: "24px" };
    case "top-left":
      return { top: "24px", left: "24px" };
  }
}
function $u(e) {
  switch (e) {
    case "bottom-left":
      return { bottom: "24px", right: "24px", left: "60px", top: "72px" };
    case "bottom-right":
      return { bottom: "24px", left: "24px", right: "60px", top: "72px" };
    case "top-right":
      return { top: "24px", left: "24px", right: "60px", bottom: "72px" };
    case "top-left":
      return { top: "24px", right: "24px", left: "60px", bottom: "72px" };
  }
}
function Du(e) {
  const t = document.createElement("div");
  Object.assign(t.style, Object.assign(Object.assign({ position: "fixed" }, $u(e.position)), { display: "block", borderRadius: "4px", border: "1px #ccc solid", boxShadow: "0px 0px 8px #00000044", backgroundColor: "#eee", zIndex: "999990" })), t.style.display = "none", t.className = "instant-devtool-container";
  function n() {
    return t.style.display !== "none";
  }
  return { element: t, isVisible: n };
}
const xu = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
};
function Uu() {
  var e;
  return globalThis.__instantDbSchemaHashStore = (e = globalThis.__instantDbSchemaHashStore) !== null && e !== void 0 ? e : /* @__PURE__ */ new WeakMap(), globalThis.__instantDbSchemaHashStore;
}
function Lu() {
  var e;
  return globalThis.__instantDbStore = (e = globalThis.__instantDbStore) !== null && e !== void 0 ? e : {}, globalThis.__instantDbStore;
}
function en(e) {
  const t = e.__adminToken;
  return e.appId + "_" + (e.websocketURI || "default_ws_uri") + "_" + (e.apiURI || "default_api_uri") + "_" + (t || "client_only") + "_" + e.useDateObjects;
}
const tn = Lu(), Mt = Uu();
class Fu {
  constructor(t) {
    this.db = t, this.sendMagicCode = (n) => this.db.sendMagicCode(n), this.signInWithMagicCode = (n) => this.db.signInWithMagicCode(n), this.signInWithToken = (n) => this.db.signInWithCustomToken(n), this.signInAsGuest = () => this.db.signInAsGuest(), this.createAuthorizationURL = (n) => this.db.createAuthorizationURL(n), this.signInWithIdToken = (n) => this.db.signInWithIdToken(n), this.exchangeOAuthCode = (n) => this.db.exchangeCodeForToken(n), this.issuerURI = () => this.db.issuerURI(), this.signOut = (n = { invalidateToken: !0 }) => this.db.signOut(n);
  }
}
class Nu {
  constructor(t) {
    this.db = t, this.uploadFile = (n, r, i = {}) => this.db.uploadFile(n, r, i), this.delete = (n) => this.db.deleteFile(n), this.upload = (n, r) => this.db.upload(n, r), this.put = this.upload, this.getDownloadUrl = (n) => this.db.getDownloadUrl(n);
  }
}
function Vu(e) {
  return JSON.parse(JSON.stringify(e));
}
class qu {
  constructor(t) {
    this.tx = yn(), this._reactor = t, this.auth = new Fu(this._reactor), this.storage = new Nu(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(db.tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(db.tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(db.tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    db.tx.todos[todoId].update({ title: 'Go on a run' }),
   *    db.tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(t) {
    return this._reactor.pushTx(t);
  }
  getLocalId(t) {
    return this._reactor.getLocalId(t);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(t, n, r) {
    return this._reactor.subscribeQuery(t, n, r);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(t) {
    return this._reactor.subscribeAuth(t);
  }
  /**
   * One time query for the logged in state. This is useful
   * for scenarios where you want to know the current auth
   * state without subscribing to changes.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const user = await db.getAuth();
   *   console.log('logged in as', user.email)
   */
  getAuth() {
    return this._reactor.getAuth();
  }
  /**
   * Listen for connection status changes to Instant. This is useful
   * for building things like connectivity indicators
   *
   * @see https://www.instantdb.com/docs/patterns#connection-status
   * @example
   *   const unsub = db.subscribeConnectionStatus((status) => {
   *     const connectionState =
   *       status === 'connecting' || status === 'opened'
   *         ? 'authenticating'
   *       : status === 'authenticated'
   *         ? 'connected'
   *       : status === 'closed'
   *         ? 'closed'
   *       : status === 'errored'
   *         ? 'errored'
   *       : 'unexpected state';
   *
   *     console.log('Connection status:', connectionState);
   *   });
   */
  subscribeConnectionStatus(t) {
    return this._reactor.subscribeConnectionStatus(t);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(t = "_defaultRoomType", n = "_defaultRoomId", r) {
    return {
      leaveRoom: this._reactor.joinRoom(n, r == null ? void 0 : r.initialPresence),
      subscribeTopic: (s, o) => this._reactor.subscribeTopic(n, s, o),
      subscribePresence: (s, o) => this._reactor.subscribePresence(t, n, s, o),
      publishTopic: (s, o) => this._reactor.publishTopic({ roomType: t, roomId: n, topic: s, data: o }),
      publishPresence: (s) => this._reactor.publishPresence(t, n, s),
      getPresence: (s) => this._reactor.getPresence(t, n, s)
    };
  }
  shutdown() {
    delete tn[en(this._reactor.config)], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(t, n) {
    return this._reactor.queryOnce(t, n);
  }
}
function fr(e) {
  if (!e)
    return "0";
  if (Mt.get(e))
    return Mt.get(e);
  const t = G(e);
  return Mt.set(e, t), t;
}
function Wu(e, t) {
  return fr(e._reactor.config.schema) !== fr(t);
}
function Ku(e, t, n, r, i) {
  const s = tn[en(e)];
  if (s)
    return Wu(s, e.schema) && s._reactor.updateSchema(e.schema), s;
  const o = new ju(Object.assign(Object.assign(Object.assign({}, xu), e), { cardinalityInference: !!e.schema }), t || Nr, n || Vr, Object.assign(Object.assign({}, r || {}), { "@instantdb/core": zr }), i), u = new qu(o);
  return tn[en(e)] = u, zu(e.appId, e.devtool), u;
}
function zu(e, t) {
  if (typeof window > "u" || typeof window.location > "u" || typeof document > "u" || typeof t == "boolean" && !t)
    return;
  const n = Object.assign({ position: "bottom-right", allowedHosts: ["localhost"] }, typeof t == "object" ? t : {});
  n.allowedHosts.includes(window.location.hostname) && Pu(e, n);
}
const Bu = {
  isLoading: !0,
  data: void 0,
  pageInfo: void 0,
  error: void 0
};
function lr(e) {
  return {
    isLoading: !e,
    data: void 0,
    pageInfo: void 0,
    error: void 0,
    ...e || {}
  };
}
function Qu(e, t, n) {
  t && n && "ruleParams" in n && (t = { $$ruleParams: n.ruleParams, ...t });
  const r = t ? Vu(t) : null, i = G(r), s = dt(
    lr(e._reactor.getPreviousResult(r))
  ), o = ye(
    (a) => (s.current = lr(
      e._reactor.getPreviousResult(r)
    ), r ? e.subscribeQuery(r, (c) => {
      s.current = {
        isLoading: !c,
        data: void 0,
        pageInfo: void 0,
        error: void 0,
        ...c
      }, a();
    }) : () => {
    }),
    // Build a new subscribe function if the query changes
    [i]
  );
  return { state: $t(
    o,
    () => s.current,
    () => Bu
  ), query: r };
}
function Gu() {
  const e = dt(null);
  we(() => {
    n();
  }, []);
  function t(r, i) {
    clearTimeout(e.current), e.current = setTimeout(i, r);
  }
  function n() {
    clearTimeout(e.current);
  }
  return { set: t, clear: n };
}
const Yu = 1e3;
function Hu(e, t, n) {
  we(() => e.core._reactor.subscribeTopic(
    e.id,
    t,
    (i, s) => {
      n(i, s);
    }
  ), [e.id, t]);
}
function Ju(e, t) {
  return we(() => e.core._reactor.joinRoom(e.id), [e.id]), ye(
    (r) => {
      e.core._reactor.publishTopic({
        roomType: e.type,
        roomId: e.id,
        topic: t,
        data: r
      });
    },
    [e.id, t]
  );
}
function Zu(e, t = {}) {
  var o, u;
  const [n, r] = pr(() => e.core._reactor.getPresence(e.type, e.id, t) ?? {
    peers: {},
    isLoading: !0
  });
  we(() => e.core._reactor.subscribePresence(
    e.type,
    e.id,
    t,
    (d) => {
      r(d);
    }
  ), [e.id, t.user, (o = t.peers) == null ? void 0 : o.join(), (u = t.keys) == null ? void 0 : u.join()]);
  const i = ye(
    (a) => {
      e.core._reactor.publishPresence(e.type, e.id, a);
    },
    [e.type, e.id]
  );
  return Dt(() => ({
    ...n,
    publishPresence: i
  }), [n, i]);
}
function Xu(e, t, n) {
  we(() => e.core._reactor.joinRoom(e.id, t), [e.id]), we(() => e.core._reactor.publishPresence(e.type, e.id, t), [e.type, e.id, n ?? JSON.stringify(t)]);
}
function ea(e, t, n = {}) {
  const r = Gu(), i = _e.usePresence(e, {
    keys: [t]
  }), s = Dt(() => {
    const c = e._core._reactor.getPresence(
      e.type,
      e.id
    );
    return n != null && n.writeOnly ? [] : Object.values((c == null ? void 0 : c.peers) ?? {}).filter(
      (f) => f[t] === !0
    );
  }, [n == null ? void 0 : n.writeOnly, i]), o = ye(
    (c) => {
      e.core._reactor.publishPresence(e.type, e.id, {
        [t]: c
      }), c && ((n == null ? void 0 : n.timeout) === null || (n == null ? void 0 : n.timeout) === 0 || r.set((n == null ? void 0 : n.timeout) ?? Yu, () => {
        e.core._reactor.publishPresence(e.type, e.id, {
          [t]: null
        });
      }));
    },
    [e.type, e.id, t, n == null ? void 0 : n.timeout, r]
  ), u = ye(
    (c) => {
      const h = !((n == null ? void 0 : n.stopOnEnter) && c.key === "Enter");
      o(h);
    },
    [n.stopOnEnter, o]
  ), a = ye(() => {
    o(!1);
  }, [o]), d = Dt(() => ({ onKeyDown: u, onBlur: a }), [u, a]);
  return {
    active: s,
    setActive: o,
    inputProps: d
  };
}
const _e = {
  useTopicEffect: Hu,
  usePublishTopic: Ju,
  usePresence: Zu,
  useSyncPresence: Xu,
  useTypingIndicator: ea
};
class ta {
  constructor(t, n, r) {
    this.useTopicEffect = (i, s) => {
      _e.useTopicEffect(this, i, s);
    }, this.usePublishTopic = (i) => _e.usePublishTopic(this, i), this.usePresence = (i = {}) => _e.usePresence(this, i), this.useSyncPresence = (i, s) => _e.useSyncPresence(this, i, s), this.useTypingIndicator = (i, s = {}) => _e.useTypingIndicator(this, i, s), this.core = t, this._core = this.core, this.type = n, this.id = r;
  }
}
const na = {
  isLoading: !0,
  user: void 0,
  error: void 0
};
class ia {
  constructor(t, n) {
    this.tx = yn(), this.getLocalId = (r) => this.core.getLocalId(r), this.useLocalId = (r) => {
      const [i, s] = pr(null);
      return we(() => {
        (async () => {
          const u = await this.getLocalId(r);
          s(u);
        })();
      }, [r]), i;
    }, this.rooms = _e, this.transact = (r) => this.core.transact(r), this.useQuery = (r, i) => Qu(
      this.core,
      r,
      i
    ).state, this.useAuth = () => {
      const r = dt(
        this.core._reactor._currentUserCached
      ), i = ye((o) => this.core.subscribeAuth((a) => {
        r.current = { isLoading: !1, ...a }, o();
      }), []);
      return $t(
        i,
        () => r.current,
        () => na
      );
    }, this.useUser = () => {
      const { user: r } = this.useAuth();
      if (!r)
        throw new ke(
          "useUser must be used within an auth-protected route"
        );
      return r;
    }, this.useConnectionStatus = () => {
      const r = dt(
        this.core._reactor.status
      ), i = ye((o) => this.core.subscribeConnectionStatus((a) => {
        a !== r.current && (r.current = a, o());
      }), []);
      return $t(
        i,
        () => r.current,
        // For SSR, always return 'connecting' as the initial state
        () => "connecting"
      );
    }, this.queryOnce = (r, i) => this.core.queryOnce(r, i), this.SignedIn = ({ children: r }) => {
      const i = this.useAuth();
      return i.isLoading || i.error || !i.user ? null : /* @__PURE__ */ rt.jsx(rt.Fragment, { children: r });
    }, this.SignedOut = ({ children: r }) => {
      const i = this.useAuth();
      return i.isLoading || i.error || i.user ? null : /* @__PURE__ */ rt.jsx(rt.Fragment, { children: r });
    }, this.core = Ku(
      t,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.Storage,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.NetworkListener,
      n,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.EventSourceImpl
    ), this._core = this.core, this.auth = this.core.auth, this.storage = this.core.storage;
  }
  /**
   * Obtain a handle to a room, which allows you to listen to topics and presence data
   *
   * If you don't provide a `type` or `id`, Instant will default to `_defaultRoomType` and `_defaultRoomId`
   * as the room type and id, respectively.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   *
   * @example
   *  const room = db.room('chat', roomId);
   *  const { peers } = db.rooms.usePresence(room);
   */
  room(t = "_defaultRoomType", n = "_defaultRoomId") {
    return new ta(this.core, t, n);
  }
  /**
   * One time query for the logged in state. This is useful
   * for scenarios where you want to know the current auth
   * state without subscribing to changes.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const user = await db.getAuth();
   *   console.log('logged in as', user.email)
   */
  getAuth() {
    return this.core.getAuth();
  }
}
export {
  ia as InstantReactAbstractDatabase
};
